{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "de527806-9c1e-4476-af62-2e206d1cc755",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Core Tree:\n",
      "       18          \n",
      "  ┏━━━━━┻━━━━━┓    \n",
      " 17           ┃    \n",
      " ┏┻━┓         ┃    \n",
      "16  ┃         ┃    \n",
      "┏┻┓ ┃         ┃    \n",
      "┃ ┃ ┃        15    \n",
      "┃ ┃ ┃       ┏━┻━━┓ \n",
      "┃ ┃ ┃      14    ┃ \n",
      "┃ ┃ ┃     ┏━┻━┓  ┃ \n",
      "┃ ┃ ┃    13   ┃  ┃ \n",
      "┃ ┃ ┃   ┏━┻━┓ ┃  ┃ \n",
      "┃ ┃ ┃  12   ┃ ┃  ┃ \n",
      "┃ ┃ ┃  ┏┻━┓ ┃ ┃  ┃ \n",
      "┃ ┃ ┃  ┃  ┃ ┃ ┃ 11 \n",
      "┃ ┃ ┃  ┃  ┃ ┃ ┃ ┏┻┓\n",
      "┃ ┃ ┃ 10  ┃ ┃ ┃ ┃ ┃\n",
      "┃ ┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃\n",
      "0 8 3 1 5 9 4 2 6 7\n",
      "\n",
      "                               70                         \n",
      "              ┏━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━┓      \n",
      "              ┃                                   69      \n",
      "              ┃                              ┏━━━━━┻━━━━━┓\n",
      "              ┃                             68           ┃\n",
      "              ┃                        ┏━━━━━┻━━━━━┓     ┃\n",
      "              ┃                        ┃           ┃    67\n",
      "              ┃                        ┃           ┃     ┃\n",
      "              ┃                       65           ┃    66\n",
      "              ┃                        ┃           ┃      \n",
      "             64                        ┃           ┃      \n",
      "          ┏━━━┻━━━┓                    ┃           ┃      \n",
      "          ┃       ┃                    ┃          63      \n",
      "          ┃       ┃                    ┃        ┏━━┻━━┓   \n",
      "          ┃       ┃                    ┃        ┃    62   \n",
      "          ┃       ┃                    ┃        ┃     ┃   \n",
      "          ┃       ┃                   60        ┃    61   \n",
      "          ┃       ┃                    ┃        ┃         \n",
      "          ┃       ┃                    ┃       59         \n",
      "          ┃       ┃                    ┃      ┏━┻━━┓      \n",
      "          ┃       ┃                   58      ┃    ┃      \n",
      "          ┃       ┃                 ┏━━┻━━┓   ┃    ┃      \n",
      "          ┃       ┃                 ┃     ┃  57    ┃      \n",
      "          ┃       ┃                 ┃     ┃  ┏┻━┓  ┃      \n",
      "          ┃       ┃                 ┃     ┃  ┃ 56  ┃      \n",
      "          ┃       ┃                 ┃     ┃  ┃  ┃  ┃      \n",
      "         54       ┃                 ┃     ┃  ┃ 55  ┃      \n",
      "          ┃       ┃                 ┃     ┃  ┃     ┃      \n",
      "          ┃       ┃                 ┃    53  ┃     ┃      \n",
      "          ┃       ┃                 ┃     ┃  ┃     ┃      \n",
      "          ┃       ┃                 ┃    52 51     ┃      \n",
      "          ┃       ┃                 ┃        ┃     ┃      \n",
      "          ┃       ┃                50        ┃     ┃      \n",
      "          ┃       ┃             ┏━━━┻━━┓     ┃     ┃      \n",
      "          ┃       ┃             ┃     49     ┃     ┃      \n",
      "          ┃       ┃             ┃      ┃     ┃     ┃      \n",
      "          ┃       ┃             ┃     48    47     ┃      \n",
      "          ┃       ┃             ┃            ┃     ┃      \n",
      "          ┃       ┃             ┃            ┃    46      \n",
      "          ┃       ┃             ┃            ┃     ┃      \n",
      "         44       ┃             ┃            ┃    45      \n",
      "          ┃       ┃             ┃            ┃            \n",
      "          ┃       ┃             ┃           43            \n",
      "          ┃       ┃             ┃            ┃            \n",
      "          ┃       ┃            41           42            \n",
      "          ┃       ┃             ┃                         \n",
      "          ┃       ┃            40                         \n",
      "          ┃       ┃          ┏━━┻━━━┓                     \n",
      "          ┃       ┃         39      ┃                     \n",
      "          ┃       ┃       ┏━━┻━━┓   ┃                     \n",
      "          ┃       ┃      38     ┃   ┃                     \n",
      "          ┃       ┃     ┏━┻┓    ┃   ┃                     \n",
      "          ┃       ┃     ┃  ┃    ┃  37                     \n",
      "          ┃       ┃     ┃  ┃    ┃   ┃                     \n",
      "         35       ┃     ┃  ┃    ┃  36                     \n",
      "          ┃       ┃     ┃  ┃    ┃                         \n",
      "          ┃       ┃     ┃ 34    ┃                         \n",
      "          ┃       ┃     ┃  ┃    ┃                         \n",
      "          ┃       ┃    32 33    ┃                         \n",
      "          ┃       ┃     ┃       ┃                         \n",
      "          ┃      31     ┃       ┃                         \n",
      "          ┃      ┏┻━┓   ┃       ┃                         \n",
      "          ┃      ┃  ┃   ┃      30                         \n",
      "          ┃      ┃  ┃   ┃     ┏━┻┓                        \n",
      "          ┃      ┃  ┃   ┃     ┃ 29                        \n",
      "          ┃      ┃  ┃   ┃     ┃  ┃                        \n",
      "          ┃      ┃  ┃  27     ┃ 28                        \n",
      "          ┃      ┃  ┃   ┃     ┃                           \n",
      "          ┃      ┃  ┃   ┃    26                           \n",
      "          ┃      ┃  ┃   ┃     ┃                           \n",
      "          ┃      ┃  ┃  24    25                           \n",
      "          ┃      ┃  ┃   ┃                                 \n",
      "          ┃      ┃  ┃  23                                 \n",
      "          ┃      ┃  ┃   ┃                                 \n",
      "          ┃     21  ┃  22                                 \n",
      "          ┃      ┃  ┃                                     \n",
      "         20      ┃  ┃                                     \n",
      "    ┏━━━━━┻━━━━┓ ┃  ┃                                     \n",
      "    ┃          ┃ ┃ 19                                     \n",
      "    ┃          ┃ ┃ ┏┻┓                                    \n",
      "   18          ┃ ┃ ┃ ┃                                    \n",
      " ┏━━┻━┓        ┃ ┃ ┃ ┃                                    \n",
      " ┃    ┃       17 ┃ ┃ ┃                                    \n",
      " ┃    ┃        ┃ ┃ ┃ ┃                                    \n",
      " ┃    ┃       16 ┃ ┃15                                    \n",
      " ┃    ┃          ┃ ┃ ┃                                    \n",
      " ┃   14          ┃ ┃ ┃                                    \n",
      " ┃  ┏━┻━┓        ┃ ┃ ┃                                    \n",
      " ┃  ┃  13        ┃ ┃ ┃                                    \n",
      " ┃  ┃ ┏━┻━┓      ┃ ┃ ┃                                    \n",
      " ┃  ┃ ┃  12      ┃ ┃ ┃                                    \n",
      " ┃  ┃ ┃ ┏━┻┓     ┃ ┃ ┃                                    \n",
      "11  ┃ ┃ ┃  ┃     ┃ ┃ ┃                                    \n",
      "┏┻┓ ┃ ┃ ┃  ┃     ┃ ┃ ┃                                    \n",
      "┃ ┃ ┃ ┃ ┃ 10     ┃ ┃ ┃                                    \n",
      "┃ ┃ ┃ ┃ ┃ ┏┻┓    ┃ ┃ ┃                                    \n",
      "0 1 2 3 4 5 6    7 8 9                                    \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\uhewm\\OneDrive\\PhD\\Project No.2\\pangenome-gene-transfer-simulation\\gene_model.py:229: RuntimeWarning: 4 double mutation (present -> present) occured. It is recommended to increase the num_sites to 10 or higher. Alternatily use the double_site_relocation or relocate_double_gene_gain option.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# We get the core tree from the data in panX later. Here we will simulate a random tree and assume it is the tree from panX.\n",
    "\n",
    "import msprime\n",
    "import tskit\n",
    "\n",
    "import torch\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "from random import randint\n",
    "\n",
    "import gene_model\n",
    "import gfs\n",
    "import color_scheme\n",
    "\n",
    "import hgt_simulation\n",
    "import hgt_mutations\n",
    "import hgt_sim_args\n",
    "from gfs import gfs_from_matrix\n",
    "\n",
    "# General parameters:\n",
    "\n",
    "theta = 2 # Gene Gain rate\n",
    "rho = 0.2 # Gene Loss rate\n",
    "gene_conv = 0 # Gene Conversion rate\n",
    "recomb = 0 # Recombination rate\n",
    "\n",
    "num_sites = 1 # Number of sites to simulate\n",
    "num_samples = 10 # Number of samples / individuals\n",
    "\n",
    "hgt_rate_max = 1 # Maximum hgt rate\n",
    "hgt_rate_min = 0 # Minimum hgt rate\n",
    "    \n",
    "\n",
    "# Simulate the core tree:\n",
    "\n",
    "core_tree = msprime.sim_ancestry(\n",
    "                samples=num_samples,\n",
    "                sequence_length=1,\n",
    "                ploidy=1,\n",
    "                recombination_rate=0,\n",
    "                gene_conversion_rate=0,\n",
    "                gene_conversion_tract_length=1,  # One gene\n",
    "            )\n",
    "\n",
    "# Visualization of core tree :\n",
    "\n",
    "print(\"Core Tree:\")\n",
    "print(core_tree.first().draw_text())\n",
    "    \n",
    "\n",
    "# Example: Add HGT events to the core tree:\n",
    "\n",
    "ce_from_nwk = core_tree.first().newick()\n",
    "ce_from_ts = core_tree\n",
    "\n",
    "mts = gene_model.gene_model(  # ACHTUNG: Sortierung der Blätter ändert sich!\n",
    "    theta=theta,\n",
    "    rho=rho,\n",
    "    gene_conversion_rate=gene_conv,\n",
    "    recombination_rate=recomb,\n",
    "    hgt_rate=1,\n",
    "    num_samples=num_samples,\n",
    "    num_sites=num_sites,\n",
    "    double_site_relocation=False, \n",
    "    ce_from_nwk = ce_from_nwk,\n",
    "    # ce_from_ts = ce_from_ts\n",
    ")\n",
    "\n",
    "print(mts.first().draw_text())\n",
    "\n",
    "# Distribution of hgt_rate:\n",
    "\n",
    "prior = torch.distributions.Uniform(low=hgt_rate_min, high=hgt_rate_max)  # The hgt_rate PER SITE (GENOME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1d9b78ac-90a4-422a-b5e1-a69cecc82d2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor(0.8741)\n"
     ]
    }
   ],
   "source": [
    "# Add HGT events to the core tree:\n",
    "\n",
    "length_distribution_simulation_size = 20\n",
    "\n",
    "hgt_rate = prior.sample()\n",
    "print(hgt_rate) \n",
    "\n",
    "mts_list = []\n",
    "for i in range (1, length_distribution_simulation_size):\n",
    "    ce_from_nwk = core_tree.first().newick()\n",
    "    mts = gene_model.gene_model(\n",
    "        theta=0, # Set to 0 since it does not matter here\n",
    "        rho=0,\n",
    "        gene_conversion_rate=0,\n",
    "        recombination_rate=0,\n",
    "        hgt_rate=hgt_rate.item(),\n",
    "        num_samples=num_samples,\n",
    "        num_sites=1,\n",
    "        double_site_relocation=False, \n",
    "        ce_from_nwk = ce_from_nwk,\n",
    "        # ce_from_ts = ce_from_ts\n",
    "    )\n",
    "    \n",
    "    mts_list.append(mts)\n",
    "\n",
    "total_branch_lengths = []\n",
    "for mts in mts_list:\n",
    "    total_length = sum(tree.total_branch_length for tree in mts.trees())\n",
    "    total_branch_lengths.append(total_length)\n",
    "\n",
    "average_tree_length = sum(total_branch_lengths) / len(total_branch_lengths)\n",
    "\n",
    "# Define the empirical distribution of the tree length:\n",
    "\n",
    "ecdf = ECDF(total_branch_lengths)\n",
    "\n",
    "def tree_length_distribution(x):\n",
    "    return ecdf.x[np.searchsorted(ecdf.y, x)]\n",
    "\n",
    "# Calculate the distribution of the number of mutations:\n",
    "\n",
    "expected_number_of_mutations = theta * average_tree_length\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "60e53203-c4c8-4c1a-b763-8930ad713329",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "# def weighted_random_selection(pool_of_trees, n):\n",
    "\n",
    "# Calculate the tree lengths\n",
    "tree_lengths = np.array([tree.total_branch_length for tree in pool_of_trees])\n",
    "\n",
    "# Make sure all lengths are positive\n",
    "if np.any(tree_lengths <= 0):\n",
    "    raise ValueError(\"Tree with length 0 found.\")\n",
    "\n",
    "# List for selected trees:\n",
    "selected_trees = []\n",
    "\n",
    "for i in range(number_of_gains):\n",
    "    # Calculate probabilities\n",
    "    probabilities = tree_lengths / tree_lengths.sum()\n",
    "    \n",
    "    # Choose an index based on the tree length\n",
    "    chosen_index = np.random.choice(len(pool_of_trees), p=probabilities)\n",
    "    selected_trees.append(pool_of_trees[chosen_index])\n",
    "    \n",
    "    # Entferne den ausgewählten Baum und seine Länge aus den Listen\n",
    "    pool_of_trees.pop(chosen_index)\n",
    "    tree_lengths = np.delete(tree_lengths, chosen_index)\n",
    "\n",
    "# return selected_trees\n",
    "\n",
    "print(np.sort([tree.total_branch_length for tree in pool_of_trees]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "860030f3-a3f4-49bf-aff4-3d494597e23a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[17, 31, 22, 15, 28, 44, 54, 38, 40, 47, 8, 48, 13, 79, 10, 27, 63, 66, 59, 60, 77, 68, 29, 78, 36, 64, 2]\n",
      "[75, 23, 46, 39, 76, 37, 51, 53, 61, 19, 30, 65, 25, 34, 52, 58, 24, 57, 69, 41, 72, 16, 56, 7, 14, 21, 35]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "number_of_simulations = 1 # Number of simulations of the gfs\n",
    "size_factor = 3 # Factor which describes how much more trees are produced compared to the needed amount.\n",
    "\n",
    "for i in range(0, number_of_simulations):\n",
    "    \n",
    "    # Würfle die gewünschte Zahl der Gains\n",
    "    \n",
    "    number_of_gains = np.random.poisson(expected_number_of_mutations)\n",
    "    \n",
    "    # Erstelle mit num_sites einen Baum der entsprechenden Größe mal size_factor.\n",
    "\n",
    "    num_size = number_of_gains * size_factor\n",
    "    \n",
    "    mts = gene_model.gene_model(\n",
    "        theta=0, # Set to 0 since it does not matter yet\n",
    "        rho=0,\n",
    "        gene_conversion_rate=0,\n",
    "        recombination_rate=0,\n",
    "        hgt_rate = hgt_rate.item(),\n",
    "        num_samples=num_samples,\n",
    "        num_sites=num_size,\n",
    "        double_site_relocation=False, \n",
    "        ce_from_nwk = ce_from_nwk,\n",
    "    )\n",
    "\n",
    "    # Calculate the tree lengths\n",
    "    tree_lengths = np.array([tree.total_branch_length for tree in mts.trees()])\n",
    "    tree_lengths_average = sum(tree_lengths)/len(tree_lengths)\n",
    "    \n",
    "    # Make sure all lengths are positive\n",
    "    if np.any(tree_lengths <= 0):\n",
    "        raise ValueError(\"Tree with length 0 found.\")\n",
    "    \n",
    "    selected_trees_indexes = []\n",
    "    deleted_trees_indexes = []\n",
    "    \n",
    "    for j in range(number_of_gains):\n",
    "        \n",
    "        probabilities = tree_lengths / tree_lengths.sum()\n",
    "        \n",
    "        chosen_index = np.random.choice(len(mts.trees()), p=probabilities)\n",
    "        while (chosen_index in selected_trees_indexes or chosen_index in deleted_trees_indexes):\n",
    "            chosen_index = np.random.choice(len(mts.trees()), p=probabilities)\n",
    "        \n",
    "        selected_trees_indexes.append(chosen_index)\n",
    "        \n",
    "        if (len(selected_trees_indexes)) >= len(mts.trees()):\n",
    "            raise ValueError(\"Not enough trees to draw.\")\n",
    "        \n",
    "        # Calculate the average of the remaining trees.\n",
    "        mask = np.ones(len(tree_lengths), dtype=bool)\n",
    "        mask[selected_trees_indexes] = False\n",
    "        mask[deleted_trees_indexes] = False\n",
    "        current_average = sum(tree_lengths[mask])/len(tree_lengths[mask])\n",
    "        \n",
    "        # Remove an additional tree such that the mean of the tree length remains roughly the same.\n",
    "        target_diff = (current_average - tree_lengths_average)*(len(tree_lengths)-len(selected_trees_indexes)-len(deleted_trees_indexes)-1)\n",
    "        \n",
    "        differences = tree_lengths - tree_lengths_average - target_diff\n",
    "        differences[selected_trees_indexes] = np.inf\n",
    "        differences[deleted_trees_indexes] = np.inf\n",
    "        \n",
    "        closest_index = np.argmin(np.abs(differences))\n",
    "        deleted_trees_indexes.append(closest_index)\n",
    "\n",
    "    print(selected_trees_indexes)\n",
    "    print(deleted_trees_indexes)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "f9f84be2-c212-4d92-83e5-47240cdc4664",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[50, 46, 40, 55, 83, 4, 72, 33, 30, 22, 58, 8, 13, 70, 88, 12, 18, 87, 78, 51, 14, 23, 59, 73, 15, 92, 37, 7, 76, 75, 35]\n",
      "[16, 81, 42, 3, 45, 0, 65, 60, 20, 47, 57, 80, 68, 66, 24, 82, 32, 71, 91, 52, 19, 17, 2, 64, 34, 49, 86, 29, 69, 67, 44]\n"
     ]
    }
   ],
   "source": [
    "from random import randint\n",
    "\n",
    "number_of_simulations = 1 # Number of simulations of the gfs\n",
    "size_factor = 3 # Factor which describes how much more trees are produced compared to the needed amount.\n",
    "\n",
    "for i in range(0, number_of_simulations):\n",
    "    \n",
    "    # Würfle die gewünschte Zahl der Gains\n",
    "    \n",
    "    number_of_gains = np.random.poisson(expected_number_of_mutations)\n",
    "    \n",
    "    # Erstelle mit num_sites einen Baum der entsprechenden Größe mal size_factor.\n",
    "\n",
    "    num_size = number_of_gains * size_factor\n",
    "\n",
    "    args = hgt_sim_args.Args(\n",
    "                sample_size=num_samples,\n",
    "                num_sites=num_size,\n",
    "                gene_conversion_rate=0,\n",
    "                recombination_rate=0,\n",
    "                hgt_rate=hgt_rate.item(),\n",
    "                ce_from_ts=None,\n",
    "                ce_from_nwk=ce_from_nwk,\n",
    "                random_seed=randint(1, int(2**32 - 2)),\n",
    "            )\n",
    "\n",
    "    ts, hgt_edges = hgt_simulation.run_simulate(args)\n",
    "\n",
    "    # Calculate the tree lengths\n",
    "    tree_lengths = np.array([tree.total_branch_length for tree in ts.trees()])\n",
    "    tree_lengths_average = sum(tree_lengths)/len(tree_lengths)\n",
    "    \n",
    "    # Make sure all lengths are positive\n",
    "    if np.any(tree_lengths <= 0):\n",
    "        raise ValueError(\"Tree with length 0 found.\")\n",
    "    \n",
    "    selected_trees_indexes = []\n",
    "    deleted_trees_indexes = []\n",
    "    \n",
    "    for j in range(number_of_gains):\n",
    "        \n",
    "        probabilities = tree_lengths / tree_lengths.sum()\n",
    "        \n",
    "        chosen_index = np.random.choice(len(ts.trees()), p=probabilities)\n",
    "        while (chosen_index in selected_trees_indexes or chosen_index in deleted_trees_indexes):\n",
    "            chosen_index = np.random.choice(len(ts.trees()), p=probabilities)\n",
    "        \n",
    "        selected_trees_indexes.append(chosen_index)\n",
    "        \n",
    "        if (len(selected_trees_indexes)) >= len(ts.trees()):\n",
    "            raise ValueError(\"Not enough trees to draw.\")\n",
    "        \n",
    "        # Calculate the average of the remaining trees.\n",
    "        mask = np.ones(len(tree_lengths), dtype=bool)\n",
    "        mask[selected_trees_indexes] = False\n",
    "        mask[deleted_trees_indexes] = False\n",
    "        current_average = sum(tree_lengths[mask])/len(tree_lengths[mask])\n",
    "        \n",
    "        # Remove an additional tree such that the mean of the tree length remains roughly the same.\n",
    "        target_diff = (current_average - tree_lengths_average)*(len(tree_lengths)-len(selected_trees_indexes)-len(deleted_trees_indexes)-1)\n",
    "        \n",
    "        differences = tree_lengths - tree_lengths_average - target_diff\n",
    "        differences[selected_trees_indexes] = np.inf\n",
    "        differences[deleted_trees_indexes] = np.inf\n",
    "        \n",
    "        closest_index = np.argmin(np.abs(differences))\n",
    "        deleted_trees_indexes.append(closest_index)\n",
    "\n",
    "    print(selected_trees_indexes)\n",
    "    print(deleted_trees_indexes)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "9b61c88d-6f73-453f-9ff1-d5e49a794bac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<hgt_mutations.HGTMutationGenerator at 0x219b77f26f0>"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alleles = [\"absent\", \"present\"]\n",
    "\n",
    "theta_total_events = theta\n",
    "rho_total_events = rho * num_sites\n",
    "\n",
    "root_proba = theta_total_events / (rho_total_events if rho_total_events != 0 else theta)\n",
    "if not (0 <= root_proba <= 1):\n",
    "    raise ValueError(f\"Invalid theta / rho resulting in a root probability of {root_proba}\")\n",
    "\n",
    "event_rate = rho_total_events + theta_total_events\n",
    "theta_proba = theta_total_events / event_rate\n",
    "rho_proba = rho_total_events / event_rate\n",
    "event_rate /= num_sites\n",
    "\n",
    "gain_loss_model = msprime.MatrixMutationModel(\n",
    "    alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [1 - theta_proba, theta_proba],\n",
    "        [rho_proba, 1 - rho_proba],\n",
    "    ],\n",
    ")\n",
    "\n",
    "tables = ts.dump_tables()\n",
    "\n",
    "# Set the ancestral state for each site.\n",
    "# poisson = np.random.poisson(theta / (rho if rho != 0 else 1))\n",
    "poisson = 0\n",
    "poisson = min(poisson, num_sites)\n",
    "\n",
    "position = np.arange(0, num_sites, dtype=\"uint32\")\n",
    "position = np.random.choice(position, poisson, replace=False)\n",
    "position.sort()\n",
    "\n",
    "ancestral_state = [alleles[1]] * poisson\n",
    "ancestral_state, ancestral_state_offset = tskit.pack_strings(ancestral_state)\n",
    "\n",
    "tables.sites.set_columns(\n",
    "    position=position,\n",
    "    ancestral_state=ancestral_state,\n",
    "    ancestral_state_offset=ancestral_state_offset,\n",
    ")\n",
    "ts = tables.tree_sequence()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Anfang hgt_mutations.sim_mutations:\n",
    "\n",
    "tables = ts.dump_tables()\n",
    "\n",
    "gene_count = tables.sequence_length\n",
    "\n",
    "gain_loss_model = hgt_mutations.PythonMutationMatrixModel(\n",
    "    alleles=gain_loss_model.alleles,\n",
    "    root_distribution=gain_loss_model.root_distribution,\n",
    "    transition_matrix=gain_loss_model.transition_matrix,\n",
    ")\n",
    "\n",
    "rate_map = msprime.RateMap(position=[0, gene_count], rate=[event_rate])\n",
    "\n",
    "edges = list(tables.edges)\n",
    "edges.extend(hgt_edges)\n",
    "\n",
    "child_ids = {e.child for e in edges}\n",
    "root_nodes = list({e.parent for e in edges if e.parent not in child_ids})\n",
    "\n",
    "hgt_generator = hgt_mutations.HGTMutationGenerator(rate_map=rate_map, model=gain_loss_model)\n",
    "\n",
    "ts = hgt_generator.generate(\n",
    "    tables,\n",
    "    edges,\n",
    "    root_nodes,\n",
    "    randint(0, 4294967295),\n",
    "    keep=True,\n",
    "    discrete_genome=True,\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "hgt_generator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "a2ada6b5-de75-4175-9c51-7539ba6cbc5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3600, 3931]"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "cda23d5a-4b95-46f9-8236-12b153b1bd66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "[]\n",
      "0.625\n",
      "[]\n",
      "[0]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <div>\n",
       "            <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "            </style>\n",
       "            <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"padding:0;line-height:21px;\">\n",
       "                                    <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                                    <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                                </th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            <tr><td>Trees</td><td>93</td></tr>\n",
       "                            <tr><td>Sequence Length</td><td>93.0</td></tr>\n",
       "                            <tr><td>Time Units</td><td>unknown</td></tr>\n",
       "                            <tr><td>Sample Nodes</td><td>10</td></tr>\n",
       "                            <tr><td>Total Size</td><td>538.5 KiB</td></tr>\n",
       "                            <tr>\n",
       "                                <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td>\n",
       "                            </tr>\n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"line-height:21px;\">Table</th>\n",
       "                                <th>Rows</th>\n",
       "                                <th>Size</th>\n",
       "                                <th>Has Metadata</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "            <tr>\n",
       "                <td>Edges</td>\n",
       "                <td>11,031</td>\n",
       "                <td>344.7 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Individuals</td>\n",
       "                <td>0</td>\n",
       "                <td>24 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Migrations</td>\n",
       "                <td>0</td>\n",
       "                <td>8 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Mutations</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Nodes</td>\n",
       "                <td>3,932</td>\n",
       "                <td>107.5 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Populations</td>\n",
       "                <td>1</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Provenances</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Sites</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th>Provenance Timestamp</th>\n",
       "                                <th>Software Name</th>\n",
       "                                <th>Version</th>\n",
       "                                <th>Command</th>\n",
       "                                <th>Full record</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "            </div>\n",
       "        </div>\n",
       "    "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x219b9ecef60>"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_backup = ts\n",
    "\n",
    "num_sites_test = 10\n",
    "# Set the ancestral state for each site.\n",
    "poisson = np.random.poisson(theta / (rho if rho != 0 else 1))\n",
    "poisson = min(poisson, num_sites_test)\n",
    "\n",
    "position = np.arange(0, num_sites_test, dtype=\"uint32\")\n",
    "position = np.random.choice(position, poisson, replace=False)\n",
    "position.sort()\n",
    "\n",
    "ancestral_state = [alleles[1]] * poisson\n",
    "ancestral_state, ancestral_state_offset = tskit.pack_strings(ancestral_state)\n",
    "\n",
    "print(poisson)\n",
    "print(position)\n",
    "print((theta / (rho if rho != 0 else 1)))\n",
    "print(ancestral_state)\n",
    "print(ancestral_state_offset)\n",
    "\n",
    "ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "9ac9d364-c10f-4b57-9c46-9e507620a2fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Duplikate und ihre Häufigkeit: {}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "96"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Schritt 1: Sortiere die Ziehungen\n",
    "selected_trees_sorted = sorted(selected_trees)\n",
    "\n",
    "# Schritt 2: Zähle Duplikate\n",
    "from collections import Counter\n",
    "tree_counts = Counter(selected_trees_sorted)\n",
    "\n",
    "# Ausgabe der Anzahl der Duplikate\n",
    "duplicates = {key: count for key, count in tree_counts.items() if count > 1}\n",
    "print(f\"Duplikate und ihre Häufigkeit: {duplicates}\")\n",
    "\n",
    "# Schritt 3: Entferne Duplikate (lasse nur einen Eintrag pro Baum übrig)\n",
    "unique_trees = list(tree_counts.keys())\n",
    "\n",
    "len(mts.trees())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "7c8321d7-5499-48e9-9f76-fb82b9c97953",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method TreeSequence.mutations of <tskit.trees.TreeSequence object at 0x0000019C39A624B0>>\n",
      "[0 0 0 0 0 0 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "# Hole die Tabellen aus dem TreeSequence-Objekt\n",
    "tables = mts_list[1].dump_tables()\n",
    "\n",
    "# 1. Füge eine zufällige Position für die Mutation hinzu\n",
    "site_position = np.random.uniform(0, mts_list[1].sequence_length)  # Zufällige Position im Bereich [0, sequence_length]\n",
    "site_id = tables.sites.add_row(position=site_position, ancestral_state=\"0\")\n",
    "\n",
    "# 2. Wähle einen zufälligen Knoten aus\n",
    "node_id = np.random.choice(range(mts_list[1].num_nodes))\n",
    "\n",
    "# 3. Füge die Mutation hinzu\n",
    "tables.mutations.add_row(site=site_id, node=node_id, derived_state=\"1\")\n",
    "\n",
    "# 4. Erstelle eine neue TreeSequence mit der Mutation\n",
    "mutated_tree = tables.tree_sequence()\n",
    "\n",
    "\n",
    "for var in mutated_tree.variants():\n",
    "    print(var.genotypes)\n",
    "\n",
    "with open(\"mutated_tree.svg\", \"w\") as svg_file:\n",
    "    svg_file.write(mutated_tree.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8ad1e3cd-c3b3-45e4-8471-497eb615d7b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n"
     ]
    }
   ],
   "source": [
    "theta = 2\n",
    "expected_number_of_mutations = theta * average_tree_length\n",
    "number_of_gains = np.random.poisson(expected_number_of_mutations)\n",
    "print(number_of_gains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4b595183-b943-49b1-ab03-7fc055d95b3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[13, 3, 13, 2, 1, 1, 2, 0, 3, 2, 3, 1, 13, 0, 12, 0, 10, 1, 11]\n",
      "0.7355953140336466\n",
      "0.2\n"
     ]
    }
   ],
   "source": [
    "number_of_mutations = []\n",
    "for mts in mts_list:\n",
    "    mutations = sum(tree.num_mutations for tree in mts.trees())\n",
    "    number_of_mutations.append(mutations)\n",
    "\n",
    "print(number_of_mutations)\n",
    "print(expected_number_of_mutations)\n",
    "\n",
    "print(average_tree_length)\n",
    "print(theta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a6bb4d7b-ee30-4a0d-8478-4e7b4be6d728",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor(0.6767)\n",
      "4.338400242348786\n",
      "7.090114687665711\n",
      "[5.097504146075549, 6.94044197628146, 8.002941158939326, 6.601718861363333, 9.48883397139835, 6.661285208587734, 9.459366707173045, 7.953704524889576, 4.33840034234883, 6.368376507637036, 7.0150088747674815, 6.642537150789993, 5.517070292067897, 7.811889534543084, 7.046312126774953, 7.648810761033703, 10.495032947571657, 5.3083393713564515, 6.314604602049042]\n"
     ]
    }
   ],
   "source": [
    "print(hgt_rate)\n",
    "print(sum(tree.total_branch_length for tree in core_tree.trees()))\n",
    "print(average_tree_length)\n",
    "print(total_branch_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "51f906b4-46a8-42de-b7c9-111b00ea1a87",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_extinction_time(L, rho, hgt_rate, num_samples):\n",
    "    \"\"\"\n",
    "    Simuliert die Zeit bis zur Extinktion eines Zweigungsprozesses.\n",
    "    \n",
    "    Parameter:\n",
    "        L: Die Länge des Core-Baums (nicht direkt verwendet)\n",
    "        rho: Mutationsrate (nicht direkt verwendet)\n",
    "        gamma: HGT-Rate (Geburtenrate)\n",
    "        initial_branches: Anzahl der anfänglichen Zweige (X_0)\n",
    "\n",
    "    Rückgabewert:\n",
    "        Zeit bis zur Extinktion\n",
    "    \"\"\"\n",
    "    branches = num_samples\n",
    "    time = 0.0\n",
    "\n",
    "    while branches > 0:\n",
    "        birth_rate = hgt_rate\n",
    "        if branches > 1:\n",
    "            death_rate = branches * (branches - 1) / 2\n",
    "        else:\n",
    "            death_rate = 0\n",
    "\n",
    "        total_rate = birth_rate + death_rate\n",
    "        if total_rate == 0:\n",
    "            break\n",
    "\n",
    "        # Wartezeit bis zum nächsten Ereignis\n",
    "        time_to_next_event = np.random.exponential(1 / total_rate)\n",
    "        time += time_to_next_event\n",
    "\n",
    "        # Entscheide ob Geburt oder Tod\n",
    "        if np.random.rand() < (birth_rate / total_rate):\n",
    "            branches += 1\n",
    "        else:\n",
    "            branches -= 1\n",
    "\n",
    "    return time\n",
    "\n",
    "# Beispielhafte Parameter\n",
    "L = 100  # Beispielkernbaum bei Länge L (nicht direkt verwendet)\n",
    "rho = 0.01  # Mutationsrate (nicht direkt verwendet)\n",
    "gamma = 0.1  # HGT-Rate\n",
    "num_samples = 5  # Anzahl der anfänglichen Zweige\n",
    "\n",
    "# Berechne die Zeit bis zur Extinktion\n",
    "extinction_time = simulate_extinction_time(L, rho, gamma, num_samples)\n",
    "print(f\"Zeit bis zur Extinktion: {extinction_time:.2f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a32258ca-cdf9-4163-90b5-e2ec9034b03d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5.771128597498799, 5.4174097449121525, 6.77904869528533, 6.399279959065085, 8.679570304346026, 8.033348397032984, 9.038932700006043, 5.4174097449121525, 6.78024517220398, 6.613619376945763, 6.298587709217051, 7.107051383156005, 9.306812231266242, 9.747508725826629, 8.028013919945538, 6.422505126013231, 7.621816499411544, 7.5179369690924975, 8.666259532719694]\n",
      "   18              \n",
      " ┏━━┻━━┓           \n",
      " ┃    17           \n",
      " ┃  ┏━━┻━━┓        \n",
      " ┃  ┃    16        \n",
      " ┃  ┃  ┏━━┻━━━┓    \n",
      " ┃  ┃  ┃     15    \n",
      " ┃  ┃  ┃    ┏━┻━━┓ \n",
      " ┃  ┃  ┃    ┃   14 \n",
      " ┃  ┃  ┃    ┃   ┏┻┓\n",
      " ┃  ┃  ┃   13   ┃ ┃\n",
      " ┃  ┃  ┃  ┏━┻┓  ┃ ┃\n",
      " ┃  ┃ 12  ┃  ┃  ┃ ┃\n",
      " ┃  ┃ ┏┻┓ ┃  ┃  ┃ ┃\n",
      " ┃  ┃ ┃ ┃ ┃ 11  ┃ ┃\n",
      " ┃  ┃ ┃ ┃ ┃ ┏┻┓ ┃ ┃\n",
      "10  ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃\n",
      "┏┻┓ ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃\n",
      "0 1 2 3 4 5 6 7 8 9\n",
      "\n",
      "             18    \n",
      "           ┏━━┻━━┓ \n",
      "          17     ┃ \n",
      "        ┏━━┻━━┓  ┃ \n",
      "       16     ┃  ┃ \n",
      "    ┏━━━┻━━┓  ┃  ┃ \n",
      "   15      ┃  ┃  ┃ \n",
      " ┏━━┻━┓    ┃  ┃  ┃ \n",
      "14    ┃    ┃  ┃  ┃ \n",
      "┏┻┓   ┃    ┃  ┃  ┃ \n",
      "┃ ┃  13    ┃  ┃  ┃ \n",
      "┃ ┃  ┏┻━┓  ┃  ┃  ┃ \n",
      "┃ ┃  ┃  ┃ 12  ┃  ┃ \n",
      "┃ ┃  ┃  ┃ ┏┻┓ ┃  ┃ \n",
      "┃ ┃ 11  ┃ ┃ ┃ ┃  ┃ \n",
      "┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃  ┃ \n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃ 10 \n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┏┻┓\n",
      "0 6 1 5 4 3 7 2 8 9\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <div>\n",
       "            <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "            </style>\n",
       "            <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"padding:0;line-height:21px;\">\n",
       "                                    <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                                    <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                                </th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            <tr><td>Trees</td><td>1</td></tr>\n",
       "                            <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "                            <tr><td>Time Units</td><td>unknown</td></tr>\n",
       "                            <tr><td>Sample Nodes</td><td>10</td></tr>\n",
       "                            <tr><td>Total Size</td><td>2.8 KiB</td></tr>\n",
       "                            <tr>\n",
       "                                <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td>\n",
       "                            </tr>\n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"line-height:21px;\">Table</th>\n",
       "                                <th>Rows</th>\n",
       "                                <th>Size</th>\n",
       "                                <th>Has Metadata</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "            <tr>\n",
       "                <td>Edges</td>\n",
       "                <td>30</td>\n",
       "                <td>968 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Individuals</td>\n",
       "                <td>0</td>\n",
       "                <td>24 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Migrations</td>\n",
       "                <td>0</td>\n",
       "                <td>8 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Mutations</td>\n",
       "                <td>14</td>\n",
       "                <td>630 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    ✅\n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Nodes</td>\n",
       "                <td>31</td>\n",
       "                <td>876 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Populations</td>\n",
       "                <td>1</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Provenances</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Sites</td>\n",
       "                <td>1</td>\n",
       "                <td>46 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th>Provenance Timestamp</th>\n",
       "                                <th>Software Name</th>\n",
       "                                <th>Version</th>\n",
       "                                <th>Command</th>\n",
       "                                <th>Full record</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "            </div>\n",
       "        </div>\n",
       "    "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x19c3f296870>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(total_branch_lengths)\n",
    "print(mts_list[1].first().draw_text())\n",
    "print(core_tree.first().draw_text())\n",
    "\n",
    "mts"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pangenome-gene-transfer-simulation",
   "language": "python",
   "name": "pangenome-gene-transfer-simulation"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
