{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "de527806-9c1e-4476-af62-2e206d1cc755",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Core Tree:\n",
      "    8    \n",
      " ┏━━┻━┓  \n",
      " ┃    7  \n",
      " ┃  ┏━┻┓ \n",
      " ┃  ┃  6 \n",
      " ┃  ┃ ┏┻┓\n",
      " 5  ┃ ┃ ┃\n",
      "┏┻┓ ┃ ┃ ┃\n",
      "0 2 1 3 4\n",
      "\n",
      "                                     96                                   \n",
      "          ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━┓        \n",
      "         95                                                      ┃        \n",
      "    ┏━━━━━┻━━━━━━┓                                               ┃        \n",
      "   94            ┃                                               ┃        \n",
      " ┏━━┻━┓          ┃                                               ┃        \n",
      " ┃    ┃         93                                               ┃        \n",
      " ┃    ┃       ┏━━┻━━┓                                            ┃        \n",
      " ┃    ┃       ┃     ┃                                           92        \n",
      " ┃    ┃       ┃     ┃                                     ┏━━━━━━┻━━━━━━┓ \n",
      " ┃    ┃      91     ┃                                     ┃             ┃ \n",
      " ┃    ┃     ┏━┻━━┓  ┃                                     ┃             ┃ \n",
      " ┃    ┃     ┃   90  ┃                                     ┃             ┃ \n",
      " ┃    ┃     ┃    ┃  ┃                                     ┃             ┃ \n",
      " ┃    ┃     ┃   89  ┃                                     ┃            88 \n",
      " ┃    ┃     ┃       ┃                                     ┃             ┃ \n",
      " ┃    ┃     ┃      87                                     ┃             ┃ \n",
      " ┃    ┃     ┃       ┃                                     ┃             ┃ \n",
      " ┃    ┃     ┃      86                                     ┃            85 \n",
      " ┃    ┃     ┃                                             ┃             ┃ \n",
      " ┃    ┃     ┃                                            84             ┃ \n",
      " ┃    ┃     ┃                                      ┏━━━━━━┻━━━━━━━┓     ┃ \n",
      " ┃    ┃     ┃                                      ┃              ┃    83 \n",
      " ┃    ┃     ┃                                      ┃              ┃   ┏━┻┓\n",
      " ┃    ┃     ┃                                     82              ┃   ┃  ┃\n",
      " ┃    ┃     ┃                               ┏━━━━━━┻━━━━━━┓       ┃   ┃  ┃\n",
      " ┃    ┃     ┃                               ┃             ┃       ┃   ┃ 81\n",
      " ┃    ┃     ┃                               ┃             ┃       ┃   ┃  ┃\n",
      " ┃    ┃     ┃                               ┃             ┃       ┃  79 80\n",
      " ┃    ┃     ┃                               ┃             ┃       ┃   ┃   \n",
      " ┃    ┃    78                               ┃             ┃       ┃   ┃   \n",
      " ┃    ┃    ┏┻━┓                             ┃             ┃       ┃   ┃   \n",
      " ┃    ┃    ┃ 77                             ┃             ┃       ┃   ┃   \n",
      " ┃    ┃    ┃  ┃                             ┃             ┃       ┃   ┃   \n",
      "75    ┃    ┃ 76                             ┃             ┃       ┃   ┃   \n",
      " ┃    ┃    ┃                                ┃             ┃       ┃   ┃   \n",
      " ┃    ┃   74                                ┃             ┃       ┃   ┃   \n",
      " ┃    ┃    ┃                                ┃             ┃       ┃   ┃   \n",
      "72    ┃   73                                ┃             ┃       ┃   ┃   \n",
      " ┃    ┃                                     ┃             ┃       ┃   ┃   \n",
      " ┃    ┃                                    71             ┃       ┃   ┃   \n",
      " ┃    ┃                             ┏━━━━━━━┻━━━━━━━┓     ┃       ┃   ┃   \n",
      " ┃    ┃                             ┃               ┃    70       ┃   ┃   \n",
      " ┃    ┃                             ┃               ┃   ┏━┻━━┓    ┃   ┃   \n",
      " ┃    ┃                             ┃               ┃   ┃    ┃    ┃  69   \n",
      " ┃    ┃                             ┃               ┃   ┃    ┃    ┃   ┃   \n",
      " ┃   67                             ┃               ┃   ┃    ┃    ┃  68   \n",
      " ┃    ┃                             ┃               ┃   ┃    ┃    ┃       \n",
      " ┃    ┃                             ┃              66   ┃    ┃    ┃       \n",
      " ┃    ┃                             ┃               ┃   ┃    ┃    ┃       \n",
      " ┃    ┃                            64              65   ┃    ┃    ┃       \n",
      " ┃    ┃                             ┃                   ┃    ┃    ┃       \n",
      " ┃    ┃                             ┃                   ┃   63    ┃       \n",
      " ┃    ┃                             ┃                   ┃    ┃    ┃       \n",
      " ┃   61                             ┃                   ┃   62    ┃       \n",
      " ┃    ┃                             ┃                   ┃         ┃       \n",
      " ┃    ┃                             ┃                  60         ┃       \n",
      " ┃    ┃                             ┃                  ┏┻━┓       ┃       \n",
      " ┃    ┃                             ┃                  ┃ 59       ┃       \n",
      " ┃    ┃                             ┃                  ┃  ┃       ┃       \n",
      " ┃   57                             ┃                  ┃ 58       ┃       \n",
      " ┃    ┃                             ┃                  ┃          ┃       \n",
      " ┃    ┃                             ┃                 56          ┃       \n",
      " ┃    ┃                             ┃                  ┃          ┃       \n",
      " ┃   54                             ┃                 55          ┃       \n",
      " ┃    ┃                             ┃                             ┃       \n",
      " ┃    ┃                             ┃                            53       \n",
      " ┃    ┃                             ┃                           ┏━┻┓      \n",
      " ┃    ┃                            52                           ┃  ┃      \n",
      " ┃    ┃                   ┏━━━━━━━━━┻━━━━━━━━━┓                 ┃  ┃      \n",
      " ┃    ┃                   ┃                   ┃                 ┃ 51      \n",
      " ┃    ┃                   ┃                   ┃                 ┃  ┃      \n",
      " ┃    ┃                   ┃                   ┃                49 50      \n",
      " ┃    ┃                   ┃                   ┃                 ┃         \n",
      " ┃    ┃                   ┃                   ┃                48         \n",
      " ┃    ┃                   ┃                   ┃                 ┃         \n",
      " ┃    ┃                   ┃                  46                47         \n",
      " ┃    ┃                   ┃                   ┃                           \n",
      " ┃    ┃                   ┃                  45                           \n",
      " ┃    ┃                   ┃               ┏━━━┻━━┓                        \n",
      " ┃    ┃                  44               ┃      ┃                        \n",
      " ┃    ┃                 ┏━┻━┓             ┃      ┃                        \n",
      " ┃    ┃                 ┃  43             ┃      ┃                        \n",
      " ┃    ┃                 ┃   ┃             ┃      ┃                        \n",
      " ┃   41                 ┃  42             ┃      ┃                        \n",
      " ┃    ┃                 ┃                 ┃      ┃                        \n",
      " ┃    ┃                 ┃                40      ┃                        \n",
      " ┃    ┃                 ┃             ┏━━━┻━━━┓  ┃                        \n",
      " ┃    ┃                 ┃             ┃       ┃ 39                        \n",
      " ┃    ┃                 ┃             ┃       ┃  ┃                        \n",
      " ┃   37                 ┃             ┃       ┃ 38                        \n",
      " ┃    ┃                 ┃             ┃       ┃                           \n",
      " ┃    ┃                36             ┃       ┃                           \n",
      " ┃    ┃               ┏━┻┓            ┃       ┃                           \n",
      " ┃    ┃               ┃ 35            ┃       ┃                           \n",
      " ┃    ┃               ┃  ┃            ┃       ┃                           \n",
      "33    ┃               ┃ 34            ┃       ┃                           \n",
      " ┃    ┃               ┃               ┃       ┃                           \n",
      " ┃    ┃               ┃              32       ┃                           \n",
      " ┃    ┃               ┃           ┏━━━┻━━━┓   ┃                           \n",
      " ┃    ┃               ┃           ┃       ┃  31                           \n",
      " ┃    ┃               ┃           ┃       ┃   ┃                           \n",
      " ┃    ┃               ┃          29       ┃  30                           \n",
      " ┃    ┃               ┃           ┃       ┃                               \n",
      " ┃   28               ┃           ┃       ┃                               \n",
      " ┃  ┏━┻┓              ┃           ┃       ┃                               \n",
      " ┃  ┃  ┃              ┃           ┃      27                               \n",
      " ┃  ┃  ┃              ┃           ┃     ┏━┻┓                              \n",
      " ┃  ┃  ┃              ┃          26     ┃  ┃                              \n",
      " ┃  ┃  ┃              ┃         ┏━┻━━┓  ┃  ┃                              \n",
      " ┃  ┃  ┃              ┃         ┃    ┃  ┃ 25                              \n",
      " ┃  ┃  ┃              ┃         ┃    ┃  ┃  ┃                              \n",
      " ┃ 23  ┃              ┃         ┃    ┃  ┃ 24                              \n",
      " ┃  ┃  ┃              ┃         ┃    ┃  ┃                                 \n",
      " ┃  ┃  ┃              ┃         ┃    ┃ 22                                 \n",
      " ┃  ┃  ┃              ┃         ┃    ┃  ┃                                 \n",
      " ┃  ┃  ┃             20         ┃    ┃ 21                                 \n",
      " ┃  ┃  ┃              ┃         ┃    ┃                                    \n",
      " ┃  ┃  ┃              ┃        19    ┃                                    \n",
      " ┃  ┃  ┃              ┃        ┏┻━┓  ┃                                    \n",
      " ┃  ┃  ┃              ┃        ┃  ┃ 18                                    \n",
      " ┃  ┃  ┃              ┃        ┃  ┃  ┃                                    \n",
      " ┃ 16  ┃              ┃        ┃  ┃ 17                                    \n",
      " ┃  ┃  ┃              ┃        ┃  ┃                                       \n",
      " ┃  ┃  ┃              ┃        ┃ 15                                       \n",
      " ┃  ┃  ┃              ┃        ┃  ┃                                       \n",
      " ┃  ┃  ┃              ┃       13 14                                       \n",
      " ┃  ┃  ┃              ┃        ┃                                          \n",
      " ┃  ┃  ┃              ┃       12                                          \n",
      " ┃  ┃  ┃              ┃        ┃                                          \n",
      " ┃  ┃  ┃             10       11                                          \n",
      " ┃  ┃  ┃              ┃                                                   \n",
      " ┃  ┃  ┃              9                                                   \n",
      " ┃  ┃  ┃              ┃                                                   \n",
      " ┃  7  ┃              8                                                   \n",
      " ┃  ┃  ┃                                                                  \n",
      " ┃  ┃  6                                                                  \n",
      " ┃  ┃ ┏┻┓                                                                 \n",
      " 5  ┃ ┃ ┃                                                                 \n",
      "┏┻┓ ┃ ┃ ┃                                                                 \n",
      "0 1 2 3 4                                                                 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# We get the core tree from the data in panX later. Here we will simulate a random tree and assume it is the tree from panX.\n",
    "\n",
    "import msprime\n",
    "import tskit\n",
    "\n",
    "import torch\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "from random import randint\n",
    "\n",
    "import gene_model\n",
    "import gfs\n",
    "import color_scheme\n",
    "\n",
    "import hgt_simulation\n",
    "import hgt_mutations\n",
    "import hgt_sim_args\n",
    "from gfs import gfs_from_matrix\n",
    "\n",
    "# General parameters:\n",
    "\n",
    "theta = 0.5 # Gene Gain rate\n",
    "rho = 0.5 # Gene Loss rate\n",
    "gene_conv = 0 # Gene Conversion rate\n",
    "recomb = 0 # Recombination rate\n",
    "\n",
    "num_sites = 1 # Number of sites to simulate\n",
    "num_samples = 5 # Number of samples / individuals\n",
    "\n",
    "hgt_rate_max = 1 # Maximum hgt rate\n",
    "hgt_rate_min = 0 # Minimum hgt rate\n",
    "    \n",
    "\n",
    "# Simulate the core tree:\n",
    "\n",
    "core_tree = msprime.sim_ancestry(\n",
    "                samples=num_samples,\n",
    "                sequence_length=1,\n",
    "                ploidy=1,\n",
    "                recombination_rate=0,\n",
    "                gene_conversion_rate=0,\n",
    "                gene_conversion_tract_length=1,  # One gene\n",
    "            )\n",
    "\n",
    "# Visualization of core tree :\n",
    "\n",
    "print(\"Core Tree:\")\n",
    "print(core_tree.first().draw_text())\n",
    "    \n",
    "\n",
    "# Example: Add HGT events to the core tree:\n",
    "\n",
    "ce_from_nwk = core_tree.first().newick()\n",
    "ce_from_ts = core_tree\n",
    "\n",
    "mts = gene_model.gene_model(  # ACHTUNG: Sortierung der Blätter ändert sich!\n",
    "    theta=theta,\n",
    "    rho=rho,\n",
    "    gene_conversion_rate=gene_conv,\n",
    "    recombination_rate=recomb,\n",
    "    hgt_rate=0.5,\n",
    "    num_samples=num_samples,\n",
    "    num_sites=num_sites,\n",
    "    double_site_relocation=False, \n",
    "    ce_from_nwk = ce_from_nwk,\n",
    "    # ce_from_ts = ce_from_ts\n",
    ")\n",
    "\n",
    "print(mts.first().draw_text())\n",
    "\n",
    "# Distribution of hgt_rate:\n",
    "\n",
    "prior = torch.distributions.Uniform(low=hgt_rate_min, high=hgt_rate_max)  # The hgt_rate PER SITE (GENOME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1d9b78ac-90a4-422a-b5e1-a69cecc82d2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor(0.3020)\n"
     ]
    }
   ],
   "source": [
    "# Add HGT events to the core tree:\n",
    "\n",
    "length_distribution_simulation_size = 20\n",
    "\n",
    "hgt_rate = prior.sample()\n",
    "print(hgt_rate) \n",
    "\n",
    "mts_list = []\n",
    "for i in range (1, length_distribution_simulation_size):\n",
    "    ce_from_nwk = core_tree.first().newick()\n",
    "    mts = gene_model.gene_model(\n",
    "        theta=0, # Set to 0 since it does not matter here\n",
    "        rho=0,\n",
    "        gene_conversion_rate=0,\n",
    "        recombination_rate=0,\n",
    "        hgt_rate=hgt_rate.item(),\n",
    "        num_samples=num_samples,\n",
    "        num_sites=1,\n",
    "        double_site_relocation=False, \n",
    "        ce_from_nwk = ce_from_nwk,\n",
    "        # ce_from_ts = ce_from_ts\n",
    "    )\n",
    "    \n",
    "    mts_list.append(mts)\n",
    "\n",
    "total_branch_lengths = []\n",
    "for mts in mts_list:\n",
    "    total_length = sum(tree.total_branch_length for tree in mts.trees())\n",
    "    total_branch_lengths.append(total_length)\n",
    "\n",
    "average_tree_length = sum(total_branch_lengths) / len(total_branch_lengths)\n",
    "\n",
    "# Define the empirical distribution of the tree length:\n",
    "\n",
    "ecdf = ECDF(total_branch_lengths)\n",
    "\n",
    "def tree_length_distribution(x):\n",
    "    return ecdf.x[np.searchsorted(ecdf.y, x)]\n",
    "\n",
    "# Calculate the distribution of the number of mutations:\n",
    "\n",
    "expected_number_of_mutations = theta * average_tree_length\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "860030f3-a3f4-49bf-aff4-3d494597e23a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 10, 15, 5, 13, 0]\n",
      "[4, 9, 3, 6, 17, 8]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "number_of_simulations = 1 # Number of simulations of the gfs\n",
    "size_factor = 3 # Factor which describes how much more trees are produced compared to the needed amount.\n",
    "\n",
    "for i in range(0, number_of_simulations):\n",
    "    \n",
    "    # Würfle die gewünschte Zahl der Gains\n",
    "    \n",
    "    number_of_gains = np.random.poisson(expected_number_of_mutations)\n",
    "    \n",
    "    # Erstelle mit num_sites einen Baum der entsprechenden Größe mal size_factor.\n",
    "\n",
    "    num_size = number_of_gains * size_factor\n",
    "    \n",
    "    mts = gene_model.gene_model(\n",
    "        theta=0, # Set to 0 since it does not matter yet\n",
    "        rho=0,\n",
    "        gene_conversion_rate=0,\n",
    "        recombination_rate=0,\n",
    "        hgt_rate = hgt_rate.item(),\n",
    "        num_samples=num_samples,\n",
    "        num_sites=num_size,\n",
    "        double_site_relocation=False, \n",
    "        ce_from_nwk = ce_from_nwk,\n",
    "    )\n",
    "\n",
    "    # Calculate the tree lengths\n",
    "    tree_lengths = np.array([tree.total_branch_length for tree in mts.trees()])\n",
    "    tree_lengths_average = sum(tree_lengths)/len(tree_lengths)\n",
    "    \n",
    "    # Make sure all lengths are positive\n",
    "    if np.any(tree_lengths <= 0):\n",
    "        raise ValueError(\"Tree with length 0 found.\")\n",
    "    \n",
    "    selected_trees_indexes = []\n",
    "    deleted_trees_indexes = []\n",
    "    \n",
    "    for j in range(number_of_gains):\n",
    "        \n",
    "        probabilities = tree_lengths / tree_lengths.sum()\n",
    "        \n",
    "        chosen_index = np.random.choice(len(mts.trees()), p=probabilities)\n",
    "        while (chosen_index in selected_trees_indexes or chosen_index in deleted_trees_indexes):\n",
    "            chosen_index = np.random.choice(len(mts.trees()), p=probabilities)\n",
    "        \n",
    "        selected_trees_indexes.append(chosen_index)\n",
    "        \n",
    "        if (len(selected_trees_indexes)) >= len(mts.trees()):\n",
    "            raise ValueError(\"Not enough trees to draw.\")\n",
    "        \n",
    "        # Calculate the average of the remaining trees.\n",
    "        mask = np.ones(len(tree_lengths), dtype=bool)\n",
    "        mask[selected_trees_indexes] = False\n",
    "        mask[deleted_trees_indexes] = False\n",
    "        current_average = sum(tree_lengths[mask])/len(tree_lengths[mask])\n",
    "        \n",
    "        # Remove an additional tree such that the mean of the tree length remains roughly the same.\n",
    "        target_diff = (current_average - tree_lengths_average)*(len(tree_lengths)-len(selected_trees_indexes)-len(deleted_trees_indexes)-1)\n",
    "        \n",
    "        differences = tree_lengths - tree_lengths_average - target_diff\n",
    "        differences[selected_trees_indexes] = np.inf\n",
    "        differences[deleted_trees_indexes] = np.inf\n",
    "        \n",
    "        closest_index = np.argmin(np.abs(differences))\n",
    "        deleted_trees_indexes.append(closest_index)\n",
    "\n",
    "    print(selected_trees_indexes)\n",
    "    print(deleted_trees_indexes)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f9f84be2-c212-4d92-83e5-47240cdc4664",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[15, 12, 0, 7, 4, 16]\n",
      "[6, 5, 2, 11, 1, 3]\n"
     ]
    }
   ],
   "source": [
    "from random import randint\n",
    "\n",
    "number_of_simulations = 1 # Number of simulations of the gfs\n",
    "size_factor = 3 # Factor which describes how much more trees are produced compared to the needed amount.\n",
    "\n",
    "for i in range(0, number_of_simulations):\n",
    "    \n",
    "    # Würfle die gewünschte Zahl der Gains\n",
    "    \n",
    "    number_of_gains = np.random.poisson(expected_number_of_mutations)\n",
    "    \n",
    "    # Erstelle mit num_sites einen Baum der entsprechenden Größe mal size_factor.\n",
    "\n",
    "    num_size = number_of_gains * size_factor\n",
    "\n",
    "    args = hgt_sim_args.Args(\n",
    "                sample_size=num_samples,\n",
    "                num_sites=num_size,\n",
    "                gene_conversion_rate=0,\n",
    "                recombination_rate=0,\n",
    "                hgt_rate=hgt_rate.item(),\n",
    "                ce_from_ts=None,\n",
    "                ce_from_nwk=ce_from_nwk,\n",
    "                random_seed=randint(1, int(2**32 - 2)),\n",
    "            )\n",
    "\n",
    "    ts, hgt_edges = hgt_simulation.run_simulate(args)\n",
    "\n",
    "    # Calculate the tree lengths\n",
    "    tree_lengths = np.array([tree.total_branch_length for tree in ts.trees()])\n",
    "    tree_lengths_average = sum(tree_lengths)/len(tree_lengths)\n",
    "    \n",
    "    # Make sure all lengths are positive\n",
    "    if np.any(tree_lengths <= 0):\n",
    "        raise ValueError(\"Tree with length 0 found.\")\n",
    "    \n",
    "    selected_trees_indexes = []\n",
    "    deleted_trees_indexes = []\n",
    "    \n",
    "    for j in range(number_of_gains):\n",
    "        \n",
    "        probabilities = tree_lengths / tree_lengths.sum()\n",
    "        \n",
    "        chosen_index = np.random.choice(len(ts.trees()), p=probabilities)\n",
    "        while (chosen_index in selected_trees_indexes or chosen_index in deleted_trees_indexes):\n",
    "            chosen_index = np.random.choice(len(ts.trees()), p=probabilities)\n",
    "        \n",
    "        selected_trees_indexes.append(chosen_index)\n",
    "        \n",
    "        if (len(selected_trees_indexes)) >= len(ts.trees()):\n",
    "            raise ValueError(\"Not enough trees to draw.\")\n",
    "        \n",
    "        # Calculate the average of the remaining trees.\n",
    "        mask = np.ones(len(tree_lengths), dtype=bool)\n",
    "        mask[selected_trees_indexes] = False\n",
    "        mask[deleted_trees_indexes] = False\n",
    "        current_average = sum(tree_lengths[mask])/len(tree_lengths[mask])\n",
    "        \n",
    "        # Remove an additional tree such that the mean of the tree length remains roughly the same.\n",
    "        target_diff = (current_average - tree_lengths_average)*(len(tree_lengths)-len(selected_trees_indexes)-len(deleted_trees_indexes)-1)\n",
    "        \n",
    "        differences = tree_lengths - tree_lengths_average - target_diff\n",
    "        differences[selected_trees_indexes] = np.inf\n",
    "        differences[deleted_trees_indexes] = np.inf\n",
    "        \n",
    "        closest_index = np.argmin(np.abs(differences))\n",
    "        deleted_trees_indexes.append(closest_index)\n",
    "\n",
    "    print(selected_trees_indexes)\n",
    "    print(deleted_trees_indexes)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "071e70df-5bf0-46f5-90b0-d4f7e75bd1c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mutation platziert auf Edge 10, Zeit: 0.29673183131808273\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"400\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l180,0 l0,138.2 l0,25 l0,5 l-180,0 l0,-5 l0,-25 l0,-138.2z\" /><path d=\"M200,0 l180,0 l0,138.2 l0,25 l0,5 l-180,0 l0,-5 l0,-25 l0,-138.2z\" /></g><g class=\"axes\"><g class=\"x-axis\"><g class=\"title\" transform=\"translate(200 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line class=\"ax-line\" x1=\"20\" x2=\"380\" y1=\"163.2\" y2=\"163.2\" /><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(200 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g><g class=\"tick\" transform=\"translate(380 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">2</text></g></g></g><g class=\"site s0\" transform=\"translate(20 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /></g><g class=\"site s1\" transform=\"translate(200 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n22 p0 root\" transform=\"translate(75.125 26.8)\"><g class=\"a22 c1 node n17 p0\" transform=\"translate(-35 71.8971)\"><g class=\"a17 c2 node n16 p0\" transform=\"translate(0.0 4.59179)\"><g class=\"a16 leaf node n0 p0 sample\" transform=\"translate(-13.125 18.1111)\"><path class=\"edge\" d=\"M 0 0 V -18.1111 H 13.125\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a16 c2 node n15 p0\" transform=\"translate(13.125 2.18125)\"><g class=\"a15 leaf node n1 p0 sample\" transform=\"translate(-12.25 15.9299)\"><path class=\"edge\" d=\"M 0 0 V -15.9299 H 12.25\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a15 c2 node n14 p0\" transform=\"translate(12.25 4.35484)\"><g class=\"a14 leaf node n2 p0 sample\" transform=\"translate(-10.5 11.5751)\"><path class=\"edge\" d=\"M 0 0 V -11.5751 H 10.5\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a14 c2 node n11 p0\" transform=\"translate(10.5 10.4317)\"><g class=\"a11 leaf node n3 p0 sample\" transform=\"translate(-7 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H 7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a11 leaf node n4 p0 sample\" transform=\"translate(7 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H -7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -10.4317 H -10.5\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -4.35484 H -12.25\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -2.18125 H -13.125\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -4.59179 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">16</text></g><path class=\"edge\" d=\"M 0 0 V -71.8971 H 35\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">17</text></g><g class=\"a22 c2 node n20 p0\" transform=\"translate(35 69.603)\"><g class=\"a20 leaf node n5 p0 sample\" transform=\"translate(-13.125 24.997)\"><path class=\"edge\" d=\"M 0 0 V -24.997 H 13.125\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a20 c2 node n13 p0\" transform=\"translate(13.125 20.0112)\"><g class=\"a13 leaf node n6 p0 sample\" transform=\"translate(-12.25 4.98585)\"><path class=\"edge\" d=\"M 0 0 V -4.98585 H 12.25\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(12.25 3.65999)\"><g class=\"a12 leaf node n7 p0 sample\" transform=\"translate(-10.5 1.32587)\"><path class=\"edge\" d=\"M 0 0 V -1.32587 H 10.5\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a12 c2 node n10 p0\" transform=\"translate(10.5 1.10642)\"><g class=\"a10 leaf node n8 p0 sample\" transform=\"translate(-7 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H 7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a10 leaf node n9 p0 sample\" transform=\"translate(7 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H -7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -1.10642 H -10.5\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -3.65999 H -12.25\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -20.0112 H -13.125\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -69.603 H -35\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">20</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">22</text></g></g></g><g class=\"tree t1\" transform=\"translate(200 0)\"><g class=\"plotbox\"><g class=\"c2 node n22 p0 root\" transform=\"translate(83.0398 26.8)\"><g class=\"a22 c1 node n17 p0\" transform=\"translate(-44.7443 71.8971)\"><g class=\"a17 c2 node n16 p0\" transform=\"translate(0.0 4.59179)\"><g class=\"a16 leaf node n0 p0 sample\" transform=\"translate(-11.9318 18.1111)\"><path class=\"edge\" d=\"M 0 0 V -18.1111 H 11.9318\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a16 c2 node n15 p0\" transform=\"translate(11.9318 2.18125)\"><g class=\"a15 leaf node n1 p0 sample\" transform=\"translate(-11.1364 15.9299)\"><path class=\"edge\" d=\"M 0 0 V -15.9299 H 11.1364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a15 c2 node n14 p0\" transform=\"translate(11.1364 4.35484)\"><g class=\"a14 leaf node n2 p0 sample\" transform=\"translate(-9.54545 11.5751)\"><path class=\"edge\" d=\"M 0 0 V -11.5751 H 9.54545\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a14 c2 node n11 p0\" transform=\"translate(9.54545 10.4317)\"><g class=\"a11 leaf node n3 p0 sample\" transform=\"translate(-6.36364 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H 6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a11 leaf node n4 p0 sample\" transform=\"translate(6.36364 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H -6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -10.4317 H -9.54545\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -4.35484 H -11.1364\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -2.18125 H -11.9318\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -4.59179 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">16</text></g><path class=\"edge\" d=\"M 0 0 V -71.8971 H 44.7443\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">17</text></g><g class=\"a22 c2 node n21 p0\" transform=\"translate(44.7443 63.9474)\"><g class=\"a21 c1 node n19 p0\" transform=\"translate(25.8523 7.94965)\"><g class=\"a19 leaf node n18 p0\" transform=\"translate(0.0 4.82226e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.82226e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -7.94965 H -25.8523\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">19</text></g><g class=\"a21 c2 node n20 p0\" transform=\"translate(-25.8523 5.65557)\"><g class=\"a20 leaf node n5 p0 sample\" transform=\"translate(-11.9318 24.997)\"><path class=\"edge\" d=\"M 0 0 V -24.997 H 11.9318\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a20 c2 node n13 p0\" transform=\"translate(11.9318 20.0112)\"><g class=\"a13 leaf node n6 p0 sample\" transform=\"translate(-11.1364 4.98585)\"><path class=\"edge\" d=\"M 0 0 V -4.98585 H 11.1364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(11.1364 3.65999)\"><g class=\"a12 leaf node n7 p0 sample\" transform=\"translate(-9.54545 1.32587)\"><path class=\"edge\" d=\"M 0 0 V -1.32587 H 9.54545\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a12 c2 node n10 p0\" transform=\"translate(9.54545 1.10642)\"><g class=\"a10 leaf node n8 p0 sample\" transform=\"translate(-6.36364 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H 6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a10 leaf node n9 p0 sample\" transform=\"translate(6.36364 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H -6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -1.10642 H -9.54545\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -3.65999 H -11.1364\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -20.0112 H -11.9318\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -5.65557 H 25.8523\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">20</text></g><path class=\"edge\" d=\"M 0 0 V -63.9474 H -44.7443\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">21</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">22</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "'<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"400\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l180,0 l0,138.2 l0,25 l0,5 l-180,0 l0,-5 l0,-25 l0,-138.2z\" /><path d=\"M200,0 l180,0 l0,138.2 l0,25 l0,5 l-180,0 l0,-5 l0,-25 l0,-138.2z\" /></g><g class=\"axes\"><g class=\"x-axis\"><g class=\"title\" transform=\"translate(200 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line class=\"ax-line\" x1=\"20\" x2=\"380\" y1=\"163.2\" y2=\"163.2\" /><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(200 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g><g class=\"tick\" transform=\"translate(380 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">2</text></g></g></g><g class=\"site s0\" transform=\"translate(20 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /></g><g class=\"site s1\" transform=\"translate(200 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n22 p0 root\" transform=\"translate(75.125 26.8)\"><g class=\"a22 c1 node n17 p0\" transform=\"translate(-35 71.8971)\"><g class=\"a17 c2 node n16 p0\" transform=\"translate(0.0 4.59179)\"><g class=\"a16 leaf node n0 p0 sample\" transform=\"translate(-13.125 18.1111)\"><path class=\"edge\" d=\"M 0 0 V -18.1111 H 13.125\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a16 c2 node n15 p0\" transform=\"translate(13.125 2.18125)\"><g class=\"a15 leaf node n1 p0 sample\" transform=\"translate(-12.25 15.9299)\"><path class=\"edge\" d=\"M 0 0 V -15.9299 H 12.25\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a15 c2 node n14 p0\" transform=\"translate(12.25 4.35484)\"><g class=\"a14 leaf node n2 p0 sample\" transform=\"translate(-10.5 11.5751)\"><path class=\"edge\" d=\"M 0 0 V -11.5751 H 10.5\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a14 c2 node n11 p0\" transform=\"translate(10.5 10.4317)\"><g class=\"a11 leaf node n3 p0 sample\" transform=\"translate(-7 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H 7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a11 leaf node n4 p0 sample\" transform=\"translate(7 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H -7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -10.4317 H -10.5\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -4.35484 H -12.25\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -2.18125 H -13.125\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -4.59179 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">16</text></g><path class=\"edge\" d=\"M 0 0 V -71.8971 H 35\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">17</text></g><g class=\"a22 c2 node n20 p0\" transform=\"translate(35 69.603)\"><g class=\"a20 leaf node n5 p0 sample\" transform=\"translate(-13.125 24.997)\"><path class=\"edge\" d=\"M 0 0 V -24.997 H 13.125\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a20 c2 node n13 p0\" transform=\"translate(13.125 20.0112)\"><g class=\"a13 leaf node n6 p0 sample\" transform=\"translate(-12.25 4.98585)\"><path class=\"edge\" d=\"M 0 0 V -4.98585 H 12.25\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(12.25 3.65999)\"><g class=\"a12 leaf node n7 p0 sample\" transform=\"translate(-10.5 1.32587)\"><path class=\"edge\" d=\"M 0 0 V -1.32587 H 10.5\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a12 c2 node n10 p0\" transform=\"translate(10.5 1.10642)\"><g class=\"a10 leaf node n8 p0 sample\" transform=\"translate(-7 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H 7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a10 leaf node n9 p0 sample\" transform=\"translate(7 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H -7\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -1.10642 H -10.5\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -3.65999 H -12.25\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -20.0112 H -13.125\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -69.603 H -35\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">20</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">22</text></g></g></g><g class=\"tree t1\" transform=\"translate(200 0)\"><g class=\"plotbox\"><g class=\"c2 node n22 p0 root\" transform=\"translate(83.0398 26.8)\"><g class=\"a22 c1 node n17 p0\" transform=\"translate(-44.7443 71.8971)\"><g class=\"a17 c2 node n16 p0\" transform=\"translate(0.0 4.59179)\"><g class=\"a16 leaf node n0 p0 sample\" transform=\"translate(-11.9318 18.1111)\"><path class=\"edge\" d=\"M 0 0 V -18.1111 H 11.9318\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a16 c2 node n15 p0\" transform=\"translate(11.9318 2.18125)\"><g class=\"a15 leaf node n1 p0 sample\" transform=\"translate(-11.1364 15.9299)\"><path class=\"edge\" d=\"M 0 0 V -15.9299 H 11.1364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a15 c2 node n14 p0\" transform=\"translate(11.1364 4.35484)\"><g class=\"a14 leaf node n2 p0 sample\" transform=\"translate(-9.54545 11.5751)\"><path class=\"edge\" d=\"M 0 0 V -11.5751 H 9.54545\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a14 c2 node n11 p0\" transform=\"translate(9.54545 10.4317)\"><g class=\"a11 leaf node n3 p0 sample\" transform=\"translate(-6.36364 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H 6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a11 leaf node n4 p0 sample\" transform=\"translate(6.36364 1.14339)\"><path class=\"edge\" d=\"M 0 0 V -1.14339 H -6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -10.4317 H -9.54545\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -4.35484 H -11.1364\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -2.18125 H -11.9318\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -4.59179 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">16</text></g><path class=\"edge\" d=\"M 0 0 V -71.8971 H 44.7443\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">17</text></g><g class=\"a22 c2 node n21 p0\" transform=\"translate(44.7443 63.9474)\"><g class=\"a21 c1 node n19 p0\" transform=\"translate(25.8523 7.94965)\"><g class=\"a19 leaf node n18 p0\" transform=\"translate(0.0 4.82226e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.82226e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -7.94965 H -25.8523\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">19</text></g><g class=\"a21 c2 node n20 p0\" transform=\"translate(-25.8523 5.65557)\"><g class=\"a20 leaf node n5 p0 sample\" transform=\"translate(-11.9318 24.997)\"><path class=\"edge\" d=\"M 0 0 V -24.997 H 11.9318\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a20 c2 node n13 p0\" transform=\"translate(11.9318 20.0112)\"><g class=\"a13 leaf node n6 p0 sample\" transform=\"translate(-11.1364 4.98585)\"><path class=\"edge\" d=\"M 0 0 V -4.98585 H 11.1364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(11.1364 3.65999)\"><g class=\"a12 leaf node n7 p0 sample\" transform=\"translate(-9.54545 1.32587)\"><path class=\"edge\" d=\"M 0 0 V -1.32587 H 9.54545\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a12 c2 node n10 p0\" transform=\"translate(9.54545 1.10642)\"><g class=\"a10 leaf node n8 p0 sample\" transform=\"translate(-6.36364 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H 6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a10 leaf node n9 p0 sample\" transform=\"translate(6.36364 0.219452)\"><path class=\"edge\" d=\"M 0 0 V -0.219452 H -6.36364\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -1.10642 H -9.54545\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -3.65999 H -11.1364\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -20.0112 H -11.9318\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -5.65557 H 25.8523\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">20</text></g><path class=\"edge\" d=\"M 0 0 V -63.9474 H -44.7443\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">21</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">22</text></g></g></g></g></g></svg>'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alleles = [\"absent\", \"present\"]\n",
    "\n",
    "# Place one mutation:\n",
    "\n",
    "num_sites = 2\n",
    "\n",
    "args = hgt_sim_args.Args(\n",
    "            sample_size=num_samples,\n",
    "            num_sites=num_sites,\n",
    "            gene_conversion_rate=0,\n",
    "            recombination_rate=0,\n",
    "            hgt_rate=0.5,\n",
    "            ce_from_ts=None,\n",
    "            ce_from_nwk=ce_from_nwk,\n",
    "            random_seed=randint(1, int(2**32 - 2)),\n",
    "        )\n",
    "\n",
    "ts, hgt_edges = hgt_simulation.run_simulate(args)\n",
    "\n",
    "edges = ts.tables.edges\n",
    "nodes = ts.tables.nodes\n",
    "node_times = nodes.time\n",
    "\n",
    "edge_lengths = np.array([\n",
    "    node_times[edges.parent[i]] - node_times[edges.child[i]]\n",
    "    for i in range(len(edges.parent))\n",
    "])\n",
    "\n",
    "# Kumulative Wahrscheinlichkeitsverteilung erstellen\n",
    "cdf = np.cumsum(edge_lengths / np.sum(edge_lengths))\n",
    "\n",
    "# Zufälligen Ast basierend auf Längen auswählen\n",
    "random_value = np.random.uniform()\n",
    "selected_edge_index = np.searchsorted(cdf, random_value)\n",
    "\n",
    "# Details des ausgewählten Edges\n",
    "selected_edge = edges[selected_edge_index]\n",
    "parent_time = node_times[selected_edge.parent]\n",
    "child_time = node_times[selected_edge.child]\n",
    "\n",
    "# Zufällige Position innerhalb des Astes\n",
    "mutation_time = np.random.uniform(child_time, parent_time)\n",
    "\n",
    "print(f\"Mutation platziert auf Edge {selected_edge_index}, Zeit: {mutation_time}\")\n",
    "\n",
    "# Initialize sites:\n",
    "\n",
    "tables = ts.dump_tables()\n",
    "\n",
    "position = np.arange(0, num_sites, dtype=\"uint32\")\n",
    "\n",
    "ancestral_state = [alleles[0]] * num_sites\n",
    "ancestral_state, ancestral_state_offset = tskit.pack_strings(ancestral_state)\n",
    "\n",
    "tables.sites.set_columns(\n",
    "    position=position,\n",
    "    ancestral_state=ancestral_state,\n",
    "    ancestral_state_offset=ancestral_state_offset,\n",
    ")\n",
    "\n",
    "ts = tables.tree_sequence()\n",
    "\n",
    "\"\"\"\n",
    "# Mutation hinzufügen: an einer spezifischen Site\n",
    "mutated_site = 1\n",
    "tables = ts.dump_tables()\n",
    "tables.mutations.add_row(site=mutated_site, \n",
    "                         node=selected_edge.child, \n",
    "                         #time=mutation_time, \n",
    "                         derived_state = \"present\", \n",
    "                         metadata=0b00.to_bytes(1),\n",
    "                        # new = True\n",
    "                        )\n",
    "\n",
    "ts = tables.tree_sequence()\n",
    "\"\"\"\n",
    "\n",
    "ts2 = ts\n",
    "\n",
    "ts.draw_svg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e2b750e-496f-4877-a2bd-b7ae783bf0a3",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ts2' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m ts \u001b[38;5;241m=\u001b[39m \u001b[43mts2\u001b[49m\n\u001b[0;32m      3\u001b[0m tables \u001b[38;5;241m=\u001b[39m ts\u001b[38;5;241m.\u001b[39mdump_tables()\n\u001b[0;32m      5\u001b[0m theta_total_events \u001b[38;5;241m=\u001b[39m theta\n",
      "\u001b[1;31mNameError\u001b[0m: name 'ts2' is not defined"
     ]
    }
   ],
   "source": [
    "ts = ts2\n",
    "\n",
    "tables = ts.dump_tables()\n",
    "\n",
    "theta_total_events = theta\n",
    "rho_total_events = rho * num_sites\n",
    "\n",
    "root_proba = theta_total_events / (rho_total_events if rho_total_events != 0 else theta)\n",
    "if not (0 <= root_proba <= 1):\n",
    "    raise ValueError(f\"Invalid theta / rho resulting in a root probability of {root_proba}\")\n",
    "\n",
    "event_rate = rho_total_events + theta_total_events\n",
    "theta_proba = theta_total_events / event_rate\n",
    "rho_proba = rho_total_events / event_rate\n",
    "event_rate /= num_sites\n",
    "\n",
    "gain_loss_model = msprime.MatrixMutationModel(\n",
    "    alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [1 - theta_proba, theta_proba],\n",
    "        [rho_proba, 1 - rho_proba],\n",
    "    ],\n",
    ")\n",
    "#gain_loss_model = PythonMutationMatrixModel(\n",
    "#        alleles=model.alleles,\n",
    "#        root_distribution=model.root_distribution,\n",
    "#        transition_matrix=model.transition_matrix,\n",
    "#    )\n",
    "\n",
    "rate_map = msprime.RateMap(position=[0, num_sites], rate=[event_rate])\n",
    "\n",
    "\n",
    "hgt_generator = hgt_mutations.HGTMutationGenerator(rate_map=rate_map, model=gain_loss_model)\n",
    "\n",
    "\n",
    "# Fix the hgt_events, such that mutations coming from hgt are counted in correctly.\n",
    "\n",
    "edges = list(tables.edges)\n",
    "edges.extend(hgt_edges)\n",
    "\n",
    "child_ids = {e.child for e in edges}\n",
    "root_nodes = list({e.parent for e in edges if e.parent not in child_ids})\n",
    "\n",
    "hgt_generator.apply_mutations(tables, hgt_edges, root_nodes)\n",
    "hgt_generator.populate_tables(tables)\n",
    "\n",
    "edges = list(e for e in edges if not e in hgt_edges)\n",
    "tables.edges.clear()\n",
    "\n",
    "for e in edges:\n",
    "    tables.edges.add_row(left=e.left, right=e.right, parent=e.parent, child=e.child)\n",
    "ts = tables.tree_sequence()\n",
    "\n",
    "\n",
    "for var in mts.variants():\n",
    "        gfs_matrix = ts.allele_frequency_spectrum(span_normalise = False, polarised = True)\n",
    "\n",
    "# print(gfs_matrix)\n",
    "\n",
    "ts.draw_svg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "050b1f10-2758-4762-8cbc-8624356118b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0. 0. 0. 0. 0. 0.]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\" /></g><g class=\"axes\"><g class=\"x-axis\"><g class=\"title\" transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line class=\"ax-line\" x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\" /><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g><g class=\"site s0\" transform=\"translate(20 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /><g class=\"mut m8\"><polyline class=\"sym\" points=\"2.5,-6.5 0,-1.5 -2.5,-6.5\" /></g><g class=\"mut m7\"><polyline class=\"sym\" points=\"2.5,-10.5 0,-5.5 -2.5,-10.5\" /></g><g class=\"mut m6\"><polyline class=\"sym\" points=\"2.5,-14.5 0,-9.5 -2.5,-14.5\" /></g><g class=\"mut m5\"><polyline class=\"sym\" points=\"2.5,-18.5 0,-13.5 -2.5,-18.5\" /></g><g class=\"mut m4\"><polyline class=\"sym\" points=\"2.5,-22.5 0,-17.5 -2.5,-22.5\" /></g><g class=\"mut m3\"><polyline class=\"sym\" points=\"2.5,-26.5 0,-21.5 -2.5,-26.5\" /></g><g class=\"mut m2\"><polyline class=\"sym\" points=\"2.5,-30.5 0,-25.5 -2.5,-30.5\" /></g><g class=\"mut m1\"><polyline class=\"sym\" points=\"2.5,-34.5 0,-29.5 -2.5,-34.5\" /></g><g class=\"mut m0\"><polyline class=\"sym\" points=\"2.5,-38.5 0,-33.5 -2.5,-38.5\" /></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n28 p0 root\" transform=\"translate(105.406 26.8)\"><g class=\"a28 c1 node n19 p0\" transform=\"translate(22.5938 28.4297)\"><g class=\"a19 c1 node n14 p0\" transform=\"translate(0.0 19.9004)\"><g class=\"a14 c2 node n13 p0\" transform=\"translate(0.0 8.66666)\"><g class=\"a13 c1 node n9 p0\" transform=\"translate(-6 11.97)\"><g class=\"a9 leaf node n8 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -11.97 H 6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">9</text></g><g class=\"a13 c1 node n12 p0\" transform=\"translate(6 5.81121)\"><g class=\"a12 leaf node n11 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -5.81121 H -6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -8.66666 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -19.9004 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -28.4297 H -22.5938\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">19</text></g><g class=\"a28 c2 m0 node n27 p0 s0\" transform=\"translate(-22.5938 7.67788)\"><g class=\"a27 c2 node n25 p0\" transform=\"translate(21.1875 12.1479)\"><g class=\"a25 c1 node n22 p0\" transform=\"translate(-6 0.968869)\"><g class=\"a22 c1 node n21 p0\" transform=\"translate(0.0 7.63507)\"><g class=\"a21 leaf node n20 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">20</text></g><path class=\"edge\" d=\"M 0 0 V -7.63507 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">21</text></g><path class=\"edge\" d=\"M 0 0 V -0.968869 H 6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">22</text></g><g class=\"a25 c1 node n24 p0\" transform=\"translate(6 0.968865)\"><g class=\"a24 leaf node n23 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">23</text></g><path class=\"edge\" d=\"M 0 0 V -0.968865 H -6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">24</text></g><path class=\"edge\" d=\"M 0 0 V -12.1479 H -21.1875\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">25</text></g><g class=\"a27 c2 node n26 p0\" transform=\"translate(-21.1875 9.56991)\"><g class=\"a26 c1 m1 node n16 p0 s0\" transform=\"translate(24.375 31.0823)\"><g class=\"a16 leaf node n15 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -31.0823 H -24.375\" /><g class=\"mut m1 s0\" transform=\"translate(0 -27.6305)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"27.6305\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab rgt\" transform=\"translate(5 0)\">1</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">16</text></g><g class=\"a26 c2 m2 node n18 p0 s0\" transform=\"translate(-24.375 19.6832)\"><g class=\"a18 leaf m4 node n0 p0 s0 sample\" transform=\"translate(-11.25 57.6691)\"><path class=\"edge\" d=\"M 0 0 V -57.6691 H 11.25\" /><g class=\"mut m4 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">4</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a18 c2 node n17 p0\" transform=\"translate(11.25 6.79265)\"><g class=\"a17 leaf m5 node n1 p0 s0 sample\" transform=\"translate(-10.5 50.8764)\"><path class=\"edge\" d=\"M 0 0 V -50.8764 H 10.5\" /><g class=\"mut m5 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">5</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a17 c1 m3 node n10 p0 s0\" transform=\"translate(10.5 19.0844)\"><g class=\"a10 c1 node n7 p0\" transform=\"translate(0.0 6.15877)\"><g class=\"a7 c2 node n6 p0\" transform=\"translate(0.0 17.9988)\"><g class=\"a6 leaf m6 node n2 p0 s0 sample\" transform=\"translate(-9 7.63447)\"><path class=\"edge\" d=\"M 0 0 V -7.63447 H 9\" /><g class=\"mut m6 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">6</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(9 1.50474)\"><g class=\"a5 leaf m7 node n3 p0 s0 sample\" transform=\"translate(-6 6.12973)\"><path class=\"edge\" d=\"M 0 0 V -6.12973 H 6\" /><g class=\"mut m7 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">7</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 leaf m8 node n4 p0 s0 sample\" transform=\"translate(6 6.12973)\"><path class=\"edge\" d=\"M 0 0 V -6.12973 H -6\" /><g class=\"mut m8 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab rgt\" transform=\"translate(5 0)\">8</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -1.50474 H -9\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -17.9988 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -6.15877 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -19.0844 H -10.5\" /><g class=\"mut m3 s0\" transform=\"translate(0 -13.1455)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"13.1455\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab rgt\" transform=\"translate(5 0)\">3</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -6.79265 H -11.25\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">17</text></g><path class=\"edge\" d=\"M 0 0 V -19.6832 H 24.375\" /><g class=\"mut m2 s0\" transform=\"translate(0 -14.6281)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"14.6281\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">2</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -9.56991 H 21.1875\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">26</text></g><path class=\"edge\" d=\"M 0 0 V -7.67788 H 22.5938\" /><g class=\"mut m0 s0\" transform=\"translate(0 -3.14934)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"3.14934\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">27</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">28</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "'<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\" /></g><g class=\"axes\"><g class=\"x-axis\"><g class=\"title\" transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line class=\"ax-line\" x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\" /><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g><g class=\"site s0\" transform=\"translate(20 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /><g class=\"mut m8\"><polyline class=\"sym\" points=\"2.5,-6.5 0,-1.5 -2.5,-6.5\" /></g><g class=\"mut m7\"><polyline class=\"sym\" points=\"2.5,-10.5 0,-5.5 -2.5,-10.5\" /></g><g class=\"mut m6\"><polyline class=\"sym\" points=\"2.5,-14.5 0,-9.5 -2.5,-14.5\" /></g><g class=\"mut m5\"><polyline class=\"sym\" points=\"2.5,-18.5 0,-13.5 -2.5,-18.5\" /></g><g class=\"mut m4\"><polyline class=\"sym\" points=\"2.5,-22.5 0,-17.5 -2.5,-22.5\" /></g><g class=\"mut m3\"><polyline class=\"sym\" points=\"2.5,-26.5 0,-21.5 -2.5,-26.5\" /></g><g class=\"mut m2\"><polyline class=\"sym\" points=\"2.5,-30.5 0,-25.5 -2.5,-30.5\" /></g><g class=\"mut m1\"><polyline class=\"sym\" points=\"2.5,-34.5 0,-29.5 -2.5,-34.5\" /></g><g class=\"mut m0\"><polyline class=\"sym\" points=\"2.5,-38.5 0,-33.5 -2.5,-38.5\" /></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n28 p0 root\" transform=\"translate(105.406 26.8)\"><g class=\"a28 c1 node n19 p0\" transform=\"translate(22.5938 28.4297)\"><g class=\"a19 c1 node n14 p0\" transform=\"translate(0.0 19.9004)\"><g class=\"a14 c2 node n13 p0\" transform=\"translate(0.0 8.66666)\"><g class=\"a13 c1 node n9 p0\" transform=\"translate(-6 11.97)\"><g class=\"a9 leaf node n8 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -11.97 H 6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">9</text></g><g class=\"a13 c1 node n12 p0\" transform=\"translate(6 5.81121)\"><g class=\"a12 leaf node n11 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -5.81121 H -6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -8.66666 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -19.9004 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -28.4297 H -22.5938\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">19</text></g><g class=\"a28 c2 m0 node n27 p0 s0\" transform=\"translate(-22.5938 7.67788)\"><g class=\"a27 c2 node n25 p0\" transform=\"translate(21.1875 12.1479)\"><g class=\"a25 c1 node n22 p0\" transform=\"translate(-6 0.968869)\"><g class=\"a22 c1 node n21 p0\" transform=\"translate(0.0 7.63507)\"><g class=\"a21 leaf node n20 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">20</text></g><path class=\"edge\" d=\"M 0 0 V -7.63507 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">21</text></g><path class=\"edge\" d=\"M 0 0 V -0.968869 H 6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">22</text></g><g class=\"a25 c1 node n24 p0\" transform=\"translate(6 0.968865)\"><g class=\"a24 leaf node n23 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">23</text></g><path class=\"edge\" d=\"M 0 0 V -0.968865 H -6\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">24</text></g><path class=\"edge\" d=\"M 0 0 V -12.1479 H -21.1875\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">25</text></g><g class=\"a27 c2 node n26 p0\" transform=\"translate(-21.1875 9.56991)\"><g class=\"a26 c1 m1 node n16 p0 s0\" transform=\"translate(24.375 31.0823)\"><g class=\"a16 leaf node n15 p0\" transform=\"translate(0.0 4.3135e-06)\"><path class=\"edge\" d=\"M 0 0 V -4.3135e-06 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 11)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -31.0823 H -24.375\" /><g class=\"mut m1 s0\" transform=\"translate(0 -27.6305)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"27.6305\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab rgt\" transform=\"translate(5 0)\">1</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">16</text></g><g class=\"a26 c2 m2 node n18 p0 s0\" transform=\"translate(-24.375 19.6832)\"><g class=\"a18 leaf m4 node n0 p0 s0 sample\" transform=\"translate(-11.25 57.6691)\"><path class=\"edge\" d=\"M 0 0 V -57.6691 H 11.25\" /><g class=\"mut m4 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">4</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a18 c2 node n17 p0\" transform=\"translate(11.25 6.79265)\"><g class=\"a17 leaf m5 node n1 p0 s0 sample\" transform=\"translate(-10.5 50.8764)\"><path class=\"edge\" d=\"M 0 0 V -50.8764 H 10.5\" /><g class=\"mut m5 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">5</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a17 c1 m3 node n10 p0 s0\" transform=\"translate(10.5 19.0844)\"><g class=\"a10 c1 node n7 p0\" transform=\"translate(0.0 6.15877)\"><g class=\"a7 c2 node n6 p0\" transform=\"translate(0.0 17.9988)\"><g class=\"a6 leaf m6 node n2 p0 s0 sample\" transform=\"translate(-9 7.63447)\"><path class=\"edge\" d=\"M 0 0 V -7.63447 H 9\" /><g class=\"mut m6 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">6</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(9 1.50474)\"><g class=\"a5 leaf m7 node n3 p0 s0 sample\" transform=\"translate(-6 6.12973)\"><path class=\"edge\" d=\"M 0 0 V -6.12973 H 6\" /><g class=\"mut m7 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">7</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 leaf m8 node n4 p0 s0 sample\" transform=\"translate(6 6.12973)\"><path class=\"edge\" d=\"M 0 0 V -6.12973 H -6\" /><g class=\"mut m8 s0\" transform=\"translate(0 -4.31356e-10)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"4.31356e-10\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab rgt\" transform=\"translate(5 0)\">8</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -1.50474 H -9\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -17.9988 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -6.15877 H 0.0\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -19.0844 H -10.5\" /><g class=\"mut m3 s0\" transform=\"translate(0 -13.1455)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"13.1455\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab rgt\" transform=\"translate(5 0)\">3</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -6.79265 H -11.25\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7)\">17</text></g><path class=\"edge\" d=\"M 0 0 V -19.6832 H 24.375\" /><g class=\"mut m2 s0\" transform=\"translate(0 -14.6281)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"14.6281\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">2</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -9.56991 H 21.1875\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">26</text></g><path class=\"edge\" d=\"M 0 0 V -7.67788 H 22.5938\" /><g class=\"mut m0 s0\" transform=\"translate(0 -3.14934)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"3.14934\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7)\">27</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">28</text></g></g></g></g></g></svg>'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alleles = [\"absent\", \"present\"]\n",
    "\n",
    "num_sites = 1\n",
    "\n",
    "theta_total_events = 0.5\n",
    "rho_total_events = 0 * num_sites\n",
    "\n",
    "root_proba = theta_total_events / (rho_total_events if rho_total_events != 0 else theta)\n",
    "#if not (0 <= root_proba <= 1):\n",
    "#    raise ValueError(f\"Invalid theta / rho resulting in a root probability of {root_proba}\")\n",
    "\n",
    "event_rate = rho_total_events + theta_total_events\n",
    "theta_proba = theta_total_events / event_rate\n",
    "rho_proba = rho_total_events / event_rate\n",
    "event_rate /= num_sites\n",
    "\n",
    "gain_loss_model = msprime.MatrixMutationModel(\n",
    "    alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [1 - theta_proba, theta_proba],\n",
    "        [rho_proba, 1 - rho_proba],\n",
    "    ],\n",
    ")\n",
    "\n",
    "\n",
    "# Place one mutation:\n",
    "\n",
    "\n",
    "args = hgt_sim_args.Args(\n",
    "            sample_size=num_samples,\n",
    "            num_sites=num_sites,\n",
    "            gene_conversion_rate=0,\n",
    "            recombination_rate=0,\n",
    "            hgt_rate=0.5,\n",
    "            ce_from_ts=None,\n",
    "            ce_from_nwk=ce_from_nwk,\n",
    "            random_seed=randint(1, int(2**32 - 2)),\n",
    "        )\n",
    "\n",
    "ts, hgt_edges = hgt_simulation.run_simulate(args)\n",
    "\n",
    "tables = ts.dump_tables()\n",
    "\n",
    "position = np.arange(0, num_sites, dtype=\"uint32\")\n",
    "\n",
    "ancestral_state = [alleles[0]] * num_sites\n",
    "ancestral_state, ancestral_state_offset = tskit.pack_strings(ancestral_state)\n",
    "\n",
    "tables.sites.set_columns(\n",
    "    position=position,\n",
    "    ancestral_state=ancestral_state,\n",
    "    ancestral_state_offset=ancestral_state_offset,\n",
    ")\n",
    "\n",
    "ts = tables.tree_sequence()\n",
    "\n",
    "\n",
    "mts = hgt_mutations.sim_mutations(\n",
    "            ts,\n",
    "            hgt_edges=hgt_edges,\n",
    "            event_rate=event_rate,\n",
    "            model=gain_loss_model,\n",
    "            one_mutation = False,\n",
    "        )\n",
    "\n",
    "for var in mts.variants():\n",
    "        gfs_matrix = mts.allele_frequency_spectrum(span_normalise = False, polarised = True)\n",
    "\n",
    "print(gfs_matrix)\n",
    "\n",
    "mts.draw_svg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7c461064-e45e-4e62-a670-feddfac41c65",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name '_unpack_tables' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[41], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m tables \u001b[38;5;241m=\u001b[39m mts\u001b[38;5;241m.\u001b[39mdump_tables()\n\u001b[1;32m----> 3\u001b[0m derived_state, parent_state, metadata_state \u001b[38;5;241m=\u001b[39m \u001b[43m_unpack_tables\u001b[49m(tables)\n\u001b[0;32m      5\u001b[0m derived_state\n",
      "\u001b[1;31mNameError\u001b[0m: name '_unpack_tables' is not defined"
     ]
    }
   ],
   "source": [
    "tables = mts.dump_tables()\n",
    "\n",
    "derived_state, parent_state, metadata_state = _unpack_tables(tables)\n",
    "\n",
    "derived_state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "9b61c88d-6f73-453f-9ff1-d5e49a794bac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<hgt_mutations.HGTMutationGenerator at 0x219b77f26f0>"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alleles = [\"absent\", \"present\"]\n",
    "\n",
    "theta_total_events = theta\n",
    "rho_total_events = rho * num_sites\n",
    "\n",
    "root_proba = theta_total_events / (rho_total_events if rho_total_events != 0 else theta)\n",
    "if not (0 <= root_proba <= 1):\n",
    "    raise ValueError(f\"Invalid theta / rho resulting in a root probability of {root_proba}\")\n",
    "\n",
    "event_rate = rho_total_events + theta_total_events\n",
    "theta_proba = theta_total_events / event_rate\n",
    "rho_proba = rho_total_events / event_rate\n",
    "event_rate /= num_sites\n",
    "\n",
    "gain_loss_model = msprime.MatrixMutationModel(\n",
    "    alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [1 - theta_proba, theta_proba],\n",
    "        [rho_proba, 1 - rho_proba],\n",
    "    ],\n",
    ")\n",
    "\n",
    "tables = ts.dump_tables()\n",
    "\n",
    "# Set the ancestral state for each site.\n",
    "# poisson = np.random.poisson(theta / (rho if rho != 0 else 1))\n",
    "poisson = 0\n",
    "poisson = min(poisson, num_sites)\n",
    "\n",
    "position = np.arange(0, num_sites, dtype=\"uint32\")\n",
    "position = np.random.choice(position, poisson, replace=False)\n",
    "position.sort()\n",
    "\n",
    "ancestral_state = [alleles[1]] * poisson\n",
    "ancestral_state, ancestral_state_offset = tskit.pack_strings(ancestral_state)\n",
    "\n",
    "tables.sites.set_columns(\n",
    "    position=position,\n",
    "    ancestral_state=ancestral_state,\n",
    "    ancestral_state_offset=ancestral_state_offset,\n",
    ")\n",
    "ts = tables.tree_sequence()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Anfang hgt_mutations.sim_mutations:\n",
    "\n",
    "tables = ts.dump_tables()\n",
    "\n",
    "gene_count = tables.sequence_length\n",
    "\n",
    "gain_loss_model = hgt_mutations.PythonMutationMatrixModel(\n",
    "    alleles=gain_loss_model.alleles,\n",
    "    root_distribution=gain_loss_model.root_distribution,\n",
    "    transition_matrix=gain_loss_model.transition_matrix,\n",
    ")\n",
    "\n",
    "rate_map = msprime.RateMap(position=[0, gene_count], rate=[event_rate])\n",
    "\n",
    "edges = list(tables.edges)\n",
    "edges.extend(hgt_edges)\n",
    "\n",
    "child_ids = {e.child for e in edges}\n",
    "root_nodes = list({e.parent for e in edges if e.parent not in child_ids})\n",
    "\n",
    "hgt_generator = hgt_mutations.HGTMutationGenerator(rate_map=rate_map, model=gain_loss_model)\n",
    "\n",
    "ts = hgt_generator.generate(\n",
    "    tables,\n",
    "    edges,\n",
    "    root_nodes,\n",
    "    randint(0, 4294967295),\n",
    "    keep=True,\n",
    "    discrete_genome=True,\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "hgt_generator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "id": "a2ada6b5-de75-4175-9c51-7539ba6cbc5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'\\x00'"
      ]
     },
     "execution_count": 469,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print(mts.tables.edges)\n",
    "# print(mts.first().draw_text())\n",
    "\n",
    "# Zugriff auf die Tabellen\n",
    "\n",
    "0b00.to_bytes(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "44923b03-354b-4c35-964d-61baef1bcec1",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Tree' object has no attribute 'dump_tables'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[352], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m tables \u001b[38;5;241m=\u001b[39m \u001b[43mmts\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mat\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdump_tables\u001b[49m()\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'Tree' object has no attribute 'dump_tables'"
     ]
    }
   ],
   "source": [
    "tables = mts.at(0).dump_tables()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "cda23d5a-4b95-46f9-8236-12b153b1bd66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "[]\n",
      "0.625\n",
      "[]\n",
      "[0]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <div>\n",
       "            <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "            </style>\n",
       "            <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"padding:0;line-height:21px;\">\n",
       "                                    <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                                    <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                                </th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            <tr><td>Trees</td><td>93</td></tr>\n",
       "                            <tr><td>Sequence Length</td><td>93.0</td></tr>\n",
       "                            <tr><td>Time Units</td><td>unknown</td></tr>\n",
       "                            <tr><td>Sample Nodes</td><td>10</td></tr>\n",
       "                            <tr><td>Total Size</td><td>538.5 KiB</td></tr>\n",
       "                            <tr>\n",
       "                                <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td>\n",
       "                            </tr>\n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"line-height:21px;\">Table</th>\n",
       "                                <th>Rows</th>\n",
       "                                <th>Size</th>\n",
       "                                <th>Has Metadata</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "            <tr>\n",
       "                <td>Edges</td>\n",
       "                <td>11,031</td>\n",
       "                <td>344.7 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Individuals</td>\n",
       "                <td>0</td>\n",
       "                <td>24 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Migrations</td>\n",
       "                <td>0</td>\n",
       "                <td>8 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Mutations</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Nodes</td>\n",
       "                <td>3,932</td>\n",
       "                <td>107.5 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Populations</td>\n",
       "                <td>1</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Provenances</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Sites</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th>Provenance Timestamp</th>\n",
       "                                <th>Software Name</th>\n",
       "                                <th>Version</th>\n",
       "                                <th>Command</th>\n",
       "                                <th>Full record</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "            </div>\n",
       "        </div>\n",
       "    "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x219b9ecef60>"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_backup = ts\n",
    "\n",
    "num_sites_test = 10\n",
    "# Set the ancestral state for each site.\n",
    "poisson = np.random.poisson(theta / (rho if rho != 0 else 1))\n",
    "poisson = min(poisson, num_sites_test)\n",
    "\n",
    "position = np.arange(0, num_sites_test, dtype=\"uint32\")\n",
    "position = np.random.choice(position, poisson, replace=False)\n",
    "position.sort()\n",
    "\n",
    "ancestral_state = [alleles[1]] * poisson\n",
    "ancestral_state, ancestral_state_offset = tskit.pack_strings(ancestral_state)\n",
    "\n",
    "print(poisson)\n",
    "print(position)\n",
    "print((theta / (rho if rho != 0 else 1)))\n",
    "print(ancestral_state)\n",
    "print(ancestral_state_offset)\n",
    "\n",
    "ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "9ac9d364-c10f-4b57-9c46-9e507620a2fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Duplikate und ihre Häufigkeit: {}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "96"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Schritt 1: Sortiere die Ziehungen\n",
    "selected_trees_sorted = sorted(selected_trees)\n",
    "\n",
    "# Schritt 2: Zähle Duplikate\n",
    "from collections import Counter\n",
    "tree_counts = Counter(selected_trees_sorted)\n",
    "\n",
    "# Ausgabe der Anzahl der Duplikate\n",
    "duplicates = {key: count for key, count in tree_counts.items() if count > 1}\n",
    "print(f\"Duplikate und ihre Häufigkeit: {duplicates}\")\n",
    "\n",
    "# Schritt 3: Entferne Duplikate (lasse nur einen Eintrag pro Baum übrig)\n",
    "unique_trees = list(tree_counts.keys())\n",
    "\n",
    "len(mts.trees())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "7c8321d7-5499-48e9-9f76-fb82b9c97953",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method TreeSequence.mutations of <tskit.trees.TreeSequence object at 0x0000019C39A624B0>>\n",
      "[0 0 0 0 0 0 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "# Hole die Tabellen aus dem TreeSequence-Objekt\n",
    "tables = mts_list[1].dump_tables()\n",
    "\n",
    "# 1. Füge eine zufällige Position für die Mutation hinzu\n",
    "site_position = np.random.uniform(0, mts_list[1].sequence_length)  # Zufällige Position im Bereich [0, sequence_length]\n",
    "site_id = tables.sites.add_row(position=site_position, ancestral_state=\"0\")\n",
    "\n",
    "# 2. Wähle einen zufälligen Knoten aus\n",
    "node_id = np.random.choice(range(mts_list[1].num_nodes))\n",
    "\n",
    "# 3. Füge die Mutation hinzu\n",
    "tables.mutations.add_row(site=site_id, node=node_id, derived_state=\"1\")\n",
    "\n",
    "# 4. Erstelle eine neue TreeSequence mit der Mutation\n",
    "mutated_tree = tables.tree_sequence()\n",
    "\n",
    "\n",
    "for var in mutated_tree.variants():\n",
    "    print(var.genotypes)\n",
    "\n",
    "with open(\"mutated_tree.svg\", \"w\") as svg_file:\n",
    "    svg_file.write(mutated_tree.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8ad1e3cd-c3b3-45e4-8471-497eb615d7b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n"
     ]
    }
   ],
   "source": [
    "theta = 2\n",
    "expected_number_of_mutations = theta * average_tree_length\n",
    "number_of_gains = np.random.poisson(expected_number_of_mutations)\n",
    "print(number_of_gains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4b595183-b943-49b1-ab03-7fc055d95b3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[13, 3, 13, 2, 1, 1, 2, 0, 3, 2, 3, 1, 13, 0, 12, 0, 10, 1, 11]\n",
      "0.7355953140336466\n",
      "0.2\n"
     ]
    }
   ],
   "source": [
    "number_of_mutations = []\n",
    "for mts in mts_list:\n",
    "    mutations = sum(tree.num_mutations for tree in mts.trees())\n",
    "    number_of_mutations.append(mutations)\n",
    "\n",
    "print(number_of_mutations)\n",
    "print(expected_number_of_mutations)\n",
    "\n",
    "print(average_tree_length)\n",
    "print(theta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a6bb4d7b-ee30-4a0d-8478-4e7b4be6d728",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor(0.6767)\n",
      "4.338400242348786\n",
      "7.090114687665711\n",
      "[5.097504146075549, 6.94044197628146, 8.002941158939326, 6.601718861363333, 9.48883397139835, 6.661285208587734, 9.459366707173045, 7.953704524889576, 4.33840034234883, 6.368376507637036, 7.0150088747674815, 6.642537150789993, 5.517070292067897, 7.811889534543084, 7.046312126774953, 7.648810761033703, 10.495032947571657, 5.3083393713564515, 6.314604602049042]\n"
     ]
    }
   ],
   "source": [
    "print(hgt_rate)\n",
    "print(sum(tree.total_branch_length for tree in core_tree.trees()))\n",
    "print(average_tree_length)\n",
    "print(total_branch_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "51f906b4-46a8-42de-b7c9-111b00ea1a87",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_extinction_time(L, rho, hgt_rate, num_samples):\n",
    "    \"\"\"\n",
    "    Simuliert die Zeit bis zur Extinktion eines Zweigungsprozesses.\n",
    "    \n",
    "    Parameter:\n",
    "        L: Die Länge des Core-Baums (nicht direkt verwendet)\n",
    "        rho: Mutationsrate (nicht direkt verwendet)\n",
    "        gamma: HGT-Rate (Geburtenrate)\n",
    "        initial_branches: Anzahl der anfänglichen Zweige (X_0)\n",
    "\n",
    "    Rückgabewert:\n",
    "        Zeit bis zur Extinktion\n",
    "    \"\"\"\n",
    "    branches = num_samples\n",
    "    time = 0.0\n",
    "\n",
    "    while branches > 0:\n",
    "        birth_rate = hgt_rate\n",
    "        if branches > 1:\n",
    "            death_rate = branches * (branches - 1) / 2\n",
    "        else:\n",
    "            death_rate = 0\n",
    "\n",
    "        total_rate = birth_rate + death_rate\n",
    "        if total_rate == 0:\n",
    "            break\n",
    "\n",
    "        # Wartezeit bis zum nächsten Ereignis\n",
    "        time_to_next_event = np.random.exponential(1 / total_rate)\n",
    "        time += time_to_next_event\n",
    "\n",
    "        # Entscheide ob Geburt oder Tod\n",
    "        if np.random.rand() < (birth_rate / total_rate):\n",
    "            branches += 1\n",
    "        else:\n",
    "            branches -= 1\n",
    "\n",
    "    return time\n",
    "\n",
    "# Beispielhafte Parameter\n",
    "L = 100  # Beispielkernbaum bei Länge L (nicht direkt verwendet)\n",
    "rho = 0.01  # Mutationsrate (nicht direkt verwendet)\n",
    "gamma = 0.1  # HGT-Rate\n",
    "num_samples = 5  # Anzahl der anfänglichen Zweige\n",
    "\n",
    "# Berechne die Zeit bis zur Extinktion\n",
    "extinction_time = simulate_extinction_time(L, rho, gamma, num_samples)\n",
    "print(f\"Zeit bis zur Extinktion: {extinction_time:.2f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a32258ca-cdf9-4163-90b5-e2ec9034b03d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5.771128597498799, 5.4174097449121525, 6.77904869528533, 6.399279959065085, 8.679570304346026, 8.033348397032984, 9.038932700006043, 5.4174097449121525, 6.78024517220398, 6.613619376945763, 6.298587709217051, 7.107051383156005, 9.306812231266242, 9.747508725826629, 8.028013919945538, 6.422505126013231, 7.621816499411544, 7.5179369690924975, 8.666259532719694]\n",
      "   18              \n",
      " ┏━━┻━━┓           \n",
      " ┃    17           \n",
      " ┃  ┏━━┻━━┓        \n",
      " ┃  ┃    16        \n",
      " ┃  ┃  ┏━━┻━━━┓    \n",
      " ┃  ┃  ┃     15    \n",
      " ┃  ┃  ┃    ┏━┻━━┓ \n",
      " ┃  ┃  ┃    ┃   14 \n",
      " ┃  ┃  ┃    ┃   ┏┻┓\n",
      " ┃  ┃  ┃   13   ┃ ┃\n",
      " ┃  ┃  ┃  ┏━┻┓  ┃ ┃\n",
      " ┃  ┃ 12  ┃  ┃  ┃ ┃\n",
      " ┃  ┃ ┏┻┓ ┃  ┃  ┃ ┃\n",
      " ┃  ┃ ┃ ┃ ┃ 11  ┃ ┃\n",
      " ┃  ┃ ┃ ┃ ┃ ┏┻┓ ┃ ┃\n",
      "10  ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃\n",
      "┏┻┓ ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃\n",
      "0 1 2 3 4 5 6 7 8 9\n",
      "\n",
      "             18    \n",
      "           ┏━━┻━━┓ \n",
      "          17     ┃ \n",
      "        ┏━━┻━━┓  ┃ \n",
      "       16     ┃  ┃ \n",
      "    ┏━━━┻━━┓  ┃  ┃ \n",
      "   15      ┃  ┃  ┃ \n",
      " ┏━━┻━┓    ┃  ┃  ┃ \n",
      "14    ┃    ┃  ┃  ┃ \n",
      "┏┻┓   ┃    ┃  ┃  ┃ \n",
      "┃ ┃  13    ┃  ┃  ┃ \n",
      "┃ ┃  ┏┻━┓  ┃  ┃  ┃ \n",
      "┃ ┃  ┃  ┃ 12  ┃  ┃ \n",
      "┃ ┃  ┃  ┃ ┏┻┓ ┃  ┃ \n",
      "┃ ┃ 11  ┃ ┃ ┃ ┃  ┃ \n",
      "┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃  ┃ \n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃ 10 \n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ ┃ ┏┻┓\n",
      "0 6 1 5 4 3 7 2 8 9\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <div>\n",
       "            <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "            </style>\n",
       "            <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"padding:0;line-height:21px;\">\n",
       "                                    <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                                    <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                                </th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            <tr><td>Trees</td><td>1</td></tr>\n",
       "                            <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "                            <tr><td>Time Units</td><td>unknown</td></tr>\n",
       "                            <tr><td>Sample Nodes</td><td>10</td></tr>\n",
       "                            <tr><td>Total Size</td><td>2.8 KiB</td></tr>\n",
       "                            <tr>\n",
       "                                <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td>\n",
       "                            </tr>\n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"line-height:21px;\">Table</th>\n",
       "                                <th>Rows</th>\n",
       "                                <th>Size</th>\n",
       "                                <th>Has Metadata</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "            <tr>\n",
       "                <td>Edges</td>\n",
       "                <td>30</td>\n",
       "                <td>968 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Individuals</td>\n",
       "                <td>0</td>\n",
       "                <td>24 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Migrations</td>\n",
       "                <td>0</td>\n",
       "                <td>8 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Mutations</td>\n",
       "                <td>14</td>\n",
       "                <td>630 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    ✅\n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Nodes</td>\n",
       "                <td>31</td>\n",
       "                <td>876 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Populations</td>\n",
       "                <td>1</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Provenances</td>\n",
       "                <td>0</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Sites</td>\n",
       "                <td>1</td>\n",
       "                <td>46 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th>Provenance Timestamp</th>\n",
       "                                <th>Software Name</th>\n",
       "                                <th>Version</th>\n",
       "                                <th>Command</th>\n",
       "                                <th>Full record</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "            </div>\n",
       "        </div>\n",
       "    "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x19c3f296870>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(total_branch_lengths)\n",
    "print(mts_list[1].first().draw_text())\n",
    "print(core_tree.first().draw_text())\n",
    "\n",
    "mts"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pangenome-gene-transfer-simulation",
   "language": "python",
   "name": "pangenome-gene-transfer-simulation"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
