{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "4e116c44-d851-4060-897c-a0a0f6c02d45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clonal node: 545\n",
      "Present_mutation at node:  433\n",
      "Number of present genes:  56\n",
      "Success: hgt_rate = 5, Total computation time = 1.165586 seconds.\n",
      "Expected Amount of sites:  303.84373192795744\n",
      "Expected Amount of mutations:  356.50562927530086\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <div>\n",
       "            <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "            </style>\n",
       "            <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"padding:0;line-height:21px;\">\n",
       "                                    <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                                    <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                                </th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            <tr><td>Trees</td><td>1</td></tr>\n",
       "                            <tr><td>Sequence Length</td><td>1,000.0</td></tr>\n",
       "                            <tr><td>Time Units</td><td>unknown</td></tr>\n",
       "                            <tr><td>Sample Nodes</td><td>100</td></tr>\n",
       "                            <tr><td>Total Size</td><td>53.1 KiB</td></tr>\n",
       "                            <tr>\n",
       "                                <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td>\n",
       "                            </tr>\n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"line-height:21px;\">Table</th>\n",
       "                                <th>Rows</th>\n",
       "                                <th>Size</th>\n",
       "                                <th>Has Metadata</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "            <tr>\n",
       "                <td>Edges</td>\n",
       "                <td>144</td>\n",
       "                <td>4.5 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Individuals</td>\n",
       "                <td>0</td>\n",
       "                <td>24 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Migrations</td>\n",
       "                <td>0</td>\n",
       "                <td>8 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Mutations</td>\n",
       "                <td>362</td>\n",
       "                <td>13.7 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Nodes</td>\n",
       "                <td>911</td>\n",
       "                <td>24.9 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Populations</td>\n",
       "                <td>1</td>\n",
       "                <td>16 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Provenances</td>\n",
       "                <td>1</td>\n",
       "                <td>758 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Sites</td>\n",
       "                <td>305</td>\n",
       "                <td>8.0 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th>Provenance Timestamp</th>\n",
       "                                <th>Software Name</th>\n",
       "                                <th>Version</th>\n",
       "                                <th>Command</th>\n",
       "                                <th>Full record</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "                <tr>\n",
       "                    <td>06 September, 2025 at 04:12:59 AM</td>\n",
       "                    <td>msprime</td>\n",
       "                    <td>1.3.1</td>\n",
       "                    <td>sim_mutations</td>\n",
       "                    <td>\n",
       "                        <details>\n",
       "                            <summary>Details</summary>\n",
       "                            \n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\"></span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    schema_version: 1.0.0<br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">software:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    name: msprime<br/>version: 1.3.1<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">parameters:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    command: sim_mutations<br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">tree_sequence:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    __constant__: __current_ts__<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>rate: 0.1<br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">model:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    __class__: msprime.mutations.InfiniteAlle&lt;br/&gt;les<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>start_time: None<br/>end_time: None<br/>discrete_genome: None<br/>keep: True<br/>random_seed: 3017823703<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">environment:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">os:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    system: Linux<br/>node: AsusF<br/>release: 4.4.0-19041-Microsoft<br/>version: #5794-Microsoft Mon Apr 07&lt;br/&gt;17:55:00 PST 2025<br/>machine: x86_64<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">python:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    implementation: CPython<br/>version: 3.12.1<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">libraries:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">kastore:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    version: 2.1.1<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">tskit:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    version: 0.6.0<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">gsl:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    version: 2.7<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            \n",
       "                        </details>\n",
       "                    </td>\n",
       "                </tr>\n",
       "            \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "            </div>\n",
       "        </div>\n",
       "    "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7f0ac82e8680>"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import advanced_simulator_h5\n",
    "import importlib\n",
    "\n",
    "importlib.reload(advanced_simulator_h5)\n",
    "\n",
    "data = advanced_simulator_h5.simulator(num_samples = 100, hgt_rate = 5, rho = 0)\n",
    "\n",
    "mts = data['mts']\n",
    "\n",
    "#mts.draw_svg()\n",
    "gene_trees_list = data['gene_trees_list']\n",
    "\n",
    "tree = gene_trees_list.first()\n",
    "\n",
    "print(\"Expected Amount of sites: \", gene_trees_list.sequence_length*(1-(1-1/gene_trees_list.sequence_length)**gene_trees_list.num_mutations))\n",
    "print(\"Expected Amount of mutations: \", gene_trees_list.sequence_length*0.1*tree.total_branch_length)\n",
    "\n",
    "gene_trees_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "ac372cb7-63b2-437f-9e23-0abe74d6bc61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "363.66148136517995\n",
      "0.30499999999999994\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "362.00000000000006"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = (1-(1-1/gene_trees_list.sequence_length)**gene_trees_list.num_mutations)\n",
    "\n",
    "y = snp_percentage_of_gen_to_absolute_number(gene_trees_list.num_sites/gene_trees_list.sequence_length, gene_trees_list.sequence_length)\n",
    "print(y)\n",
    "\n",
    "z = absolute_number_to_snp_percentage_of_gen(y, gene_length, 0.1)\n",
    "\n",
    "print(z)\n",
    "\n",
    "np.log(1 - x) / np.log(1 - 1/gene_trees_list.sequence_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "f63a2d3c-d386-4c8c-b3a0-ba81514dfa69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1.6809711999999877"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "86.73 - (76.32+13.09-76.32*13.09/1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "01fda2b2-0003-46f5-994a-8ba07cf71168",
   "metadata": {},
   "outputs": [],
   "source": [
    "def snp_percentage_of_gen_to_absolute_number(x, gene_length):\n",
    "    if x == 1:\n",
    "        return np.inf\n",
    "    else:\n",
    "        y = np.log(1 - x) / np.log(1 - 1/gene_length)\n",
    "    return y\n",
    "\n",
    "def absolute_number_to_snp_percentage_of_gen(y, gene_length, nucleotide_mutation_rate):\n",
    "    L = gene_length\n",
    "    mu = nucleotide_mutation_rate\n",
    "    if y == np.inf:\n",
    "        return 1\n",
    "    else:\n",
    "        return 1 - np.exp(y * np.log(1 - 1/L))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b2cacc11-1101-4db5-b79e-e4aad8cd004c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import msprime\n",
    "import tskit\n",
    "import hgt_simulation\n",
    "import hgt_sim_args\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import random\n",
    "import torch\n",
    "import os\n",
    "import re\n",
    "import secrets\n",
    "import copy\n",
    "import json\n",
    "import h5py\n",
    "import uuid\n",
    "import glob\n",
    "import shutil\n",
    "import pickle\n",
    "import time\n",
    "\n",
    "from random import randint\n",
    "from collections import namedtuple\n",
    "from collections import defaultdict\n",
    "from collections import deque\n",
    "from sbi.utils import BoxUniform\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from numba import njit\n",
    "from typing import Union, Sequence, List, Tuple\n",
    "from networkx.readwrite import json_graph\n",
    "\n",
    "from concurrent.futures import ProcessPoolExecutor\n",
    "\n",
    "def distance_core(tree_sequence: tskit.TreeSequence, num_samples: int) -> np.ndarray:\n",
    "    \n",
    "    tree = tree_sequence.first()\n",
    "    distance_matrix = np.zeros((num_samples, num_samples))\n",
    "\n",
    "    for i in range(num_samples):\n",
    "        time_i = tree.get_time(i)\n",
    "        for j in range(i):\n",
    "            time_j = tree.get_time(j)\n",
    "            mrca = tree.mrca(i, j)\n",
    "            time_mrca = tree.get_time(mrca)\n",
    "\n",
    "            # Distance is the time between each sample and their MRCA\n",
    "            distance = (time_mrca - time_i) + (time_mrca - time_j)\n",
    "            distance_matrix[i, j] = distance\n",
    "            distance_matrix[j, i] = distance  # symmetric\n",
    "\n",
    "    return distance_matrix\n",
    "\n",
    "def multidimensional_scaling(D, multidimensional_scaling_dimensions=100):\n",
    "\n",
    "    # 1. Quadrat der Abstände\n",
    "    D2 = D ** 2\n",
    "\n",
    "    # 2. Zentrierungsmatrix\n",
    "    n = D.shape[0]\n",
    "    J = np.eye(n) - np.ones((n, n)) / n\n",
    "\n",
    "    # 3. Doubly centered inner product matrix\n",
    "    B = -0.5 * J @ D2 @ J\n",
    "\n",
    "    # 4. Eigenzerlegung\n",
    "    eigvals, eigvecs = np.linalg.eigh(B)\n",
    "\n",
    "    # 5. Sortieren (absteigend)\n",
    "    idx = np.argsort(eigvals)[::-1]\n",
    "    eigvals = eigvals[idx]\n",
    "    eigvecs = eigvecs[:, idx]\n",
    "\n",
    "    # 6. Komponenten auswählen\n",
    "    L = np.diag(np.sqrt(np.maximum(eigvals[:multidimensional_scaling_dimensions], 0)))\n",
    "    V = eigvecs[:, :multidimensional_scaling_dimensions]\n",
    "\n",
    "    out = V @ L  # → [n, multidimensional_scaling_dimensions]\n",
    "\n",
    "    if out.shape[1] < multidimensional_scaling_dimensions:\n",
    "        padding = np.zeros((out.shape[0], multidimensional_scaling_dimensions - out.shape[1]))\n",
    "        out = np.hstack([out, padding])\n",
    "    \n",
    "    return out\n",
    "\n",
    "def count_blocks_sym(mat: np.ndarray,\n",
    "                     threshold: Union[float, Sequence[float]] = 0.05\n",
    "                    ) -> List[Tuple[np.ndarray, int]]:\n",
    "\n",
    "    assert mat.ndim == 2 and mat.shape[0] == mat.shape[1], \"Matrix muss quadratisch sein.\"\n",
    "    thresholds = np.atleast_1d(threshold).astype(np.float64)\n",
    "\n",
    "    if mat.size == 0:\n",
    "        return np.zeros_like(thresholds)\n",
    "    \n",
    "    # Vorarbeit außerhalb der Schleife\n",
    "    idx = np.where(~np.isnan(mat[0]))[0]\n",
    "\n",
    "    #results_matrices = []\n",
    "    results_number_of_clusters = []\n",
    "    for th in thresholds:\n",
    "        labels, k = _count_blocks_single(mat, idx, th)\n",
    "        #results_matrices.append(labels)\n",
    "        results_number_of_clusters.append(int(k))\n",
    "    results_number_of_clusters = np.array(results_number_of_clusters, dtype=int)\n",
    "    #return results_matrices, results_number_of_clusters\n",
    "    return results_number_of_clusters\n",
    "\n",
    "def _count_blocks_single(mat, idx, threshold):\n",
    "    n = idx.size\n",
    "    label_matrix = -np.ones((n, n), dtype=np.int32)\n",
    "    num_labels   = 0\n",
    "\n",
    "    for ii in range(n):\n",
    "        i = idx[ii]\n",
    "        for jj in range(ii, n):\n",
    "            j = idx[jj]\n",
    "            if ii == 0 and jj == 0:\n",
    "                label_matrix[ii, jj] = 0\n",
    "                num_labels += 1\n",
    "            elif ii == 0:\n",
    "                if abs(mat[i, j] - mat[i, idx[jj-1]]) < threshold:\n",
    "                    label_matrix[ii, jj] = label_matrix[ii, jj-1]\n",
    "                else:\n",
    "                    label_matrix[ii, jj] = num_labels\n",
    "                    num_labels += 1\n",
    "            else:\n",
    "                if abs(mat[i, j] - mat[idx[ii-1], j]) < threshold:\n",
    "                    label_matrix[ii, jj] = label_matrix[ii-1, jj]\n",
    "                elif abs(mat[i, j] - mat[i, idx[jj-1]]) < threshold:\n",
    "                    label_matrix[ii, jj] = label_matrix[ii, jj-1]\n",
    "                else:\n",
    "                    label_matrix[ii, jj] = num_labels\n",
    "                    num_labels += 1\n",
    "    return label_matrix, num_labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "04482f70-21fd-4c76-bbd4-6d1e8edeaa49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clonal node: 240\n",
      "Present_mutation at node:  209\n",
      "Number of present genes:  23\n",
      "Success: hgt_rate = 1, Total computation time = 0.507256 seconds.\n"
     ]
    }
   ],
   "source": [
    "num_samples = 100\n",
    "theta = 1\n",
    "rho = 1\n",
    "hgt_rate = 1\n",
    "num_genes = 1\n",
    "nucleotide_mutation_rate = None\n",
    "gene_length = 1000\n",
    "pca_dimensions = 10\n",
    "ce_from_nwk = None\n",
    "seed = None\n",
    "distance_matrix = None\n",
    "multidimensional_scaling_dimensions = 100\n",
    "block_clustering_threshold = np.arange(0, 1.0001, 0.025)\n",
    "clonal_root_mutation = None\n",
    "hgt_difference_removal_threshold = 0\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "if ce_from_nwk is None and num_samples is None:\n",
    "    raise ValueError(\n",
    "        \"Neither a core tree or parameters for simulation were provided. Choose either.\"\n",
    "    )\n",
    "\n",
    "if ce_from_nwk is None:\n",
    "    core_tree = msprime.sim_ancestry(\n",
    "            samples=num_samples,\n",
    "            sequence_length=1,\n",
    "            ploidy=1,\n",
    "            recombination_rate=0,\n",
    "            gene_conversion_rate=0,\n",
    "            gene_conversion_tract_length=1,  # One gene\n",
    "            random_seed=seed,\n",
    "        )\n",
    "\n",
    "    ce_from_nwk = core_tree.first().newick()\n",
    "\n",
    "random.seed(seed)\n",
    "np.random.seed(seed)\n",
    "\n",
    "if seed is None:\n",
    "    seed = secrets.randbelow(2**32 - 4) + 2\n",
    "\n",
    "if nucleotide_mutation_rate is None:\n",
    "    #nucleotide_mutation_rate = np.random.uniform(0.033, 0.066)\n",
    "    nucleotide_mutation_rate = np.random.uniform(0.05)\n",
    "\n",
    "if num_samples < multidimensional_scaling_dimensions:\n",
    "    multidimensional_scaling_dimensions = num_samples\n",
    "\n",
    "\n",
    "### Calculate hgt events:\n",
    "\n",
    "args = hgt_sim_args.Args(\n",
    "    sample_size=num_samples,\n",
    "    num_sites=num_genes,\n",
    "    gene_conversion_rate=0,\n",
    "    recombination_rate=0,\n",
    "    hgt_rate=hgt_rate,\n",
    "    ce_from_ts=None,\n",
    "    ce_from_nwk=ce_from_nwk,\n",
    "    random_seed=seed,\n",
    "    #random_seed=84,\n",
    ")\n",
    "\n",
    "ts, hgt_edges = hgt_simulation.run_simulate(args)\n",
    "        \n",
    "### Place mutations\n",
    "\n",
    "alleles = [\"absent\", \"present\"]\n",
    "\n",
    "# Place one mutation per site, e.g. gen\n",
    "\n",
    "gains_model = msprime.MatrixMutationModel(\n",
    "    alleles = alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [0, 1],\n",
    "        [0, 1],\n",
    "    ],\n",
    ")\n",
    "\n",
    "ts_gains = msprime.sim_mutations(ts, rate=1, model = gains_model, keep = True, random_seed=seed)\n",
    "\n",
    "k = 1\n",
    "while (ts_gains.num_sites < ts_gains.sequence_length):\n",
    "    ts_gains = msprime.sim_mutations(ts_gains, rate=1, model = gains_model, keep = True, random_seed=seed+k)\n",
    "    k = k+1\n",
    "\n",
    "# Remove superfluous mutations\n",
    "\n",
    "tables = ts_gains.dump_tables()\n",
    "\n",
    "mutations_by_site = {}\n",
    "for mut in tables.mutations:\n",
    "    if mut.site not in mutations_by_site:\n",
    "        mutations_by_site[mut.site] = []\n",
    "    mutations_by_site[mut.site].append(mut)\n",
    "\n",
    "tables.mutations.clear()\n",
    "\n",
    "clonal_root_node = ts_gains.first().mrca(*list(range(num_samples)))\n",
    "\n",
    "print(\"Clonal node:\", clonal_root_node)\n",
    "\n",
    "if clonal_root_mutation == None:\n",
    "    clonal_root_mutation = random.random() < 0.5\n",
    "\n",
    "present_mutations = [[] for i in range(num_genes)]\n",
    "for site, mutations in mutations_by_site.items():\n",
    "\n",
    "    if clonal_root_mutation == True:\n",
    "        tables.mutations.add_row(\n",
    "            site=site,\n",
    "            node=clonal_root_node,\n",
    "            derived_state=\"present\",\n",
    "            parent=-1,\n",
    "            metadata=None,\n",
    "            time=tables.nodes.time[clonal_root_node],\n",
    "        )\n",
    "        \n",
    "    else:\n",
    "        selected_mutation = random.choice(mutations)\n",
    "        tables.mutations.add_row(\n",
    "            site=selected_mutation.site,\n",
    "            node=selected_mutation.node,\n",
    "            derived_state=selected_mutation.derived_state,\n",
    "            parent=-1,\n",
    "            metadata=None,\n",
    "            time=selected_mutation.time,\n",
    "        )\n",
    "        present_mutations[site].append(selected_mutation)\n",
    "\n",
    "        # Mutation at clonal root:\n",
    "        tables.mutations.add_row(\n",
    "            site=site,\n",
    "            node=clonal_root_node,\n",
    "            derived_state=\"absent\",\n",
    "            parent=-1,\n",
    "            metadata=None,\n",
    "            time=tables.nodes.time[clonal_root_node],\n",
    "        )\n",
    "\n",
    "    # Add sentinel mutations at the leafs:\n",
    "\n",
    "    for leaf_position in range(num_samples):\n",
    "        tables.mutations.add_row(\n",
    "            site = site,\n",
    "            node = leaf_position,\n",
    "            derived_state = \"absent\",\n",
    "            time = 0.00000000001,\n",
    "        )\n",
    "\n",
    "tables.sort()\n",
    "ts_gains = tables.tree_sequence()\n",
    "\n",
    "\n",
    "# Place losses:\n",
    "\n",
    "losses_model = msprime.MatrixMutationModel(\n",
    "    alleles = alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [1, 0],\n",
    "        [1, 0],\n",
    "    ],\n",
    ")\n",
    "\n",
    "ts_gains_losses = msprime.sim_mutations(ts_gains, rate = rho, model = losses_model, keep = True, random_seed=seed-1)\n",
    "\n",
    "### Calculate the gene absence presence matrix:\n",
    "\n",
    "MutationRecord = namedtuple('MutationRecord', ['site_id', 'mutation_id', 'node', 'is_hgt'])\n",
    "\n",
    "tables = ts_gains_losses.dump_tables()\n",
    "\n",
    "tables.mutations.clear() # SIMPLE VERSION!\n",
    "\n",
    "hgt_parent_nodes = [edge.parent-1 for edge in hgt_edges]\n",
    "hgt_children_nodes = [edge.child for edge in hgt_edges]\n",
    "hgt_parent_children = defaultdict(list)\n",
    "\n",
    "for parent in hgt_parent_nodes:\n",
    "    hgt_parent_children[parent].append(parent-1)\n",
    "\n",
    "\n",
    "\n",
    "# Initialize tables for the diversity of each gene\n",
    "tables_gene = tskit.TableCollection(sequence_length=gene_length)\n",
    "tables_gene.nodes.replace_with(ts_gains_losses.tables.nodes)\n",
    "tables_gene.populations.replace_with(ts_gains_losses.tables.populations)\n",
    "\n",
    "\n",
    "tables_gene_list = [copy.deepcopy(tables_gene) for _ in range(num_genes)]\n",
    "gene_number_hgt_events_passed = [0 for _ in range(num_genes)]\n",
    "gene_number_loss_events = [0 for _ in range(num_genes)]\n",
    "gene_nodes_loss_events = [[] for _ in range(num_genes)]\n",
    "gene_trees_list = []\n",
    "\n",
    "\n",
    "for tree in ts_gains_losses.trees():\n",
    "\n",
    "    for site in tree.sites():\n",
    "        \n",
    "        hgt_parent_children_passed = [False] * ts_gains_losses.num_nodes\n",
    "        mutations = site.mutations\n",
    "        #site = list(tree.sites())[0]\n",
    "        \n",
    "        present_mutation = [m for m in mutations if m.derived_state == \"present\"][0]\n",
    "        absent_mutation_nodes = {m.node for m in mutations if m.derived_state == \"absent\"}\n",
    "\n",
    "        absent_mutations = defaultdict(list)\n",
    "        for m in mutations:\n",
    "            if m.derived_state == \"absent\" and m.time < present_mutation.time:\n",
    "                absent_mutations[m.node].append(m)\n",
    "\n",
    "        branching_nodes_reached_before = defaultdict(list)\n",
    "        for node_id in range(ts_gains_losses.num_nodes):\n",
    "            branching_nodes_reached_before[node_id] = False\n",
    "    \n",
    "        print(\"Present_mutation at node: \", present_mutation.node)\n",
    "    \n",
    "        branching_nodes_to_process = deque([(present_mutation.node, False, False, 0)])\n",
    "        # The second variable describes if a hgt edge was passed the whole way down to the actual node. \n",
    "        # The third describes if a hgt edge was passed in the last step.\n",
    "        # The fourth is the number of hgt edges that were passed.\n",
    "    \n",
    "        child_mutations = []\n",
    "    \n",
    "        if present_mutation.node < num_samples: # Gain directly above leaf:\n",
    "            if present_mutation.id == sorted([mut for mut in mutations if mut.node == present_mutation.node], key=lambda m: m.time)[1].id:\n",
    "                sentinel_mutation = min([mut for mut in absent_mutations[present_mutation.node]], key=lambda m: m.time)\n",
    "                child_mutations.append(MutationRecord(\n",
    "                    site_id=site.id,\n",
    "                    mutation_id=sentinel_mutation.id,\n",
    "                    node=sentinel_mutation.node,\n",
    "                    is_hgt=False\n",
    "                ))\n",
    "                if not tree.parent(present_mutation.node) == -1: # Can occur if its a clonal root mutation\n",
    "                    tables_gene_list[site.id].edges.add_row(\n",
    "                            left = 0, right = gene_length, parent = tree.parent(present_mutation.node), child = sentinel_mutation.node\n",
    "                    )\n",
    "            else:\n",
    "                gene_number_loss_events[site.id] += 1\n",
    "                gene_nodes_loss_events[site.id].append(present_mutation.node)\n",
    "    \n",
    "        else: \n",
    "            \n",
    "            # To see, which HGT edges are passed, we have to go through the tree two times. \n",
    "            # First, we detect all passed HGT edges, then we calculate the presence of mutations in the leaves.\n",
    "    \n",
    "            # First time going through:\n",
    "            while branching_nodes_to_process:\n",
    "                \n",
    "                last_branching_node = branching_nodes_to_process.popleft()\n",
    "                selected_branch_nodes_to_process = deque([last_branching_node])\n",
    "                branching_nodes_reached_before[tree.parent(last_branching_node[0])] = True\n",
    "                \n",
    "                while selected_branch_nodes_to_process:\n",
    "                \n",
    "                    child_node = selected_branch_nodes_to_process.popleft()    \n",
    "        \n",
    "                    # If there is a mutation on the edge, find the earliest one.\n",
    "                    if not child_node[2] and child_node[0] in absent_mutation_nodes:\n",
    "                        \n",
    "                        absent_mutation_after_gain_at_node = absent_mutations[child_node[0]]\n",
    "                        \n",
    "                        if not absent_mutation_after_gain_at_node: # empty\n",
    "                            children = tree.children(child_node[0])\n",
    "                            if len(children) > 1:\n",
    "                                for child in reversed(children):\n",
    "                                    if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                        branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                                    #else:\n",
    "                                    #    print(\"Child passed before\", \" Child: \", child)\n",
    "                            else:\n",
    "                                for child in reversed(children):\n",
    "                                    selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                            if hgt_parent_children[child_node[0]]:\n",
    "                                selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])\n",
    "                                hgt_parent_children_passed[hgt_parent_children[child_node[0]][0]] = True # The child of the hgt_edge is marked\n",
    "                                #print(\"HGT at Node: \", hgt_parent_children[child_node[0]][0])\n",
    "                            \n",
    "            \n",
    "                    # If there is no mutation, add child nodes.\n",
    "                    else:\n",
    "                        children = tree.children(child_node[0])\n",
    "                        if len(children) > 1:\n",
    "                            for child in reversed(children):\n",
    "                                #print(\"Child: \", child)\n",
    "                                if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                    branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                                #else:\n",
    "                                #    print(\"Child passed before\", \" Child: \", child)\n",
    "                        else:\n",
    "                            for child in reversed(children):\n",
    "                                #print(\"Child: \", child)\n",
    "                                selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                        if hgt_parent_children[child_node[0]]:\n",
    "                            selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])\n",
    "                            hgt_parent_children_passed[hgt_parent_children[child_node[0]][0]] = True # The child of the hgt_edge is marked\n",
    "                            #print(\"HGT at Node: \", hgt_parent_children[child_node[0]][0])\n",
    "    \n",
    "            # Second time going through:\n",
    "            \n",
    "            branching_nodes_to_process = deque([(present_mutation.node, False, False, 0)])\n",
    "\n",
    "            branching_nodes_reached_before = defaultdict(list)\n",
    "            for node_id in range(ts_gains_losses.num_nodes):\n",
    "                branching_nodes_reached_before[node_id] = False\n",
    "            \n",
    "            while branching_nodes_to_process:\n",
    "                \n",
    "                last_branching_node = branching_nodes_to_process.popleft()\n",
    "                selected_branch_nodes_to_process = deque([last_branching_node])\n",
    "                branching_nodes_reached_before[tree.parent(last_branching_node[0])] = True\n",
    "                \n",
    "                while selected_branch_nodes_to_process:\n",
    "                \n",
    "                    child_node = selected_branch_nodes_to_process.popleft()\n",
    "    \n",
    "                    if not child_node[2] and hgt_parent_children_passed[child_node[0]]:\n",
    "                        #print(\"Incoming HGT edge registered at node: \", child_node[0])\n",
    "                        continue\n",
    "        \n",
    "                    # If there is a mutation on the edge, find the earliest one.\n",
    "                    if not child_node[2] and child_node[0] in absent_mutation_nodes:\n",
    "                        \n",
    "                        absent_mutation_after_gain_at_node = absent_mutations[child_node[0]]\n",
    "                        \n",
    "                        if absent_mutation_after_gain_at_node: # not empty\n",
    "                            earliest_mutation = max(\n",
    "                                absent_mutations[child_node[0]], \n",
    "                                key=lambda m: m.time\n",
    "                            )\n",
    "            \n",
    "                            if earliest_mutation.time == 0.00000000001:\n",
    "                                earliest_mutation.derived_state = \"present\"\n",
    "                                child_mutations.append(MutationRecord(\n",
    "                                    site_id=site.id,\n",
    "                                    mutation_id=earliest_mutation.id,\n",
    "                                    node=earliest_mutation.node,\n",
    "                                    is_hgt=child_node[1]\n",
    "                                ))\n",
    "                                tables_gene_list[site.id].edges.add_row(\n",
    "                                    left = 0, right = gene_length, parent = tree.parent(last_branching_node[0]), child = earliest_mutation.node\n",
    "                                )\n",
    "                                gene_number_hgt_events_passed[site.id] += child_node[3]\n",
    "                            else:\n",
    "                                gene_number_loss_events[site.id] += 1\n",
    "                                gene_nodes_loss_events[site.id].append(earliest_mutation.node)\n",
    "                        else:\n",
    "                            children = tree.children(child_node[0])\n",
    "                            if len(children) > 1:\n",
    "                                for child in reversed(children):\n",
    "                                    if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                        branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                                if not tree.parent(last_branching_node[0]) == -1:\n",
    "                                    tables_gene_list[site.id].edges.add_row(\n",
    "                                            left = 0, right = gene_length, parent = tree.parent(last_branching_node[0]), child = child_node[0]\n",
    "                                    )\n",
    "                                gene_number_hgt_events_passed[site.id] += child_node[3]\n",
    "                            else:\n",
    "                                for child in reversed(children):\n",
    "                                    selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                            if hgt_parent_children[child_node[0]]:\n",
    "                                selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])  \n",
    "                            \n",
    "            \n",
    "                    # If there is no mutation, add child nodes.\n",
    "                    else:\n",
    "                        children = tree.children(child_node[0])\n",
    "                        if len(children) > 1:\n",
    "                            for child in reversed(children):\n",
    "                                if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                    branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                            if not tree.parent(last_branching_node[0]) == -1:\n",
    "                                tables_gene_list[site.id].edges.add_row(\n",
    "                                        left = 0, right = gene_length, parent = tree.parent(last_branching_node[0]), child = child_node[0]\n",
    "                                )\n",
    "                            gene_number_hgt_events_passed[site.id] += child_node[3]\n",
    "                        else:\n",
    "                            for child in reversed(children):\n",
    "                                selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                        if hgt_parent_children[child_node[0]]:\n",
    "                            selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])\n",
    "\n",
    "        child_mutations.sort(key=lambda mut: not mut.is_hgt) # Will set is_hgt to False later if there are paths without hgt events to the leaf.\n",
    "    \n",
    "        # We have to adress multiple paths to the same destiny, some with hgt and other without it:\n",
    "        unique_mutations = {}\n",
    "    \n",
    "        for mut in child_mutations:\n",
    "            if mut.node not in unique_mutations:\n",
    "                unique_mutations[mut.node] = mut\n",
    "            else:\n",
    "                existing_mut = unique_mutations[mut.node]\n",
    "                if not existing_mut.is_hgt or not mut.is_hgt:\n",
    "                    unique_mutations[mut.node] = mut._replace(is_hgt=False)\n",
    "        \n",
    "        child_mutations_filtered = list(unique_mutations.values())\n",
    "\n",
    "        for mutation in mutations:\n",
    "            if mutation.time > 0.00000000001:\n",
    "                if mutation.derived_state == \"absent\":\n",
    "                    metadata_value = bytes([3]) \n",
    "                elif mutation.derived_state == \"present\":\n",
    "                    metadata_value = bytes([7])\n",
    "                tables.mutations.add_row(\n",
    "                    site=site.id,\n",
    "                    node=mutation.node,\n",
    "                    derived_state=mutation.derived_state,\n",
    "                    parent=-1,\n",
    "                    metadata=metadata_value,\n",
    "                    time=mutation.time,\n",
    "                )\n",
    "        \n",
    "        for mutation in child_mutations_filtered:\n",
    "    \n",
    "            tables.mutations.add_row(\n",
    "                site=site.id,\n",
    "                node=mutation.node,\n",
    "                derived_state=\"present\",\n",
    "                parent=-1,\n",
    "                metadata=bytes([mutation.is_hgt]),\n",
    "                time=0.00000000001,\n",
    "            )\n",
    "    \n",
    "\n",
    "mts = tables.tree_sequence()\n",
    "\n",
    "for i in range(num_genes): \n",
    "\n",
    "    tables_gene_list[i].sort()\n",
    "    \n",
    "    gene_trees_list.append(tables_gene_list[i].tree_sequence())\n",
    "\n",
    "### Compute the gene presence and absence:\n",
    "\n",
    "gene_absence_presence_matrix = []\n",
    "\n",
    "for var in mts.variants():\n",
    "    gene_absence_presence_matrix.append(var.genotypes)\n",
    "gene_absence_presence_matrix = np.array(gene_absence_presence_matrix)\n",
    "\n",
    "print(\"Number of present genes: \", sum(gene_absence_presence_matrix[0]))\n",
    "\n",
    "#fitch_scores = fitch_parsimony_score(mts, gene_absence_presence_matrix)\n",
    "\n",
    "### Calculate the HGT events:\n",
    "\n",
    "gene_number_hgt_events_passed = [0 for _ in range(num_genes)]\n",
    "nodes_hgt_events = [[] for _ in range(num_genes)]\n",
    "hgt_trees = [[] for _ in range(num_genes)]\n",
    "\n",
    "for tree in mts.trees():\n",
    "    for site in tree.sites():\n",
    "        reached_nodes_from_leaves = defaultdict(list)\n",
    "        for node_id in range(mts.num_nodes):\n",
    "            reached_nodes_from_leaves[node_id] = False\n",
    "        \n",
    "        stack = []\n",
    "        for node_id in range(mts.num_samples):\n",
    "            if gene_absence_presence_matrix[site.id][node_id]:\n",
    "                stack.append((node_id, node_id))\n",
    "                \n",
    "        while stack:\n",
    "            node, leaf_node = stack.pop()\n",
    "            parent = tree.parent(node)\n",
    "            if not hgt_parent_children_passed[node] and not reached_nodes_from_leaves[node] and node < clonal_root_node:\n",
    "                stack.append((parent, leaf_node))\n",
    "                reached_nodes_from_leaves[node] = True\n",
    "            elif hgt_parent_children_passed[node] and not reached_nodes_from_leaves[node] and node < clonal_root_node:\n",
    "                #print(\"HGT event spotted at node: \", node)\n",
    "                nodes_hgt_events[site.id].append((node, leaf_node))\n",
    "                gene_number_hgt_events_passed[site.id] += 1\n",
    "                reached_nodes_from_leaves[node] = True\n",
    "\n",
    "        tables = gene_trees_list[site.id].tables\n",
    "        for subtree in gene_trees_list[site.id].trees():\n",
    "            for node, leaf_node in nodes_hgt_events[site.id]:\n",
    "                temp_time = 0\n",
    "                parent = leaf_node\n",
    "                time_of_node = gene_trees_list[site.id].node(node).time\n",
    "                while temp_time < time_of_node:\n",
    "                    child = parent\n",
    "                    parent = subtree.parent(parent)\n",
    "                    temp_time = gene_trees_list[site.id].node(parent).time\n",
    "\n",
    "                mask = [\n",
    "                    not (edge.parent == parent and edge.child == child)\n",
    "                    for edge in tables.edges\n",
    "                ]\n",
    "                \n",
    "                tables.edges.set_columns(\n",
    "                    left=[e.left for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                    right=[e.right for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                    parent=[e.parent for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                    child=[e.child for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                )\n",
    "                \n",
    "                tables.edges.add_row(left = 0, right = gene_length, parent=node, child=child)\n",
    "                tables.edges.add_row(left = 0, right = gene_length, parent=parent, child=node)\n",
    "\n",
    "        tables.sort()\n",
    "        gene_trees_list[site.id] = tables.tree_sequence()\n",
    "\n",
    "        valid_nodes_hgt_events = []\n",
    "        for subtree in gene_trees_list[site.id].trees():\n",
    "            for node, leaf_node in nodes_hgt_events[site.id]:\n",
    "\n",
    "                tables = mts.tables\n",
    "                \n",
    "                temp_time = 0\n",
    "                parent = leaf_node\n",
    "                time_of_node = gene_trees_list[site.id].node(node).time\n",
    "\n",
    "                mask = [\n",
    "                    not (edge.parent == tree.parent(node) and edge.child == node)\n",
    "                    for edge in tables.edges\n",
    "                ]\n",
    "\n",
    "                tables.edges.set_columns(\n",
    "                    left=[e.left for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                    right=[e.right for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                    parent=[e.parent for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                    child=[e.child for i, e in enumerate(tables.edges) if mask[i]],\n",
    "                )\n",
    "                \n",
    "                tables.edges.add_row(left = site.id, right = site.id+1, parent=subtree.parent(node), child=node)\n",
    "\n",
    "                tables.sort()\n",
    "                tables.mutations.clear()\n",
    "\n",
    "                hgt_tree = tables.tree_sequence()\n",
    "\n",
    "                hgt_trees[site.id].append((hgt_tree, (node, leaf_node)))\n",
    "                \n",
    "                keep = False\n",
    "                for leaf in range(num_samples):\n",
    "                    if gene_absence_presence_matrix[site.id][leaf] == 1 and abs(\n",
    "                        hgt_tree.node(hgt_tree.at(hgt_tree.sites()[site.id].position).mrca(node, leaf)).time - mts.node(tree.mrca(node, leaf)).time) > hgt_difference_removal_threshold:\n",
    "                        keep = True\n",
    "                        break\n",
    "                if keep:\n",
    "                    valid_nodes_hgt_events.append((node, leaf_node))\n",
    "                else:\n",
    "                    print(\"Removed HGT edge\")\n",
    "        nodes_hgt_events[site.id] = valid_nodes_hgt_events\n",
    "            \n",
    "\n",
    "# Simulate the tree for each gene:\n",
    "\n",
    "\"\"\"\n",
    "nucleotide_mutation = msprime.MatrixMutationModel(\n",
    "        alleles=[\"C\", \"M1\", \"M2\", \"M3\"],\n",
    "        root_distribution=[1.0, 0.0, 0.0, 0.0],  # nur C als Wurzel\n",
    "        transition_matrix=[\n",
    "            # C     M1     M2     M3\n",
    "            [ 0.0,  1/3,  1/3,  1/3 ], \n",
    "            [ 1/3,  0,    1/3,  1/3 ],  \n",
    "            [ 1/3,  1/3,  0.0,  1/3 ],  \n",
    "            [ 1/3,  1/3,  1/3,  0.0 ],  \n",
    "        ]\n",
    "    )\n",
    "\"\"\"\n",
    "nucleotide_mutation = msprime.MatrixMutationModel(\n",
    "    alleles=[\"C\", \"M\"],\n",
    "    root_distribution=[1.0, 0.0],  # nur C als Wurzel\n",
    "    transition_matrix=[\n",
    "        # C     M     \n",
    "        [ 0.0,  1 ], \n",
    "        [ 1/3,  2/3],  \n",
    "    ]\n",
    ")\n",
    "\n",
    "\n",
    "nucleotide_mutation = msprime.InfiniteAlleles()\n",
    "\n",
    "\n",
    "for i in range(num_genes): \n",
    "    \n",
    "    gene_trees_list[i] = msprime.sim_mutations(gene_trees_list[i], \n",
    "                                               rate = nucleotide_mutation_rate, model = nucleotide_mutation, keep = True, random_seed=seed-1)\n",
    "\n",
    "# Calculate the different alleles:\n",
    "\n",
    "alleles_list = []\n",
    "\n",
    "for i in range(num_genes):\n",
    "    alleles_list.append([])\n",
    "    for var in gene_trees_list[i].variants():\n",
    "        alleles_list[i].append(var.genotypes)\n",
    "    alleles_list[i] = np.array(alleles_list[i])\n",
    "\n",
    "### Calculate the distances between the leaves\n",
    "\n",
    "if distance_matrix is None:\n",
    "    distance_matrix = distance_core(tree_sequence = mts, num_samples = num_samples)\n",
    "scaled_distance_matrix = multidimensional_scaling(distance_matrix, multidimensional_scaling_dimensions = multidimensional_scaling_dimensions)\n",
    "\n",
    "# Distances in alleles (no pca)\n",
    "\n",
    "alleles_list_pca = []\n",
    "core_allel_distance_list = []\n",
    "core_allel_distance_number_of_clusters = []\n",
    "\n",
    "for A in alleles_list:\n",
    "\n",
    "    if A.shape[0] == 0:\n",
    "        A_reconstructed = np.full((num_samples, multidimensional_scaling_dimensions), -1)\n",
    "        alleles_list_pca.append(A_reconstructed)\n",
    "        continue\n",
    "\n",
    "    gene_absent_vector = (A[0] == -1)\n",
    "    gene_present_vector = ~gene_absent_vector\n",
    "\n",
    "    # Berechne die paarweisen Hamming-Distanzen\n",
    "    euclidean_distances = squareform(pdist(A.T, metric='euclidean')) # Will be transformed into Hamming distances later.\n",
    "    \n",
    "    scaled_euclidean_distances = np.full((num_samples, multidimensional_scaling_dimensions), 0.0)\n",
    "    \n",
    "    if gene_present_vector.sum() > 0:\n",
    "        scaled_euclidean_distances[gene_present_vector, :] = multidimensional_scaling(\n",
    "            euclidean_distances[np.ix_(gene_present_vector, gene_present_vector)], multidimensional_scaling_dimensions = multidimensional_scaling_dimensions)\n",
    "\n",
    "    euclidean_distances[gene_absent_vector, : ] = -1\n",
    "    euclidean_distances[:, gene_absent_vector ] = -1\n",
    "\n",
    "    alleles_list_pca.append(scaled_euclidean_distances)\n",
    "\n",
    "    distance_matrix_valid = distance_matrix[np.ix_(gene_present_vector, gene_present_vector)]\n",
    "    euclidean_distances_valid = euclidean_distances[np.ix_(gene_present_vector, gene_present_vector)]\n",
    "    euclidean_distances_valid = euclidean_distances_valid ** 2 # Return to Hamming distances.\n",
    "    \n",
    "    if gene_present_vector.sum() > 0 and np.max(euclidean_distances_valid) > 0:\n",
    "        euclidean_distances_valid = euclidean_distances_valid / np.mean(euclidean_distances_valid) * np.mean(distance_matrix_valid)\n",
    "\n",
    "    core_allel_distance_valid = distance_matrix_valid - euclidean_distances_valid\n",
    "\n",
    "    core_allel_distance_list.append(core_allel_distance_valid)\n",
    "\n",
    "    core_allel_distance_cluster_results = count_blocks_sym(core_allel_distance_valid, block_clustering_threshold)\n",
    "    \n",
    "    core_allel_distance_number_of_clusters.append(core_allel_distance_cluster_results)\n",
    "\n",
    "### Construct the graph:\n",
    "\n",
    "tree = mts.first()\n",
    "children = defaultdict(list)\n",
    "for leaf in range(num_samples):\n",
    "    node = leaf\n",
    "    while node < clonal_root_node and len(children[node]) < 2:\n",
    "        parent = tree.parent(node)\n",
    "        children[parent].append(node)\n",
    "        node = parent\n",
    "        if parent == tskit.NULL:\n",
    "            break\n",
    "              \n",
    "for node in list(children):\n",
    "    if len(children[node]) == 1:\n",
    "        parent = tree.parent(node)\n",
    "        if parent == tskit.NULL:\n",
    "            continue  # kein Parent → überspringen\n",
    "        child = children[node][0]  # das eine Kind\n",
    "\n",
    "        # 1. altes Kind durch neues Kind ersetzen\n",
    "        if node in children[parent]:\n",
    "            children[parent].remove(node)\n",
    "        children[parent].append(child)\n",
    "\n",
    "        # 2. entferne den Zwischenknoten\n",
    "        del children[node]\n",
    "\n",
    "parents = defaultdict(int)\n",
    "for parent, child_list in children.items():\n",
    "    for child in child_list:\n",
    "        parents[child] = parent\n",
    "\n",
    "node_to_leaf = defaultdict(list)\n",
    "for child in sorted(list(parents)):\n",
    "    if child < num_samples:\n",
    "        node_to_leaf[child] = [child]\n",
    "    node_to_leaf[parents[child]].extend(node_to_leaf[child])\n",
    "\n",
    "Graphs = []\n",
    "graph_properties = []\n",
    "parental_nodes_hgt_events = [[] for _ in range(num_genes)]\n",
    "parental_nodes_hgt_events_corrected = [[] for _ in range(num_genes)]\n",
    "children_gene_nodes_loss_events = [[] for _ in range(num_genes)]\n",
    "\n",
    "for tree in mts.trees():\n",
    "    for site in tree.sites():\n",
    "        \n",
    "        G = nx.DiGraph()\n",
    "        # Alle Knoten sammeln\n",
    "        all_nodes = set(children.keys())\n",
    "        for child_list in children.values():\n",
    "            all_nodes.update(child_list)\n",
    "        \n",
    "        # Sortierte Nodes hinzufügen\n",
    "        for node in sorted(all_nodes):\n",
    "            G.add_node(node, core_distance=0, allele_distance=0, time = tree.get_time(node))\n",
    "            \n",
    "        for parent, child_list in children.items():\n",
    "            for child in child_list:\n",
    "                G.add_edge(child, parent)\n",
    "\n",
    "        #### Add allele distances:\n",
    "        \n",
    "        # Calculate for every node the number of SNPs below it.\n",
    "        allele_distances = defaultdict(list)\n",
    "        gene_present_bool = gene_absence_presence_matrix[site.id] == 1\n",
    "        \n",
    "        for node, leaves in node_to_leaf.items():\n",
    "            if alleles_list[site.id].ndim == 1:\n",
    "               alleles_list[site.id] = np.zeros((1, num_samples), dtype=int)\n",
    "            leaves_arr = np.array(leaves)\n",
    "            valid_cols = leaves_arr[gene_present_bool[leaves_arr]]\n",
    "            if valid_cols.size == 0:\n",
    "                distance = 0\n",
    "            else:\n",
    "                subset = alleles_list[site.id][:, valid_cols]\n",
    "                distance = np.count_nonzero(np.ptp(subset, axis=1)) # ptp = max - min\n",
    "            allele_distances[tuple(leaves)] = distance\n",
    "\n",
    "        # Calculate for every node the number of SNPs below it which were not already SNPs in at least one of the children (new SNPs).\n",
    "        allele_distances_only_new = defaultdict(list)\n",
    "        gene_present_bool = gene_absence_presence_matrix[site.id] == 1\n",
    "        \n",
    "        for node, leaves in node_to_leaf.items():\n",
    "            if alleles_list[site.id].ndim == 1:\n",
    "                alleles_list[site.id] = np.zeros((1, num_samples), dtype=int)\n",
    "        \n",
    "            leaves_arr = np.array(leaves)\n",
    "            valid_cols = leaves_arr[gene_present_bool[leaves_arr]]\n",
    "        \n",
    "            if valid_cols.size == 0:\n",
    "                distance = 0\n",
    "            else:\n",
    "                subset = alleles_list[site.id][:, valid_cols]\n",
    "                \n",
    "                # Polymorphe Positionen im aktuellen Knoten\n",
    "                polymorph_mask = np.ptp(subset, axis=1) > 0   # shape = [num_sites]\n",
    "                \n",
    "                # Kinder-Masken sammeln\n",
    "                child_polymorph_mask_list = []\n",
    "                for child in G.predecessors(node): \n",
    "                    child_leaves = np.array(node_to_leaf[child])\n",
    "                    child_valid = child_leaves[gene_present_bool[child_leaves]]\n",
    "                    child_subset = alleles_list[site.id][:, child_valid]\n",
    "                \n",
    "                    if child_valid.size > 0:\n",
    "                        child_mask = np.ptp(child_subset, axis=1) > 0\n",
    "                    else:\n",
    "                        child_mask = np.zeros(subset.shape[0], dtype=bool)\n",
    "                \n",
    "                    child_polymorph_mask_list.append(child_mask)\n",
    "                \n",
    "                if child_polymorph_mask_list:\n",
    "                    # OR über alle Kinder -> Position polymorph in mind. einem Kind\n",
    "                    child_any_mask = np.logical_or.reduce(child_polymorph_mask_list)\n",
    "                else:\n",
    "                    # keine Kinder -> leerer OR = False\n",
    "                    child_any_mask = np.zeros_like(polymorph_mask)\n",
    "                \n",
    "                # Nur Positionen, die polymorph sind UND in allen Kindern noch monomorph\n",
    "                new_polymorph_mask = polymorph_mask & (~child_any_mask)\n",
    "                \n",
    "                distance = np.count_nonzero(new_polymorph_mask)\n",
    "                \n",
    "            allele_distances_only_new[tuple(leaves)] = distance\n",
    "            \n",
    "\n",
    "        for node in list(parents) + [clonal_root_node]:\n",
    "            G.nodes[node][\"allele_distance\"] = allele_distances[tuple(node_to_leaf[node])] / gene_length\n",
    "            G.nodes[node][\"allele_distance_only_new\"] = allele_distances_only_new[tuple(node_to_leaf[node])] / gene_length\n",
    "\n",
    "        for gen_tree in gene_trees_list[site.id].trees():\n",
    "            for node in list(parents) + [clonal_root_node]:\n",
    "                \n",
    "                # Calculate the expected number of alleles, which also can be seen as the true allele distance:\n",
    "                leaves = [leaf for leaf in node_to_leaf[node] if gene_absence_presence_matrix[site.id][leaf] == 1]\n",
    "                if len(leaves) <= 1:\n",
    "                    G.nodes[node][\"true_allele_distance\"] = 0\n",
    "                else:\n",
    "                    m = leaves[0]\n",
    "                    for u in leaves[1:]:\n",
    "                        m = gen_tree.mrca(m, u) # The most recent common ancestor of all leaves from the node\n",
    "                    \n",
    "                    visited_edges = set()\n",
    "                    total_length = 0.0 \n",
    "                    for leaf in leaves:\n",
    "                        u = leaf\n",
    "                        while u != tskit.NULL and u < m:\n",
    "                            p = gen_tree.parent(u)\n",
    "                            if p == tskit.NULL:\n",
    "                                break\n",
    "                            edge = (p, u)\n",
    "                            if edge not in visited_edges:\n",
    "                                visited_edges.add(edge)\n",
    "                                total_length += gen_tree.time(p) - gen_tree.time(u)\n",
    "                            # weiter hoch\n",
    "                            u = p\n",
    "          \n",
    "                    # The expected proportion of SNPS:\n",
    "                    G.nodes[node][\"true_allele_distance\"] = (1 - (1 - 1/gene_length)**round(total_length * nucleotide_mutation_rate * gene_length))\n",
    "\n",
    "        ### Add clonal distances:\n",
    "        subtree_has_gene = {}\n",
    "        for node in range(clonal_root_node):\n",
    "            if node < num_samples:\n",
    "                if gene_absence_presence_matrix[site.id][node] == 1:\n",
    "                    subtree_has_gene[node] = True\n",
    "                else:\n",
    "                    subtree_has_gene[node] = False\n",
    "            else:\n",
    "                child_list = children[node]\n",
    "                if any(subtree_has_gene[child] for child in child_list):\n",
    "                    subtree_has_gene[node] = True\n",
    "                else:\n",
    "                    subtree_has_gene[node] = False\n",
    "\n",
    "        for node in range(num_samples, clonal_root_node + 1):\n",
    "            if node in G.nodes:\n",
    "                c0, c1 = children[node]  # Entpacke einmal\n",
    "                c0_has, c1_has = subtree_has_gene[c0], subtree_has_gene[c1]\n",
    "            \n",
    "                # Falls beide Subtrees kein Gen enthalten, können wir überspringen\n",
    "                if not (c0_has or c1_has):\n",
    "                    continue\n",
    "            \n",
    "                if c0_has and c1_has:\n",
    "                    node_time = tree.get_time(node)\n",
    "                    core_distance = (\n",
    "                        2 * node_time\n",
    "                        - tree.get_time(c0) - tree.get_time(c1)\n",
    "                        + G.nodes[c0][\"core_distance\"]\n",
    "                        + G.nodes[c1][\"core_distance\"]\n",
    "                    )\n",
    "                if c0_has and c1_has:\n",
    "                    # Finde tiefste Kinderknoten, bei denen beide Subtrees True sind\n",
    "                    if c0 >= num_samples:\n",
    "                        d0, d1 = children[c0]\n",
    "                        while (not subtree_has_gene[d0] and subtree_has_gene[d1]) or (subtree_has_gene[d0] and not subtree_has_gene[d1]):\n",
    "                            if subtree_has_gene[d0]:\n",
    "                                c0 = d0\n",
    "                                if d0 < num_samples:\n",
    "                                    break\n",
    "                            elif subtree_has_gene[d1]:\n",
    "                                c0 = d1\n",
    "                                if d1 < num_samples:\n",
    "                                    break\n",
    "                            d0, d1 = children[c0]\n",
    "                                    \n",
    "                    if c1 >= num_samples:\n",
    "                        d0, d1 = children[c1]\n",
    "                        while (not subtree_has_gene[d0] and subtree_has_gene[d1]) or (subtree_has_gene[d0] and not subtree_has_gene[d1]):\n",
    "                            if subtree_has_gene[d0]:\n",
    "                                c1 = d0\n",
    "                                if d0 < num_samples:\n",
    "                                    break\n",
    "                            elif subtree_has_gene[d1]:\n",
    "                                c1 = d1\n",
    "                                if d1 < num_samples:\n",
    "                                    break\n",
    "                            d0, d1 = children[c1]\n",
    "         \n",
    "                    node_time = tree.get_time(node)\n",
    "                    core_distance = (\n",
    "                        2 * node_time\n",
    "                        - tree.get_time(c0) - tree.get_time(c1)\n",
    "                        + G.nodes[c0][\"core_distance\"]\n",
    "                        + G.nodes[c1][\"core_distance\"]\n",
    "                    )\n",
    "                elif c0_has:\n",
    "                    core_distance = G.nodes[c0][\"core_distance\"]\n",
    "                else:  # c1_has\n",
    "                    core_distance = G.nodes[c1][\"core_distance\"]\n",
    "\n",
    "                G.nodes[node][\"core_distance\"] = core_distance\n",
    "\n",
    "       \n",
    "        # Set the numbers of the nodes in the graph to 0 to num_samples-1 :\n",
    "        G_nodes_reordering = {old_label: new_label for new_label, old_label in enumerate(G.nodes())}\n",
    "        nx.relabel_nodes(G, G_nodes_reordering, copy=False)\n",
    "        \n",
    "        Graphs.append(G)\n",
    "\n",
    "        # Process the graph to save it more easily:\n",
    "        graph_nodes = list(G.nodes)\n",
    "        graph_edges = list(G.edges)\n",
    "        graph_edge_index = torch.tensor(np.array(graph_edges).T)\n",
    "\n",
    "        node_features = []\n",
    "        for node in graph_nodes:\n",
    "            core = G.nodes[node].get(\"core_distance\", 0.0)\n",
    "            allele = G.nodes[node].get(\"allele_distance\", 0.0)\n",
    "            allele_only_new = G.nodes[node].get(\"allele_distance_only_new\", 0.0)\n",
    "            node_time = G.nodes[node].get(\"time\", 0.0)\n",
    "            node_features.append([core, allele, allele_only_new, node_time])\n",
    "        node_features = torch.tensor(node_features, dtype=torch.float32)\n",
    "\n",
    "        graph_properties.append([graph_nodes, graph_edge_index, node_features])\n",
    "\n",
    "        nodes_in_simplified_tree = list(parents) + [clonal_root_node]\n",
    "        for node, leaf in nodes_hgt_events[site.id]:\n",
    "            while node not in nodes_in_simplified_tree:\n",
    "                node = tree.parent(node)\n",
    "            parental_nodes_hgt_events[site.id].append(node)\n",
    "        for node in gene_nodes_loss_events[site.id]:\n",
    "            stack = [node]\n",
    "            while stack:\n",
    "                node = stack.pop()\n",
    "                if node in nodes_in_simplified_tree:\n",
    "                    children_gene_nodes_loss_events[site.id].append(node)\n",
    "                    break\n",
    "                else:\n",
    "                    if node >= num_samples:\n",
    "                        child_list = tree.children(node)\n",
    "                        stack.extend(child_list)\n",
    "                    #if node in hgt_parent_nodes:\n",
    "                    #    stack.append(node-1)\n",
    "\n",
    "        # Move nodes with HGT events to a higher node, if there are not enough genes present under it.\n",
    "        for hgt_node in parental_nodes_hgt_events[site.id]:\n",
    "            node = hgt_node\n",
    "            c0, c1 = children[node]\n",
    "            while not (subtree_has_gene[c0] and subtree_has_gene[c1]) and node < clonal_root_node:\n",
    "                node = parents[node]\n",
    "                c0, c1 = children[node]\n",
    "            if node < clonal_root_node or hgt_node == clonal_root_node:\n",
    "                parental_nodes_hgt_events_corrected[site.id].append(node)\n",
    "        \n",
    "        parental_nodes_hgt_events_corrected[site.id] = [G_nodes_reordering[node] for node in parental_nodes_hgt_events_corrected[site.id]]\n",
    "        parental_nodes_hgt_events[site.id] = [G_nodes_reordering[node] for node in parental_nodes_hgt_events[site.id]]\n",
    "        children_gene_nodes_loss_events[site.id] = [G_nodes_reordering[node] for node in children_gene_nodes_loss_events[site.id]]\n",
    "   \n",
    "### Print the computation time.\n",
    "\n",
    "end_time = time.time()\n",
    "elapsed_time = end_time - start_time\n",
    "\n",
    "print(f\"Success: hgt_rate = {hgt_rate}, Total computation time = {elapsed_time:.6f} seconds.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "9d7e67a1-40c9-456c-95a9-890f8665fe7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "1 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "2 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "3 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "4 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "5 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "6 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "7 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "8 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "9 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "10 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "11 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "12 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "13 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "14 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "15 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "16 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "17 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "18 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "19 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "20 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "21 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "22 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "23 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "24 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "25 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "26 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "27 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "28 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "29 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "30 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "31 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "32 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "33 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "34 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "35 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "36 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "37 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "38 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "39 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "40 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "41 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "42 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "43 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "44 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "45 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "46 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "47 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "48 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "49 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "50 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "51 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "52 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "53 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "54 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "55 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "56 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "57 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "58 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "59 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "60 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "61 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "62 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "63 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "64 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "65 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "66 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "67 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "68 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "69 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "70 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "71 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "72 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "73 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "74 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "75 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "76 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "77 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "78 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "79 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "80 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "81 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "82 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "83 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "84 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "85 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "86 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "87 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "88 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "89 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "90 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "91 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "92 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "93 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "94 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "95 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "96 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "97 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "98 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "99 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "100 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0001, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "101 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0002, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "102 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0002, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "103 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0002, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "104 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0003, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "105 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0004, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "106 {'core_distance': 0.0009, 'allele_distance': 0.0, 'time': 0.0005, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0.0}\n",
      "107 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0006, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "110 {'core_distance': 0.0016, 'allele_distance': 0.002, 'time': 0.0008, 'allele_distance_only_new': 0.002, 'true_allele_distance': 0.001}\n",
      "109 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0008, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "108 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0007, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "113 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0025, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "112 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0022, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "111 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0014, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "116 {'core_distance': 0.0064, 'allele_distance': 0.003, 'time': 0.0032, 'allele_distance_only_new': 0.003, 'true_allele_distance': 0.003}\n",
      "115 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0029, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "114 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0025, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "119 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0038, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "118 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0034, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "117 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0033, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "122 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0048, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "121 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0044, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "120 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0038, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "126 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0064, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "125 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.006, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "124 {'core_distance': 0.0117, 'allele_distance': 0.003, 'time': 0.0056, 'allele_distance_only_new': 0.003, 'true_allele_distance': 0.006}\n",
      "123 {'core_distance': 0.0144, 'allele_distance': 0.006, 'time': 0.0056, 'allele_distance_only_new': 0.003, 'true_allele_distance': 0.007}\n",
      "132 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0079, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "131 {'core_distance': 0.0224, 'allele_distance': 0.011, 'time': 0.0077, 'allele_distance_only_new': 0.003, 'true_allele_distance': 0.0109}\n",
      "130 {'core_distance': 0.0232, 'allele_distance': 0.011, 'time': 0.0072, 'allele_distance_only_new': 0.005, 'true_allele_distance': 0.0109}\n",
      "129 {'core_distance': 0.014, 'allele_distance': 0.008, 'time': 0.007, 'allele_distance_only_new': 0.008, 'true_allele_distance': 0.007}\n",
      "128 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0067, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "127 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0065, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "138 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0097, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "137 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0095, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "136 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0094, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "135 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0087, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "134 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0085, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "133 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0082, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "144 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0127, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "143 {'core_distance': 0.0394, 'allele_distance': 0.02, 'time': 0.0123, 'allele_distance_only_new': 0.009, 'true_allele_distance': 0.0188}\n",
      "142 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0123, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "141 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0109, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "140 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0104, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "139 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0098, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "150 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0175, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "149 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0173, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "148 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0173, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "147 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0155, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "146 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.013, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "145 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0129, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "156 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0197, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "155 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0194, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "154 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0194, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "153 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0193, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "152 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0187, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "151 {'core_distance': 0.0354, 'allele_distance': 0.015, 'time': 0.0177, 'allele_distance_only_new': 0.015, 'true_allele_distance': 0.0169}\n",
      "162 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.024, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "161 {'core_distance': 0.0484, 'allele_distance': 0.034, 'time': 0.0238, 'allele_distance_only_new': 0.032, 'true_allele_distance': 0.0237}\n",
      "160 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.023, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "159 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0228, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "158 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0201, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "157 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0199, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "173 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0385, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "172 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0383, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "171 {'core_distance': 0.0943, 'allele_distance': 0.041, 'time': 0.0383, 'allele_distance_only_new': 0.026, 'true_allele_distance': 0.045}\n",
      "170 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0334, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "169 {'core_distance': 0.1085, 'allele_distance': 0.062, 'time': 0.0327, 'allele_distance_only_new': 0.031, 'true_allele_distance': 0.0516}\n",
      "168 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0321, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "167 {'core_distance': 0.0633, 'allele_distance': 0.021, 'time': 0.0317, 'allele_distance_only_new': 0.021, 'true_allele_distance': 0.0305}\n",
      "166 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0306, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "165 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0304, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "164 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0294, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "163 {'core_distance': 0.055, 'allele_distance': 0.027, 'time': 0.0275, 'allele_distance_only_new': 0.027, 'true_allele_distance': 0.0267}\n",
      "185 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0823, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "184 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.08, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "183 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0597, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "182 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0588, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "181 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0538, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "180 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0524, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "179 {'core_distance': 0.2117, 'allele_distance': 0.095, 'time': 0.0512, 'allele_distance_only_new': 0.014, 'true_allele_distance': 0.0988}\n",
      "178 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0506, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "177 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0487, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "176 {'core_distance': 0.2158, 'allele_distance': 0.105, 'time': 0.0475, 'allele_distance_only_new': 0.011, 'true_allele_distance': 0.1006}\n",
      "175 {'core_distance': 0.1214, 'allele_distance': 0.06, 'time': 0.0439, 'allele_distance_only_new': 0.03, 'true_allele_distance': 0.0573}\n",
      "174 {'core_distance': 0.1363, 'allele_distance': 0.06, 'time': 0.0401, 'allele_distance_only_new': 0.019, 'true_allele_distance': 0.0648}\n",
      "198 {'core_distance': 1.075, 'allele_distance': 0.392, 'time': 1.4866, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0.4092}\n",
      "197 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.5509, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "196 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.4181, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "195 {'core_distance': 1.075, 'allele_distance': 0.392, 'time': 0.4064, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0.4092}\n",
      "194 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.2808, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "193 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.2753, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "192 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.2283, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "191 {'core_distance': 1.075, 'allele_distance': 0.392, 'time': 0.2065, 'allele_distance_only_new': 0.049, 'true_allele_distance': 0.4092}\n",
      "190 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.1937, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "189 {'core_distance': 0.6676, 'allele_distance': 0.273, 'time': 0.1659, 'allele_distance_only_new': 0.081, 'true_allele_distance': 0.279}\n",
      "188 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.1193, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n",
      "187 {'core_distance': 0.4275, 'allele_distance': 0.192, 'time': 0.0918, 'allele_distance_only_new': 0.034, 'true_allele_distance': 0.1887}\n",
      "186 {'core_distance': 0, 'allele_distance': 0.0, 'time': 0.0886, 'allele_distance_only_new': 0.0, 'true_allele_distance': 0}\n"
     ]
    }
   ],
   "source": [
    "for node, data in G.nodes(data=True):\n",
    "    rounded_data = {\n",
    "        key: round(value, 4) if isinstance(value, (int, float)) else value\n",
    "        for key, value in data.items()\n",
    "    }\n",
    "    print(node, rounded_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b42cfd6d-3cd8-4dab-b271-c3194d137860",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
