{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "6d159872-ad3c-4750-98c0-95c853373c41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clonal node: 6147\n",
      "Present_mutation at node:  9282\n",
      "Number of present genes:  843\n",
      "gene_number_hgt_events_passed:  [914]\n"
     ]
    }
   ],
   "source": [
    "import msprime\n",
    "import tskit\n",
    "import hgt_simulation\n",
    "import hgt_sim_args\n",
    "import numpy as np\n",
    "import random\n",
    "import torch\n",
    "import os\n",
    "import time\n",
    "import re\n",
    "import secrets\n",
    "import copy\n",
    "import json\n",
    "import h5py\n",
    "import uuid\n",
    "import glob\n",
    "import shutil\n",
    "\n",
    "from random import randint\n",
    "from collections import namedtuple\n",
    "from collections import defaultdict\n",
    "from collections import deque\n",
    "from typing import List, Union\n",
    "from sbi.utils import BoxUniform\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "from concurrent.futures import ProcessPoolExecutor\n",
    "\n",
    "\n",
    "num_samples = 1000\n",
    "theta = 1\n",
    "rho = 1\n",
    "hgt_rate = 20\n",
    "num_genes = 1\n",
    "nucleotide_mutation_rate = 0.1\n",
    "gene_length = 1000\n",
    "pca_dimensions = 10\n",
    "ce_from_nwk = None\n",
    "seed = None\n",
    "distance_matrix = None\n",
    "multidimensional_scaling_dimensions = 100\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "    \n",
    "if ce_from_nwk is None and num_samples is None:\n",
    "    raise ValueError(\n",
    "        \"Neither a core tree or parameters for simulation were provided. Choose either.\"\n",
    "    )\n",
    "\n",
    "if ce_from_nwk is None:\n",
    "    core_tree = msprime.sim_ancestry(\n",
    "            samples=num_samples,\n",
    "            sequence_length=1,\n",
    "            ploidy=1,\n",
    "            recombination_rate=0,\n",
    "            gene_conversion_rate=0,\n",
    "            gene_conversion_tract_length=1,  # One gene\n",
    "            random_seed=seed,\n",
    "        )\n",
    "\n",
    "    ce_from_nwk = core_tree.first().newick()\n",
    "\n",
    "if seed is None:\n",
    "    seed = secrets.randbelow(2**32 - 4) + 2\n",
    "\n",
    "if num_samples < multidimensional_scaling_dimensions:\n",
    "    multidimensional_scaling_dimensions = num_samples\n",
    "\n",
    "random.seed(seed)\n",
    "np.random.seed(seed)\n",
    "#print(\"Seed: \", seed)\n",
    "\n",
    "### Calculate hgt events:\n",
    "\n",
    "args = hgt_sim_args.Args(\n",
    "    sample_size=num_samples,\n",
    "    num_sites=num_genes,\n",
    "    gene_conversion_rate=0,\n",
    "    recombination_rate=0,\n",
    "    hgt_rate=hgt_rate,\n",
    "    ce_from_ts=None,\n",
    "    ce_from_nwk=ce_from_nwk,\n",
    "    random_seed=seed,\n",
    "    #random_seed=84,\n",
    ")\n",
    "\n",
    "ts, hgt_edges = hgt_simulation.run_simulate(args)\n",
    "        \n",
    "### Place mutations\n",
    "\n",
    "alleles = [\"absent\", \"present\"]\n",
    "\n",
    "# Place one mutation per site, e.g. genome position\n",
    "\n",
    "gains_model = msprime.MatrixMutationModel(\n",
    "    alleles = alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [0, 1],\n",
    "        [0, 1],\n",
    "    ],\n",
    ")\n",
    "\n",
    "ts_gains = msprime.sim_mutations(ts, rate=1, model = gains_model, keep = True, random_seed=seed)\n",
    "\n",
    "k = 1\n",
    "while (ts_gains.num_sites < ts_gains.sequence_length):\n",
    "    ts_gains = msprime.sim_mutations(ts_gains, rate=1, model = gains_model, keep = True, random_seed=seed+k)\n",
    "    k = k+1\n",
    "\n",
    "# Remove superfluous mutations\n",
    "\n",
    "tables = ts_gains.dump_tables()\n",
    "\n",
    "mutations_by_site = {}\n",
    "for mut in tables.mutations:\n",
    "    if mut.site not in mutations_by_site:\n",
    "        mutations_by_site[mut.site] = []\n",
    "    mutations_by_site[mut.site].append(mut)\n",
    "\n",
    "tables.mutations.clear()\n",
    "\n",
    "for site, mutations in mutations_by_site.items():\n",
    "    selected_mutation = random.choice(mutations)\n",
    "\n",
    "    tables.mutations.add_row(\n",
    "        site=selected_mutation.site,\n",
    "        node=selected_mutation.node,\n",
    "        derived_state=selected_mutation.derived_state,\n",
    "        parent=-1,\n",
    "        metadata=None,\n",
    "        time=selected_mutation.time,\n",
    "    )\n",
    "\n",
    "    # Add sentinel mutations at the leafs:\n",
    "\n",
    "    for leaf_position in range(num_samples):\n",
    "        tables.mutations.add_row(\n",
    "            site = selected_mutation.site,\n",
    "            node = leaf_position,\n",
    "            derived_state = \"absent\",\n",
    "            time = 0.00000000001,\n",
    "        )\n",
    "\n",
    "ts_gains = tables.tree_sequence()\n",
    "\n",
    "\n",
    "# Place losses:\n",
    "\n",
    "losses_model = msprime.MatrixMutationModel(\n",
    "    alleles = alleles,\n",
    "    root_distribution=[1, 0],\n",
    "    transition_matrix=[\n",
    "        [1, 0],\n",
    "        [1, 0],\n",
    "    ],\n",
    ")\n",
    "\n",
    "ts_gains_losses = msprime.sim_mutations(ts_gains, rate = rho, model = losses_model, keep = True, random_seed=seed-1)\n",
    "\n",
    "tables = ts_gains_losses.dump_tables()\n",
    "\n",
    "# Find the node of the root of the clonal tree:\n",
    "clonal_node = ts_gains.first().mrca(*list(range(num_samples)))\n",
    "\n",
    "print(\"Clonal node:\", clonal_node)\n",
    "\n",
    "for site, mutations in mutations_by_site.items():\n",
    "    # Mutation at clonal root:\n",
    "    tables.mutations.add_row(\n",
    "        site=site,\n",
    "        node=clonal_node,\n",
    "        derived_state=\"absent\",\n",
    "        parent=-1,\n",
    "        metadata=None,\n",
    "        time=tables.nodes.time[clonal_node],\n",
    "    )\n",
    "\n",
    "tables.sort()\n",
    "ts_gains_losses = tables.tree_sequence()\n",
    "\n",
    "### Calculate the gene absence presence matrix:\n",
    "\n",
    "MutationRecord = namedtuple('MutationRecord', ['site_id', 'mutation_id', 'node', 'is_hgt'])\n",
    "\n",
    "tables = ts_gains_losses.dump_tables()\n",
    "\n",
    "tables.mutations.clear() # SIMPLE VERSION!\n",
    "\n",
    "hgt_parent_nodes = [edge.parent-1 for edge in hgt_edges]\n",
    "hgt_children_nodes = [edge.child for edge in hgt_edges]\n",
    "hgt_parent_children = defaultdict(list)\n",
    "#hgt_children_parent = defaultdict(list)\n",
    "\n",
    "for parent in hgt_parent_nodes:\n",
    "    hgt_parent_children[parent].append(parent-1)\n",
    "\n",
    "\n",
    "#for child in hgt_children_nodes:\n",
    "#    hgt_children_parent[child].append(child+1)\n",
    "\n",
    "# Initialize tables for the diversity of each gene\n",
    "tables_gene = tskit.TableCollection(sequence_length=gene_length)\n",
    "tables_gene.nodes.replace_with(ts_gains_losses.tables.nodes)\n",
    "tables_gene.populations.replace_with(ts_gains_losses.tables.populations)\n",
    "\n",
    "\n",
    "tables_gene_list = [copy.deepcopy(tables_gene) for _ in range(num_genes)]\n",
    "gene_number_hgt_events_passed = [0 for _ in range(num_genes)]\n",
    "gene_trees_list = []\n",
    "\n",
    "\n",
    "for tree in ts_gains_losses.trees():\n",
    "\n",
    "    for site in tree.sites():\n",
    "        \n",
    "        hgt_parent_children_passed = [False] * ts_gains_losses.num_nodes\n",
    "        mutations = site.mutations\n",
    "        #site = list(tree.sites())[0]\n",
    "        \n",
    "        present_mutation = [m for m in mutations if m.derived_state == \"present\"][0]\n",
    "        absent_mutation_nodes = {m.node for m in mutations if m.derived_state == \"absent\"}\n",
    "\n",
    "        absent_mutations = defaultdict(list)\n",
    "        for m in mutations:\n",
    "            if m.derived_state == \"absent\" and m.time < present_mutation.time:\n",
    "                absent_mutations[m.node].append(m)\n",
    "\n",
    "        branching_nodes_reached_before = defaultdict(list)\n",
    "        for node_id in range(ts_gains_losses.num_nodes):\n",
    "            branching_nodes_reached_before[node_id] = False\n",
    "    \n",
    "        print(\"Present_mutation at node: \", present_mutation.node)\n",
    "    \n",
    "        branching_nodes_to_process = deque([(present_mutation.node, False, False, 0)])\n",
    "        # The second variable describes if a hgt edge was passed the whole way down to the actual node. \n",
    "        # The third describes if a hgt edge was passed in the last step.\n",
    "        # The fourth is the number of hgt edges that were passed.\n",
    "    \n",
    "        child_mutations = []\n",
    "    \n",
    "        if present_mutation.node < num_samples: # Gain directly above leaf:\n",
    "            if present_mutation.id == sorted([mut for mut in mutations if mut.node == present_mutation.node], key=lambda m: m.time)[1].id:\n",
    "                sentinel_mutation = min([mut for mut in absent_mutations[present_mutation.node]], key=lambda m: m.time)\n",
    "                child_mutations.append(MutationRecord(\n",
    "                    site_id=site.id,\n",
    "                    mutation_id=sentinel_mutation.id,\n",
    "                    node=sentinel_mutation.node,\n",
    "                    is_hgt=False\n",
    "                ))\n",
    "                tables_gene_list[site.id].edges.add_row(\n",
    "                        left = 0, right = gene_length, parent = tree.parent(present_mutation.node), child = sentinel_mutation.node\n",
    "                )\n",
    "    \n",
    "        else: \n",
    "            \n",
    "            # To see, which HGT edges are passed, we have to go through the tree two times. \n",
    "            # First, we detect all passed HGT edges, then we calculate the presence of mutations in the leaves.\n",
    "    \n",
    "            # First time going through:\n",
    "            while branching_nodes_to_process:\n",
    "                \n",
    "                last_branching_node = branching_nodes_to_process.popleft()\n",
    "                selected_branch_nodes_to_process = deque([last_branching_node])\n",
    "                branching_nodes_reached_before[tree.parent(last_branching_node[0])] = True\n",
    "                \n",
    "                while selected_branch_nodes_to_process:\n",
    "                \n",
    "                    child_node = selected_branch_nodes_to_process.popleft()    \n",
    "        \n",
    "                    # If there is a mutation on the edge, find the earliest one.\n",
    "                    if not child_node[2] and child_node[0] in absent_mutation_nodes:\n",
    "                        \n",
    "                        absent_mutation_after_gain_at_node = absent_mutations[child_node[0]]\n",
    "                        \n",
    "                        if not absent_mutation_after_gain_at_node: # empty\n",
    "                            children = tree.children(child_node[0])\n",
    "                            if len(children) > 1:\n",
    "                                for child in reversed(children):\n",
    "                                    if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                        branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                                    #else:\n",
    "                                    #    print(\"Child passed before\", \" Child: \", child)\n",
    "                            else:\n",
    "                                for child in reversed(children):\n",
    "                                    selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                            if hgt_parent_children[child_node[0]]:\n",
    "                                selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])\n",
    "                                hgt_parent_children_passed[hgt_parent_children[child_node[0]][0]] = True # The child of the hgt_edge is marked\n",
    "                                #print(\"HGT at Node: \", hgt_parent_children[child_node[0]][0])\n",
    "                            \n",
    "            \n",
    "                    # If there is no mutation, add child nodes.\n",
    "                    else:\n",
    "                        children = tree.children(child_node[0])\n",
    "                        if len(children) > 1:\n",
    "                            for child in reversed(children):\n",
    "                                #print(\"Child: \", child)\n",
    "                                if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                    branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                                #else:\n",
    "                                #    print(\"Child passed before\", \" Child: \", child)\n",
    "                        else:\n",
    "                            for child in reversed(children):\n",
    "                                #print(\"Child: \", child)\n",
    "                                selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                        if hgt_parent_children[child_node[0]]:\n",
    "                            selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])\n",
    "                            hgt_parent_children_passed[hgt_parent_children[child_node[0]][0]] = True # The child of the hgt_edge is marked\n",
    "                            #print(\"HGT at Node: \", hgt_parent_children[child_node[0]][0])\n",
    "    \n",
    "            # Second time going through:\n",
    "            \n",
    "            branching_nodes_to_process = deque([(present_mutation.node, False, False, 0)])\n",
    "\n",
    "            branching_nodes_reached_before = defaultdict(list)\n",
    "            for node_id in range(ts_gains_losses.num_nodes):\n",
    "                branching_nodes_reached_before[node_id] = False\n",
    "            \n",
    "            while branching_nodes_to_process:\n",
    "                \n",
    "                last_branching_node = branching_nodes_to_process.popleft()\n",
    "                selected_branch_nodes_to_process = deque([last_branching_node])\n",
    "                branching_nodes_reached_before[tree.parent(last_branching_node[0])] = True\n",
    "                \n",
    "                while selected_branch_nodes_to_process:\n",
    "                \n",
    "                    child_node = selected_branch_nodes_to_process.popleft()\n",
    "    \n",
    "                    if not child_node[2] and hgt_parent_children_passed[child_node[0]]:\n",
    "                        #print(\"Incoming HGT edge registered at node: \", child_node[0])\n",
    "                        continue\n",
    "        \n",
    "                    # If there is a mutation on the edge, find the earliest one.\n",
    "                    if not child_node[2] and child_node[0] in absent_mutation_nodes:\n",
    "                        \n",
    "                        absent_mutation_after_gain_at_node = absent_mutations[child_node[0]]\n",
    "                        \n",
    "                        if absent_mutation_after_gain_at_node: # not empty\n",
    "                            earliest_mutation = max(\n",
    "                                absent_mutations[child_node[0]], \n",
    "                                key=lambda m: m.time\n",
    "                            )\n",
    "            \n",
    "                            if earliest_mutation.time == 0.00000000001:\n",
    "                                earliest_mutation.derived_state = \"present\"\n",
    "                                child_mutations.append(MutationRecord(\n",
    "                                    site_id=site.id,\n",
    "                                    mutation_id=earliest_mutation.id,\n",
    "                                    node=earliest_mutation.node,\n",
    "                                    is_hgt=child_node[1]\n",
    "                                ))\n",
    "                                tables_gene_list[site.id].edges.add_row(\n",
    "                                    left = 0, right = gene_length, parent = tree.parent(last_branching_node[0]), child = earliest_mutation.node\n",
    "                                )\n",
    "                                gene_number_hgt_events_passed[site.id] += child_node[3]\n",
    "                        else:\n",
    "                            children = tree.children(child_node[0])\n",
    "                            if len(children) > 1:\n",
    "                                for child in reversed(children):\n",
    "                                    if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                        branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                                tables_gene_list[site.id].edges.add_row(\n",
    "                                        left = 0, right = gene_length, parent = tree.parent(last_branching_node[0]), child = child_node[0]\n",
    "                                )\n",
    "                                gene_number_hgt_events_passed[site.id] += child_node[3]\n",
    "                            else:\n",
    "                                for child in reversed(children):\n",
    "                                    selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                            if hgt_parent_children[child_node[0]]:\n",
    "                                selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])  \n",
    "                            \n",
    "            \n",
    "                    # If there is no mutation, add child nodes.\n",
    "                    else:\n",
    "                        children = tree.children(child_node[0])\n",
    "                        if len(children) > 1:\n",
    "                            for child in reversed(children):\n",
    "                                if not branching_nodes_reached_before[child_node[0]]:\n",
    "                                    branching_nodes_to_process.extendleft([(child, child_node[1], False, 0)])\n",
    "                            tables_gene_list[site.id].edges.add_row(\n",
    "                                    left = 0, right = gene_length, parent = tree.parent(last_branching_node[0]), child = child_node[0]\n",
    "                            )\n",
    "                            gene_number_hgt_events_passed[site.id] += child_node[3]\n",
    "                        else:\n",
    "                            for child in reversed(children):\n",
    "                                selected_branch_nodes_to_process.extendleft([(child, child_node[1], False, child_node[3])])\n",
    "                        if hgt_parent_children[child_node[0]]:\n",
    "                            selected_branch_nodes_to_process.extendleft([(hgt_parent_children[child_node[0]][0], True, True, child_node[3] + 1)])\n",
    "\n",
    "\n",
    "        child_mutations.sort(key=lambda mut: not mut.is_hgt) # Will set is_hgt to False later if there are paths without hgt events to the leaf.\n",
    "    \n",
    "        # We have to adress multiple paths to the same destiny, some with hgt and other without it:\n",
    "        unique_mutations = {}\n",
    "    \n",
    "        for mut in child_mutations:\n",
    "            if mut.node not in unique_mutations:\n",
    "                unique_mutations[mut.node] = mut\n",
    "            else:\n",
    "                existing_mut = unique_mutations[mut.node]\n",
    "                if not existing_mut.is_hgt or not mut.is_hgt:\n",
    "                    unique_mutations[mut.node] = mut._replace(is_hgt=False)\n",
    "        \n",
    "        child_mutations_filtered = list(unique_mutations.values())\n",
    "\n",
    "        for mutation in mutations:\n",
    "            if mutation.time > 0.00000000001:\n",
    "                if mutation.derived_state == \"absent\":\n",
    "                    metadata_value = bytes([3]) \n",
    "                elif mutation.derived_state == \"present\":\n",
    "                    metadata_value = bytes([7])\n",
    "                tables.mutations.add_row(\n",
    "                    site=site.id,\n",
    "                    node=mutation.node,\n",
    "                    derived_state=mutation.derived_state,\n",
    "                    parent=-1,\n",
    "                    metadata=metadata_value,\n",
    "                    time=mutation.time,\n",
    "                )\n",
    "        \n",
    "        for mutation in child_mutations_filtered:\n",
    "    \n",
    "            tables.mutations.add_row(\n",
    "                site=site.id,\n",
    "                node=mutation.node,\n",
    "                derived_state=\"present\",\n",
    "                parent=-1,\n",
    "                metadata=bytes([mutation.is_hgt]),\n",
    "                time=0.00000000001,\n",
    "            )\n",
    "    \n",
    "\n",
    "mts = tables.tree_sequence()\n",
    "\n",
    "# Simulate the tree for each gene:\n",
    "\n",
    "\n",
    "nucleotide_mutation = msprime.MatrixMutationModel(\n",
    "    alleles=[\"C\", \"M\"],\n",
    "    root_distribution=[1.0, 0.0],  # nur C als Wurzel\n",
    "    transition_matrix=[\n",
    "        # C     M1     \n",
    "        [ 0.0,  1 ], \n",
    "        [ 1/3,  2/3],  \n",
    "    ]\n",
    ")\n",
    "\n",
    "for i in range(num_genes): \n",
    "            \n",
    "    tables_gene_list[i].sort()\n",
    "    \n",
    "    gene_trees_list.append(tables_gene_list[i].tree_sequence())\n",
    "    \n",
    "    gene_trees_list[i] = msprime.sim_mutations(gene_trees_list[i], \n",
    "                                               rate = nucleotide_mutation_rate, model = nucleotide_mutation, keep = True, random_seed=seed-1)\n",
    "\n",
    "# Calculate the different alleles:\n",
    "\n",
    "alleles_list = []\n",
    "\n",
    "for i in range(num_genes):\n",
    "    alleles_list.append([])\n",
    "    for var in gene_trees_list[i].variants():\n",
    "        alleles_list[i].append(var.genotypes)\n",
    "    alleles_list[i] = np.array(alleles_list[i])\n",
    "\n",
    "\n",
    "### Compute the gene presence and absence:\n",
    "\n",
    "gene_absence_presence_matrix = []\n",
    "\n",
    "for var in mts.variants():\n",
    "    gene_absence_presence_matrix.append(var.genotypes)\n",
    "gene_absence_presence_matrix = np.array(gene_absence_presence_matrix)\n",
    "\n",
    "print(\"Number of present genes: \", sum(gene_absence_presence_matrix[0]))\n",
    "print(\"gene_number_hgt_events_passed: \", gene_number_hgt_events_passed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "4f3a4aaa-baea-4b66-87af-9fd071e0db44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.019999027252197266\n",
      "Number of genes in leaves 843\n",
      "Number of HGT: 78\n"
     ]
    }
   ],
   "source": [
    "clonal_root_node = clonal_node\n",
    "gene_number_hgt_events_passed = [0 for _ in range(num_genes)]\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "for tree in mts.trees():\n",
    "    for site in tree.sites():\n",
    "        \n",
    "        clonal_nodes = defaultdict(list)\n",
    "        for node_id in range(mts.num_nodes):\n",
    "            clonal_nodes[node_id] = False\n",
    "        reached_nodes_from_leaves = copy.deepcopy(clonal_nodes)\n",
    "            \n",
    "        stack = [clonal_root_node]\n",
    "        clonal_nodes[clonal_root_node] = True\n",
    "        while stack:\n",
    "            node = stack.pop()\n",
    "            children = tree.children(node)\n",
    "            clonal_nodes[node] = True\n",
    "            stack.extend(children)\n",
    "\n",
    "        stack = []\n",
    "        for node_id in range(mts.num_samples):\n",
    "            if gene_absence_presence_matrix[site.id][node_id]:\n",
    "                #reached_nodes_from_leaves[node_id] = True\n",
    "                stack.append(node_id)\n",
    "                \n",
    "        while stack:\n",
    "            node = stack.pop()\n",
    "            #print(node)\n",
    "            parent = tree.parent(node)\n",
    "            if not hgt_parent_children_passed[node] and not reached_nodes_from_leaves[parent] and node < clonal_root_node:\n",
    "                stack.append(parent)\n",
    "                reached_nodes_from_leaves[node] = True\n",
    "            elif hgt_parent_children_passed[node] and not reached_nodes_from_leaves[parent] and node < clonal_root_node:\n",
    "                #print(\"HGT event spotted at node: \", node)\n",
    "                gene_number_hgt_events_passed[site.id] += 1\n",
    "                reached_nodes_from_leaves[node] = True\n",
    "        \n",
    "\n",
    "end_time = time.time()\n",
    "print(end_time-start_time)\n",
    "\n",
    "print(\"Number of genes in leaves\", sum(gene_absence_presence_matrix[0]))\n",
    "print(\"Number of HGT:\" , gene_number_hgt_events_passed[0])\n",
    "\n",
    "#mts.draw_svg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "fdd05f77-2084-4ff3-8613-b5a78ee28f77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False,\n",
       " False]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgt_parent_children_passed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2d2c4b3c-66f1-4f9c-ac4f-c8f2b762b597",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {0: True,\n",
       "             1: True,\n",
       "             2: True,\n",
       "             3: True,\n",
       "             4: True,\n",
       "             5: True,\n",
       "             6: False,\n",
       "             7: False,\n",
       "             8: True,\n",
       "             9: True,\n",
       "             10: True,\n",
       "             11: True,\n",
       "             12: False,\n",
       "             13: False,\n",
       "             14: False,\n",
       "             15: False,\n",
       "             16: False,\n",
       "             17: False,\n",
       "             18: False,\n",
       "             19: False,\n",
       "             20: False,\n",
       "             21: False,\n",
       "             22: False,\n",
       "             23: False,\n",
       "             24: False,\n",
       "             25: True,\n",
       "             26: False,\n",
       "             27: False,\n",
       "             28: False,\n",
       "             29: False,\n",
       "             30: False,\n",
       "             31: True,\n",
       "             32: False})"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reached_nodes_from_leaves"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pangenome-gene-transfer-simulation",
   "language": "python",
   "name": "pangenome-gene-transfer-simulation"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
