{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "14581ae4-b582-4e48-b551-0f8b9242a69a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clonal node: 2161\n",
      "Present_mutation at node:  2143\n",
      "Number of present genes:  57\n",
      "Success: hgt_rate = 2, Total computation time = 2.415885 seconds.\n",
      "HGT edges passed:  6\n",
      "[1653   72   22   16   14   14   13   10   10    8    6    6    6    6\n",
      "    4    4    4    4    4    4    4    4    4    4    4    4    3    3\n",
      "    3    3    3    3    3    3    3    3    3    3    3    3    3]\n",
      "Allel differences:  [[ 0.  3.  4. ... 92. 88. 84.]\n",
      " [ 3.  0.  1. ... 91. 87. 83.]\n",
      " [ 4.  1.  0. ... 92. 88. 84.]\n",
      " ...\n",
      " [92. 91. 92. ...  0. 22. 34.]\n",
      " [88. 87. 88. ... 22.  0. 30.]\n",
      " [84. 83. 84. ... 34. 30.  0.]]\n"
     ]
    }
   ],
   "source": [
    "import advanced_simulator_h5\n",
    "\n",
    "\n",
    "\n",
    "mts, gene_trees_list, gene_absence_presence_matrix, alleles_list_pca, scaled_distance_matrix, fitch_scores, gene_number_hgt_events_passed, core_allel_distance_list, core_allel_distance_number_of_clusters = advanced_simulator_h5.simulator(\n",
    "    num_samples = 1000, hgt_rate = 2, rho = 0\n",
    ")\n",
    "\n",
    "print(\"HGT edges passed: \", gene_number_hgt_events_passed[0])\n",
    "print(core_allel_distance_number_of_clusters[0])\n",
    "gene_present_vector = gene_absence_presence_matrix[0] == 1\n",
    "print(\"Allel differences: \", (squareform(pdist(alleles_list_pca[0]))**2)[np.ix_(gene_present_vector, gene_present_vector)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "48e1855d-c831-4ff7-b658-ac132c48ec13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([28,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,\n",
       "        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,\n",
       "        1,  1,  1,  1,  1,  1,  1])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "core_allel_distance_number_of_clusters[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8eac2d16-c612-4fe8-959c-424e75313afe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.   0.   0.05 0.05 0.05 0.05 0.05]\n",
      " [0.   0.   0.05 0.05 0.05 0.05 0.05]\n",
      " [0.05 0.05 0.   0.   0.   0.   0.  ]\n",
      " [0.05 0.05 0.   0.   0.   0.   0.  ]\n",
      " [0.05 0.05 0.   0.   0.   0.   0.  ]\n",
      " [0.05 0.05 0.   0.   0.   0.   0.  ]\n",
      " [0.05 0.05 0.   0.   0.   0.   0.  ]]\n",
      "[[0.   0.   0.06 0.04 0.02 0.02 0.02]\n",
      " [0.   0.   0.06 0.04 0.02 0.02 0.02]\n",
      " [0.06 0.06 0.   0.06 0.04 0.04 0.04]\n",
      " [0.04 0.04 0.06 0.   0.02 0.02 0.02]\n",
      " [0.02 0.02 0.04 0.02 0.   0.   0.  ]\n",
      " [0.02 0.02 0.04 0.02 0.   0.   0.  ]\n",
      " [0.02 0.02 0.04 0.02 0.   0.   0.  ]]\n",
      "[[ 0.    0.02 -0.02  0.01  0.03  0.03  0.03]\n",
      " [ 0.02  0.   -0.02  0.01  0.03  0.03  0.03]\n",
      " [-0.02 -0.02  0.   -0.05 -0.03 -0.03 -0.03]\n",
      " [ 0.01  0.01 -0.05  0.   -0.02 -0.02 -0.02]\n",
      " [ 0.03  0.03 -0.03 -0.02  0.    0.01  0.01]\n",
      " [ 0.03  0.03 -0.03 -0.02  0.01  0.    0.01]\n",
      " [ 0.03  0.03 -0.03 -0.02  0.01  0.01  0.  ]]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "np.set_printoptions(precision=2, suppress=True)\n",
    "\n",
    "gene_present_vector = gene_absence_presence_matrix[0] == 1\n",
    "\n",
    "distances = squareform(pdist(scaled_distance_matrix))[np.ix_(gene_present_vector, gene_present_vector)]\n",
    "alleles_distances = (squareform(pdist(alleles_list_pca[0]))**2)[np.ix_(gene_present_vector, gene_present_vector)] / np.mean((squareform(pdist(alleles_list_pca[0]))**2)[np.ix_(gene_present_vector, gene_present_vector)]) * np.mean(distances)\n",
    "\n",
    "print(distances)\n",
    "print(alleles_distances)\n",
    "\n",
    "print(core_allel_distance_list[0])\n",
    "\n",
    "#mts.draw_svg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2ca7970d-0356-44e7-8bd9-a6c499c16eb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 1., 3., 3., 3., 3., 3.],\n",
       "       [1., 0., 2., 2., 2., 2., 2.],\n",
       "       [3., 2., 0., 0., 2., 2., 2.],\n",
       "       [3., 2., 0., 0., 2., 2., 2.],\n",
       "       [3., 2., 2., 2., 0., 2., 2.],\n",
       "       [3., 2., 2., 2., 2., 0., 2.],\n",
       "       [3., 2., 2., 2., 2., 2., 0.]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(squareform(pdist(alleles_list_pca[0]))**2)[np.ix_(gene_present_vector, gene_present_vector)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pangenome-gene-transfer-simulation",
   "language": "python",
   "name": "pangenome-gene-transfer-simulation"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
